{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "422d2eb9-aae2-4a6a-920d-48feff2b073c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# adapted from:\n",
    "# https://ucdavis-bioinformatics-training.github.io/2018-June-RNA-Seq-Workshop/thursday/DE.html\n",
    "# https://rpubs.com/jrgonzalezISGlobal/transcriptomic_analyses\n",
    "# https://ucdavis-bioinformatics-training.github.io/2018-September-Bioinformatics-Prerequisites/friday/limma_biomart_vignettes.html\n",
    "# https://github.com/kevinblighe/EnhancedVolcano"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7062ba5a-31fd-455c-af18-93112ea0d081",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Dependencies\n",
    "\n",
    "suppressWarnings(library(edgeR))\n",
    "suppressWarnings(library(EnhancedVolcano))\n",
    "suppressWarnings(library(patchwork)) # combine plots\n",
    "suppressWarnings(library(magrittr))\n",
    "suppressWarnings(library(tibble))\n",
    "suppressWarnings(library(repr))\n",
    "suppressWarnings(library(stringr))\n",
    "suppressWarnings(library(dplyr))\n",
    "suppressWarnings(library(ggplot2))\n",
    "suppressWarnings(library(extrafont))\n",
    "suppressWarnings(library(svglite))\n",
    "\n",
    "suppressMessages(extrafont::font_import(pattern=\"Arial\",prompt=FALSE))\n",
    "suppressMessages(extrafont::loadfonts())\n",
    "\n",
    "sessionInfo()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "342b4325-8ef2-4ad9-bdef-9c7e21845cf4",
   "metadata": {},
   "source": [
    "# Data import and preprocessing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "689f10fd-1db9-40ed-8db1-d7b48d86a8e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Data\n",
    "dir.create(\"figures\", showWarnings = FALSE)\n",
    "dir.create(\"out\", showWarnings = FALSE)\n",
    "\n",
    "path_to_counts =  \"../data/cloud/gex/pbta-rsem-genes.expected_count.tsv\"\n",
    "path_to_annotation = \"../data/cloud/gex/sample_phenotypes.csv\"\n",
    "#path_to_annotation = \"out/sample_phenotypes.csv\"\n",
    "filter_genes <- function(gene_matrix){\n",
    "    patterns <- \"^ENSG|^LINC|-AS\\\\d$|^PAR_Y_|-DT$|-OT\\\\d$\"\n",
    "    rows_to_remove <- grep(patterns, rownames(gene_matrix))\n",
    "\n",
    "    # Subset the matrix to exclude these rows\n",
    "    filtered_data <- gene_matrix[-rows_to_remove, ]\n",
    "    return (filtered_data)\n",
    "}\n",
    "load_inputs <- function(cts_path,annot_path){\n",
    "    # Returns a list with named elements data$annot and data$cts\n",
    "    message(\"loading data...\")\n",
    "    cts = as.matrix(read.csv(cts_path,sep='\\t',row.names=\"Gene_or_Transcript_ID\",check.names=FALSE))\n",
    "    annot = read.csv(annot_path,row.names=1)\n",
    "\n",
    "    cts = round(cts) # counts must be integer\n",
    "    cts <- cts[, rownames(annot)] # only include samples with metadata\n",
    "    new_rownames <- sub(\"^ENSG\\\\d*\\\\.\\\\d*_\", \"\", rownames(cts)) # Remove prefixes ending with '_' from the row names\n",
    "    rownames(cts) <- new_rownames # Update the row names of the matrix\n",
    "    cts = filter_genes(cts) # remove common noncoding RNAs\n",
    "    annot$age_at_diagnosis <- (annot$age_at_diagnosis - mean(annot$age_at_diagnosis)) / sd(annot$age_at_diagnosis)\n",
    "    annot$amplified <- annot$amplicon_class %in% c('ecDNA','chromosomal')\n",
    "    annot$ecDNA <- annot$amplicon_class == 'ecDNA'\n",
    "    #annot <- annot[,c(\"sex\",\"tumor_history\",\"age_at_diagnosis\",\"extent_of_tumor_resection\",\"cancer_type\",\"amplicon_class\",'amplified','ecDNA')]\n",
    "\n",
    "    stopifnot(all(rownames(annot) == colnames(cts)))\n",
    "    return(list(cts=cts,annot=annot))\n",
    "}\n",
    "data=load_inputs(path_to_counts,path_to_annotation) # this will take awhile"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e6afefa9-b714-491b-8a4f-6f9da3cc9b9b",
   "metadata": {},
   "outputs": [],
   "source": [
    "head(data$annot)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4469e1e8-a93d-4ea3-93af-982660fbd680",
   "metadata": {},
   "outputs": [],
   "source": [
    "head(data$cts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6dafa16-eeec-49fe-b618-19c997ba3d47",
   "metadata": {},
   "outputs": [],
   "source": [
    "## lots of R pacakges work by creating a data *object* and performing matematical transformations on the structured data within.\n",
    "# https://www.rdocumentation.org/packages/edgeR/versions/3.14.0/topics/DGEList-class\n",
    "\n",
    "\n",
    "formula = ~ data$annot$cohort + data$annot$sex + data$annot$tumor_history + data$annot$age_at_diagnosis + data$annot$extent_of_tumor_resection + \n",
    "            data$annot$cancer_type + data$annot$amplicon_class\n",
    "formula = ~ data$annot$cohort + data$annot$sex + data$annot$tumor_history + data$annot$age_at_diagnosis + data$annot$extent_of_tumor_resection + \n",
    "            data$annot$cancer_type + data$annot$amplified + data$annot$ecDNA\n",
    "\n",
    "setup_preprocess_dge <- function(data,formula){\n",
    "    design <- model.matrix(formula)\n",
    "    dge <- DGEList(data$cts)\n",
    "    # remove low-expressed genes\n",
    "    keep <- filterByExpr(dge,design=design)\n",
    "    paste(\"Removing\",length(keep)-sum(keep),\"of\",length(keep),\"genes with low or invariant gene expression\") %>% message\n",
    "    dge <- dge[keep,]\n",
    "    \n",
    "    # Normalize by library size\n",
    "    dge <- calcNormFactors(dge,method=\"TMM\")\n",
    "    return(dge)\n",
    "}\n",
    "\n",
    "dge = setup_preprocess_dge(data,formula)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4a40d1fc-565f-4ca3-96cb-fb511bb82236",
   "metadata": {},
   "source": [
    "# Batch correction and outlier detection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c9a264b6-b618-4a6a-80ce-31a1b06d3134",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Looks like samples cluster by cohort, with a few odd outliers.\n",
    "options(repr.plot.width = 8, repr.plot.height = 6)\n",
    "colors <- c(\"blue\",\"red\",\"dark green\")[as.factor(data$annot$cohort)]\n",
    "mds = plotMDS(dge, gene.selection = \"common\", col = colors, pch=16)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "376115b0-a172-4fd7-bd03-b675de67888b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Corrected expression\n",
    "# NB: no need to log transform before correction; identical results.\n",
    "# PCs 1 and 2 driven by extreme values in a handful of outliers\n",
    "dgec <- removeBatchEffect(dge,batch=as.factor(data$annot$cohort))\n",
    "options(repr.plot.width = 8, repr.plot.height = 6)\n",
    "colors <- c(\"blue\",\"red\",\"dark green\")[as.factor(data$annot$cohort)]\n",
    "mdsc = plotMDS(dgec, gene.selection = \"common\", col = colors, pch=16)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cb555ba8-fb7e-41e3-ab52-6e9e8bf65441",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "mx = median(mdsc$x)\n",
    "sx = sd(mdsc$x)\n",
    "my = median(mdsc$y)\n",
    "sy = sd(mdsc$y)\n",
    "outlier_mask <- (mdsc$y < my-3*sy) | (mdsc$y > my+3*sy) | (mdsc$x < mx-3*sx) | (mdsc$x > mx+3*sx)\n",
    "paste(\"Identified\",sum(outlier_mask),\"outlier samples\")\n",
    "data$annot[outlier_mask,]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d8a15676-f6f6-4be3-9af5-fa2ce382c1d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Corrected expression sans outliers\n",
    "# Cohorts now overlap, but PC1 describes cohort 1 variance and PC2 describes cohort 2.\n",
    "options(repr.plot.width = 8, repr.plot.height = 6)\n",
    "colors <- c(\"blue\",\"red\",\"dark green\")[as.factor(data$annot$cohort[!outlier_mask])]\n",
    "mdsc = plotMDS(dgec[,!outlier_mask], gene.selection = \"common\", col = colors, pch=16)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8056f93d-7fda-421b-8318-c2886dc08322",
   "metadata": {},
   "outputs": [],
   "source": [
    "data$annot[which(!(rownames(data$annot) %in% outliers)),] %>% dim"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb94249a-2fb3-4adc-b729-2b7f35a30880",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Regenerate the dataset sans outliers\n",
    "# Outliers are hardcoded here if you don't want to run the previous lines of code\n",
    "if (exists(\"outlier_mask\")){\n",
    "    outliers <- data$annot[outlier_mask,] %>% rownames\n",
    "} else {\n",
    "    outliers=c('7316-13','7316-1082','7316-2744','7316-3645','7316-170','7316-9449','7316-1089','7316-290','7316-2144','7316-1886',\n",
    "               '7316-3028','7316-2589','7316-2071','7316-3657','7316-3061','7316-477','7316-2614','7316-2138','7316-291')\n",
    "}\n",
    "data$cts <- data$cts[,which(!(colnames(data$cts) %in% outliers))] \n",
    "data$annot <- data$annot[which(!(rownames(data$annot) %in% outliers)),]\n",
    "formula = ~ data$annot$cohort + data$annot$sex + data$annot$tumor_history + data$annot$age_at_diagnosis + data$annot$extent_of_tumor_resection + \n",
    "            data$annot$cancer_type + data$annot$amplified + data$annot$ecDNA\n",
    "dge = setup_preprocess_dge(data,formula)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "779855d7-7ce9-4568-aaa8-9e6a05a876ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "fit_dge_lm <- function(dge,design){\n",
    "    # This takes a minute\n",
    "    options(repr.plot.width=7, repr.plot.height=7)\n",
    "    message(\"Fitting voom normalization...\")\n",
    "    v <- voom(dge, design, plot=TRUE)\n",
    "    \n",
    "    message(\"Fitting linear model...\")\n",
    "    fit <- lmFit(v,design)\n",
    "    \n",
    "    message(\"Calculating emperical bayes statistics...\")\n",
    "    fit <- eBayes(fit,robust=TRUE)\n",
    "    return(fit)\n",
    "}\n",
    "fit <- fit_dge_lm(dge,design)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "24331bb2-763f-4729-8e76-4e5c34be8eb2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Apply multiple testing correction and obtain stats\n",
    "ecDNA_comparison = ncol(design)\n",
    "amp_comparison = ncol(design)-1\n",
    "get_de_genes <- function(fit,comparison_index){\n",
    "    print(paste(\"Getting DE genes w.r.t.\",colnames(design)[comparison_index]))\n",
    "    stats_df <- topTable(fit,n=Inf,coef=comparison_index) %>% tibble\n",
    "    print(stats_df %>% head)\n",
    "    return(stats_df)\n",
    "}\n",
    "ec_stats_df = get_de_genes(fit,ecDNA_comparison)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f2cbc4b8-57a7-494e-abf4-6e6ba4f42490",
   "metadata": {},
   "outputs": [],
   "source": [
    "amp_stats_df = get_de_genes(fit,amp_comparison)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4d6e5903-e913-4d9c-839f-a18750e2d1e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Seal, R.L., Denny, P., Bruford, E.A. et al. A standardized nomenclature for mammalian histone genes. Epigenetics & Chromatin 15, 34 \n",
    "# (2022). https://doi.org/10.1186/s13072-022-00467-2\n",
    "# changes in v2: H1-1, H1-2, H1-3, H1-4, H1-5, H1-6, H4-16 to rdh\n",
    "#                \n",
    "\n",
    "H_pseudogenes = c('H1-9P','H1-12P',\n",
    "                  'H2AC2P','H2AC3P','H2AC5P','H2AC9P','H2AC10P','H2AQ1P','H2AL1MP',\n",
    "                  'H2BC2P','H2BC16P','H2BC19P','H2BC20P','H2BU2P','H2BL1P', # H2BU2P = H2BC27P\n",
    "                  'H2BW3P',#'H2BW4P', # note H2BW4P and aliases not in gex dataset\n",
    "                  'H3C5P','H3C9P','H3P16','H3P44',\n",
    "                  'H4C10P',\n",
    "                  'H2AZP7','H2AZ2P1','H2AZP2','H2AZP5','H2BP2','H2BP3','H2BP9','H4P1','H2BP1','H2AL1QP','H3P18','H3P21','H3P3','H3P1','H3P4',\n",
    "                  'H3P37','H3P31','H3P14','H3P47','H3P6','H3P39','H3P13',\n",
    "                  \"H2AZP1\",\"H2AZP3\",\"H3P28\",\"H3P27\",\"H2ACP1\",\"H2ACP2\",\"H3P26\" ,\"H2BP7\",\"H3P43\",\"H2BP6\",\"H3P45\",\"H3P9\",\"H3P30\",\"H3P29\",\"H3P10\",\"H3P11\",\n",
    "                  \"H3P12\",\"H3P5\",\"H2AZP6\",\"H3P32\",\"H2BP8\",\"H3P36\",\"H3P7\",\"H3P2\",\"H3P17\",\"H3P20\",\"H3P23\",\"H3P22\",\"H3P15\",\"H3P24\",\"H3P34\",\"H2AZP4\",\n",
    "                  \"H3P33\",\"H3P35\",\"H3P40\",\"H3P38\",\"H3P41\",\"H3P8\",\"H3P19\",\"H3P42\",\"H3P25\",\"H2BP5\",\"H3P46\" ) \n",
    "H1 = c('H1-0','H1-7','H1-8','H1-10')\n",
    "H1_clustered = c('H1-1','H1-2','H1-3','H1-4','H1-5','H1-6')\n",
    "H2A_clustered = c('H2AC1','H2AC4','H2AC6','H2AC7','H2AC8','H2AC11','H2AC12',\n",
    "                'H2AC13','H2AC14','H2AC15','H2AC16','H2AC17','H2AC18','H2AC19','H2AC20','H2AC21','H2AW') # H2AW = H2AC25\n",
    "H2A = c('H2AZ1','H2AZ2','MACROH2A1','MACROH2A2','H2AX','H2AJ','H2AB1','H2AB2','H2AB3','H2AP','H2AL3')\n",
    "H2B_clustered = c('H2BC1','H2BC3','H2BC4','H2BC5','H2BC6','H2BC7','H2BC8','H2BC9','H2BC10','H2BC11','H2BC12',\n",
    "                 'H2BC13','H2BC14','H2BC15','H2BC17','H2BC18','H2BC21','H2BU1') # H2BU1 = H2BC26\n",
    "H2B = c('H2BE1','H2BW1','H2BW2','H2BS1') # H2BS1 = H2BC12L; H2BE1 = H2BK1; H2BN1 and aliases not in gex dataset\n",
    "H3_clustered = c('H3C1','H3C2','H3C3','H3C4','H3C6','H3C7','H3C8','H3C10','H3C11','H3C12','H3C13','H3C14','H3C15',\n",
    "                 'H3-4') # H3-4 has a stem loop and is in a cluster\n",
    "H3 = c('H3-3A','H3-3B','H3-5','H3-2','H3Y1','H3Y2','CENPA') # H3-2 = H3-7\n",
    "H4_clustered = c('H4C1','H4C2','H4C3','H4C4','H4C5','H4C6','H4C7','H4C8','H4C9','H4C11','H4C12','H4C13','H4C14','H4C15',\n",
    "                'H4-16') # H4-16 = H4C16 is outside the clusters but has a stem-loop and is cell-cycle regulated.\n",
    "H4 = c()\n",
    "rdh = c(H1_clustered,H2A_clustered,H2B_clustered,H3_clustered,H4_clustered)\n",
    "nrdh = c(H1,H2A,H2B,H3,H4)\n",
    "all_histones = c(rdh,nrdh,H_pseudogenes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f64e8dcb-fee1-4010-9c4f-aa100ae0b448",
   "metadata": {},
   "outputs": [],
   "source": [
    "# gene set validation checks\n",
    "paste(\"replication-dependent histone count:\",length(rdh))\n",
    "paste(\"replication-independent histone count:\",length(nrdh))\n",
    "paste(\"histone pseudogene count:\",length(H_pseudogenes))\n",
    "\n",
    "# all genes in my sets should be in my gene expression dataset\n",
    "genes <- data$cts %>% rownames\n",
    "check_set <- function(name,set){\n",
    "    pass = all(v %in% genes)\n",
    "    print(paste(name,'...',pass))\n",
    "    if (!pass){\n",
    "        missing = setdiff(v,genes)\n",
    "        print(paste(\"  missing\",missing,\"from gene expression data\"))\n",
    "    }\n",
    "    return(pass)\n",
    "}\n",
    "vectors = list(\"H1ri\"=H1,\n",
    "          \"H2Ari\"=H2A,\n",
    "          \"H2Bri\"=H2B,\n",
    "          \"H3ri\"=H3,\n",
    "          \"H4ri\"=H4,\n",
    "          \"H1rd\"=H1_clustered,\n",
    "          \"H2Ard\"=H2A_clustered,\n",
    "          \"H2Brd\"=H2B_clustered,\n",
    "          \"H3rd\"=H3_clustered,\n",
    "          \"H4rd\"=H3_clustered,\n",
    "          \"HP\"=H_pseudogenes\n",
    "    )\n",
    "print(\"Checking each histone class for genes not in expression dataset...\")\n",
    "flags=c()\n",
    "for (name in names(vectors)){\n",
    "    v = vectors[[name]]\n",
    "    pass = check_set(name,v)\n",
    "    if (!pass){\n",
    "        flag=c(flag,name)\n",
    "    }\n",
    "}\n",
    "\n",
    "print(\"Checking for possible unannotated histone genes in expression dataset...\")\n",
    "regex = \"^H(1|2|3|4)|^HIST\"\n",
    "candidates = grep(regex,genes,value=TRUE)\n",
    "candidates = setdiff(candidates,all_histones)\n",
    "not_histones=c('H19')\n",
    "candidates = setdiff(candidates,not_histones)\n",
    "print(candidates)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f710f120-99c9-41af-a294-f3cf76aab57e",
   "metadata": {},
   "outputs": [],
   "source": [
    "hits <- ec_stats_df %>% \n",
    "    #filter(adj.P.Val < 0.10) %>%\n",
    "    arrange(desc(logFC))\n",
    "hits %>% head(n=30)\n",
    "write.table(hits, file='out/differential_expression_ecDNA.tsv',quote=FALSE,sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4c0adf7b-10ec-400a-bee7-d6371ed45fb2",
   "metadata": {},
   "outputs": [],
   "source": [
    "hits %>% tail(n=30)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "91674dc8-5987-4de3-8b26-0fe84df0aacb",
   "metadata": {},
   "outputs": [],
   "source": [
    "write_gene_sets <- function(){\n",
    "    vectors = list(rdh, nrdh, H_pseudogenes)\n",
    "    names = c(\"replication_dependent_histones\",\"replication_independent_histones\",\"histone_pseudogenes\")\n",
    "    # Open a connection to the output file\n",
    "    file_conn <- file(\"out/histone-sets.gmt\", open = \"wt\")\n",
    "    # Write each set to the file\n",
    "    for (i in seq_along(names)) {\n",
    "      # Combine the name, description, and vector into a single row\n",
    "      row <- c(names[i], \"custom_gene_set\", vectors[[i]])\n",
    "      # Write the row to the file, separated by tabs\n",
    "      writeLines(paste(row, collapse = \"\\t\"), file_conn)\n",
    "    }\n",
    "    \n",
    "    # Close the file connection\n",
    "    close(file_conn)\n",
    "}\n",
    "write_gene_sets()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "53a80a58-23e3-4ac3-8d17-5766bbeb9338",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Plotting code\n",
    "\n",
    "color_code <- list(\n",
    "  \"replication-dependent histones\" = \"blue\",\n",
    "  \"replication-independent histones\" = \"darkgreen \",\n",
    "  \"histone pseudogenes\" = \"darkorange\",\n",
    "  \"other\" = \"grey50\",\n",
    "  'c-NHEJ' = 'royalblue3',\n",
    "  'Alt-EJ' = 'darkblue',\n",
    "  'SSA' = 'blue3',\n",
    "  'HR' = 'purple4',\n",
    "  'HOXA@' = 'black',\n",
    "  'HOXB@' = 'cyan',\n",
    "  'HOXC@' = 'mediumorchid4',\n",
    "  'HOXD@' = 'slateblue4'\n",
    ")\n",
    "ylabel=expression(-Log[10]*\"(\"*italic(q)*\")\")\n",
    "\n",
    "base_theme <- theme_classic(base_size=7, base_family=\"Arial\",) +\n",
    "    theme(axis.text = element_text(size=7,colour=\"black\"))\n",
    "theme_set(base_theme)\n",
    "\n",
    "osc_volcano_ii <- function(stats_df,gene_set,gene_set_name){\n",
    "    stats_df$highlight = stats_df$ID %in% gene_set\n",
    "    stats_df <- stats_df[order(stats_df$highlight),]\n",
    "    colCustom <- c(ifelse(stats_df$highlight, color_code[[gene_set_name]], \"grey50\"))\n",
    "    names(colCustom) <- c(ifelse(stats_df$highlight, gene_set_name, \"other\"))\n",
    "    plt <- EnhancedVolcano(stats_df,\n",
    "                    lab = stats_df$'ID',\n",
    "                    title = NULL,\n",
    "                    subtitle = NULL,\n",
    "                    caption = NULL,\n",
    "                    axisLabSize = 14,\n",
    "                    x = 'logFC',\n",
    "                    y = \"adj.P.Val\",\n",
    "                    #xlim = c(-2.75,2.75),\n",
    "                    #ylim = c(0,4),\n",
    "                    pCutoff = 0.05,\n",
    "                    FCcutoff = 10,\n",
    "                    labSize=0,\n",
    "                    pointSize = c(ifelse(stats_df$highlight, 3, 1)),\n",
    "                    colCustom = colCustom,\n",
    "                    colAlpha = .6\n",
    "                    ) %>% suppressWarnings\n",
    "    options(repr.plot.width=18, repr.plot.height=7)\n",
    "    return(plt + ylab(ylabel)) #+ lims(x=c(0,4),y=c(-4,4))\n",
    "}\n",
    "\n",
    "osc_volcano_v <- function(stats_df){\n",
    "    highlight <- c('GFAP','HOXA6','HOXA9','HOXA10','S100B','XRCC4','CCNB1','CCNA2','CIP2A','FEN1','MYCN',\n",
    "                  'RAD51','SMC2','RAD17','PRIM2','TIMELESS','CDK7','CDK1',\n",
    "                  #'CENPK','AURKA','CENPN','CENPL',\n",
    "                  'H3C15','H2AW','H2BU1', 'H3C13')\n",
    "    stats_df$highlight = ifelse(stats_df$ID %in% highlight,\"notable significant\",\"other\")\n",
    "    stats_df <- stats_df[order(stats_df$highlight=='other',decreasing=TRUE),]\n",
    "    stats_df$color <- ifelse(stats_df$highlight=='other','grey50','black')\n",
    "    sapply(stats_df$highlight, function(x) color_code[[x]])\n",
    "       \n",
    "    plt <- EnhancedVolcano(stats_df,\n",
    "                lab = stats_df$'ID',\n",
    "                title = NULL,\n",
    "                subtitle = NULL,\n",
    "                caption = NULL,\n",
    "                axisLabSize = 14,\n",
    "                x = 'logFC',\n",
    "                y = \"adj.P.Val\",\n",
    "                #xlim = c(-2.75,2.75),\n",
    "                ylim = c(0,4),\n",
    "                pCutoff = 0.05,\n",
    "                selectLab = highlight,\n",
    "                labSize = 4.21644413212,#3.37315530569,#\n",
    "                FCcutoff = 10,\n",
    "                vline = NULL, \n",
    "                vlineType = \"blank\",\n",
    "                legendPosition = \"none\",\n",
    "                pointSize = c(ifelse(stats_df$highlight == \"other\", 1, 3)),\n",
    "                colCustom = setNames(stats_df$color,stats_df$highlight),\n",
    "                drawConnectors = TRUE,\n",
    "                maxoverlapsConnectors = Inf,\n",
    "                lengthConnectors = unit(0, \"npc\"),\n",
    "                colAlpha = .6\n",
    "                ) %>% suppressWarnings\n",
    "    options(repr.plot.width=5, repr.plot.height=5)\n",
    "    return(plt + ylab(ylabel))\n",
    "}\n",
    "\n",
    "write_plot <- function(plt,outfile,width,height){\n",
    "    pdf.options(encoding='ISOLatin2.enc')\n",
    "    pdfName = paste(outfile, \".pdf\", sep=\"\")\n",
    "    pngName = paste(outfile, \".png\", sep=\"\")\n",
    "    svgName = paste(outfile, \".svg\", sep = \"\")\n",
    "    ggsave(path=\"figures\", filename=pdfName, device=\"pdf\", width=width, height=height, units='in')\n",
    "    ggsave(path=\"figures\", device=\"png\", filename=pngName, width=width, height=height, units='in')\n",
    "    ggsave(path=\"figures\", device=\"svg\", filename=svgName, width=width, height=height, units='in')\n",
    "\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb406591-cb66-4b25-8975-f62b178f1919",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read the tab-separated file\n",
    "read_gmt = function(file_path){\n",
    "    data <- readLines(file_path)\n",
    "    \n",
    "    # Initialize an empty list to store named vectors\n",
    "    named_vectors <- list()\n",
    "    \n",
    "    # Process each line\n",
    "    for (line in data) {\n",
    "      # Split the line by tab\n",
    "      parts <- strsplit(line, \"\\t\")[[1]]\n",
    "      \n",
    "      # The first part is the vector name\n",
    "      vector_name <- parts[1]\n",
    "      \n",
    "      # The third part and onwards are the vector values\n",
    "      vector_values <- (parts[3:length(parts)])\n",
    "      \n",
    "      # Create the named vector and add it to the list\n",
    "      named_vectors[[vector_name]] <- vector_values\n",
    "    }\n",
    "    return(named_vectors)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "96e6d33f-78db-4cca-bb1b-fba25f8f791e",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(warn=-1)\n",
    "plt <- osc_volcano_v(ec_stats_df)\n",
    "w=14.4/4;h=3.55\n",
    "options(repr.plot.width=2*w, repr.plot.height=2*h)\n",
    "#write_plot(plt,\"selected_volcano\",w,h)\n",
    "plt\n",
    "options(warn=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "92c33065-83a8-4a18-b406-c993a69ea943",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(warn=-1)\n",
    "p1 = osc_volcano_ii(ec_stats_df,rdh,'replication-dependent histones')\n",
    "p2 = osc_volcano_ii(ec_stats_df,nrdh,'replication-independent histones')\n",
    "p3 = osc_volcano_ii(ec_stats_df,H_pseudogenes, 'histone pseudogenes')\n",
    "p4 = osc_volcano_ii(amp_stats_df,rdh,'replication-dependent histones')\n",
    "p5 = osc_volcano_ii(amp_stats_df,nrdh,'replication-independent histones')\n",
    "p6 = osc_volcano_ii(amp_stats_df,H_pseudogenes, 'histone pseudogenes')\n",
    "plt <- (p1+p2+p3)/(p4+p5+p6)\n",
    "w=14.4/4*3;h=9\n",
    "options(repr.plot.width=w, repr.plot.height=h)\n",
    "#write_plot(plt,\"rdh_volcano\",w,h)\n",
    "plt\n",
    "options(warn=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d2e21d80-fc54-4eac-91ed-0e55b7025d01",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(warn=-1)\n",
    "dsbr_sets = read_gmt(\"out/dsbr-sets.gmt\")\n",
    "plt <- osc_volcano_ii(stats_df,dsbr_sets[[\"c-NHEJ\"]],'c-NHEJ') |\n",
    "osc_volcano_ii(stats_df,dsbr_sets[[\"Alt-EJ\"]],'Alt-EJ') |\n",
    "osc_volcano_ii(stats_df,dsbr_sets[[\"SSA\"]], 'SSA') | \n",
    "osc_volcano_ii(stats_df,dsbr_sets[[\"HR\"]], 'HR')\n",
    "w=14.4;h=4.5\n",
    "options(repr.plot.width=w, repr.plot.height=h)\n",
    "#write_plot(plt,\"dsbr_volcano\",w,h)\n",
    "\n",
    "plt \n",
    "options(warn=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "788e01a3-ba91-4d0d-9c7f-088b5efcce08",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(warn=-1)\n",
    "hox_sets = read_gmt(\"out/hox-sets.gmt\")\n",
    "plt <- osc_volcano_ii(stats_df,hox_sets[[\"HOXA@\"]],'HOXA@') |\n",
    "osc_volcano_ii(stats_df,hox_sets[[\"HOXB@\"]],'HOXB@') |\n",
    "osc_volcano_ii(stats_df,hox_sets[[\"HOXC@\"]], 'HOXC@') | \n",
    "osc_volcano_ii(stats_df,hox_sets[[\"HOXD@\"]], 'HOXD@')\n",
    "w=14.4;h=4.5\n",
    "options(repr.plot.width=w, repr.plot.height=h)\n",
    "#write_plot(plt,\"hox_volcano\",w,h)\n",
    "\n",
    "plt\n",
    "options(warn=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "355e2fdf-5a56-4469-b5f7-a9b6462bde0b",
   "metadata": {},
   "source": [
    "# Old plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e92b868e-893f-479b-8bab-19d746125a54",
   "metadata": {},
   "outputs": [],
   "source": [
    "osc_volcano_i <- function(stats_df,gene_set,gene_set_name){\n",
    "    highlight <- stats_df$ID %in% gene_set\n",
    "    plt <- EnhancedVolcano(stats_df,\n",
    "                    lab = stats_df$'ID',\n",
    "                    title = NULL,\n",
    "                    subtitle = NULL,\n",
    "                    caption = NULL,\n",
    "                    axisLabSize = 14,\n",
    "                    x = 'logFC',\n",
    "                    y = \"adj.P.Val\",\n",
    "                    xlim = c(-3,3),\n",
    "                    ylim = c(0,3),\n",
    "                    pCutoff = 0.05,\n",
    "                    drawConnectors = TRUE,\n",
    "                    maxoverlapsConnectors = Inf,\n",
    "                    #lengthConnectors = unit(2, \"npc\"),\n",
    "                    selectLab = gene_set,\n",
    "                    pointSize = c(ifelse(highlight, 3, 1)),\n",
    "                    )\n",
    "    options(repr.plot.width=18, repr.plot.height=7)\n",
    "    return(plt)\n",
    "}\n",
    "\n",
    "osc_volcano_iii <- function(stats_df){\n",
    "    stats_df$highlight = ifelse(stats_df$ID %in% rdh,\"replication-dependent histones\",\n",
    "                                ifelse(stats_df$ID %in% nrdh, \"canonical histones\",\n",
    "                                       ifelse(stats_df$ID %in% H_pseudogenes, \"histone pseudogenes\", \"other\")))\n",
    "    stats_df <- stats_df[order(stats_df$highlight=='other',decreasing=TRUE),]\n",
    "    stats_df$color <- sapply(stats_df$highlight, function(x) color_code[[x]])\n",
    "                             \n",
    "    plt <- EnhancedVolcano(stats_df,\n",
    "                    lab = stats_df$'ID',\n",
    "                    title = NULL,\n",
    "                    subtitle = NULL,\n",
    "                    caption = NULL,\n",
    "                    axisLabSize = 14,\n",
    "                    x = 'logFC',\n",
    "                    y = \"adj.P.Val\",\n",
    "                    xlim = c(-2.75,2.75),\n",
    "                    ylim = c(0,3),\n",
    "                    pCutoff = 0.05,\n",
    "                    FCcutoff = 10,\n",
    "                    labSize=0,\n",
    "                    pointSize = c(ifelse(stats_df$highlight == \"other\", 1, 3)),\n",
    "                    colCustom = setNames(stats_df$color,stats_df$highlight)\n",
    "                    )\n",
    "    options(repr.plot.width=8, repr.plot.height=9)\n",
    "    return(plt + ylab(ylabel)) #+ lims(x=c(0,4),y=c(-4,4))\n",
    "}\n",
    "\n",
    "osc_volcano_iv <- function(stats_df){\n",
    "    plt <- EnhancedVolcano(stats_df,\n",
    "                lab = stats_df$'ID',\n",
    "                title = NULL,\n",
    "                subtitle = NULL,\n",
    "                caption = NULL,\n",
    "                axisLabSize = 14,\n",
    "                x = 'logFC',\n",
    "                y = \"adj.P.Val\",\n",
    "                xlim = c(-2.75,2.75),\n",
    "                ylim = c(0,3),\n",
    "                pCutoff = 0.05,\n",
    "                drawConnectors = TRUE,\n",
    "                maxoverlapsConnectors = Inf,\n",
    "                lengthConnectors = unit(0, \"npc\"),   \n",
    "                )\n",
    "    options(repr.plot.width=12, repr.plot.height=12)\n",
    "    return(plt)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "73d01d91-29b7-4fe0-a20f-fe8707c0b70a",
   "metadata": {},
   "outputs": [],
   "source": [
    "osc_volcano_i(ec_stats_df,rdh) +\n",
    "osc_volcano_i(ec_stats_df,nrdh) + \n",
    "osc_volcano_i(ec_stats_df,H_pseudogenes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "00c132ca-a80d-43a0-bc4e-974021dcfee7",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt <- osc_volcano_iv(ec_stats_df)\n",
    "#write_plot(plt,\"sig_volcano\",11,11)\n",
    "plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "67edf283-ff11-41b8-9963-9fe263d69594",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "differential-expression",
   "language": "R",
   "name": "differential-expression"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.4.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
