{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "ba342204-889e-4415-b6fd-4983b2b3c9dc",
   "metadata": {},
   "source": [
    "# create inputs for gsea\n",
    "env: py3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9df18a22-2858-4457-8b80-8c979f86f111",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import scipy.spatial.distance\n",
    "import sys\n",
    "from pathlib import Path\n",
    "from collections import OrderedDict\n",
    "\n",
    "sys.path.append('../src')\n",
    "Path(\"out\").mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "import data_imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0280be98-963a-4e52-9f90-9fea735ef998",
   "metadata": {},
   "outputs": [],
   "source": [
    "def clean_cbtn_rnaseq(df):\n",
    "    # Drop ENSG prefixes \n",
    "    df.index = df.index.map(lambda x: '_'.join(x.split('_')[1:]))\n",
    "    # We want these two ENSG entries (C19MC) but no others\n",
    "    #save = df.loc[['ENSG00000269842','ENSG00000269564']]\n",
    "    # drop Drop entries without HUGO entries or with nonunique names.\n",
    "    df = df[~ (df.index.str.startswith('ENSG') | df.index.str.startswith('PAR_Y') | df.index.str.startswith('LINC'))]\n",
    "    #f = pd.concat([df,save])\n",
    "    # drop duplicates\n",
    "    df = df[~ df.index.duplicated(False)]\n",
    "    # Samples in the cohort\n",
    "    samples = data_imports.import_biosamples().external_sample_id\n",
    "    df = df.loc[:,df.columns.isin(samples)]\n",
    "    return df\n",
    "\n",
    "def import_cbtn_rnaseq(path='/Users/ochapman/Library/CloudStorage/OneDrive-SanfordBurnhamPrebysMedicalDiscoveryInstitute/projects/2023-pedpancan/data/gex/CBTN-gex.tpm.tsv'):\n",
    "    df = pd.read_csv(path,sep='\\t',index_col=0)\n",
    "    df=clean_cbtn_rnaseq(df)\n",
    "    return df\n",
    "\n",
    "def import_biosamples():\n",
    "    df = data_imports.import_biosamples()\n",
    "    # preprocessing\n",
    "    df['amplicon_class']=df['amplicon_class'].replace(\n",
    "        {'Complex noncyclic':'chromosomal',\n",
    "         'BFB':'chromosomal',\n",
    "         'Linear':'chromosomal',\n",
    "         'No amplification':'no_amp',\n",
    "        })\n",
    "    # drop duplicate samples\n",
    "    df=df.dropna(subset='external_sample_id')\n",
    "    return df\n",
    "\n",
    "def import_genes():\n",
    "    df = data_imports.import_genes()\n",
    "    # preprocessing\n",
    "    df['feature']=df['feature'].map(lambda x: x.split('_')[0]) # drop suffix\n",
    "    df['feature']=df['feature'].replace(\n",
    "        {'Complex-non-cyclic':'chromosomal',\n",
    "         'BFB':'chromosomal',\n",
    "         'Linear':'chromosomal',\n",
    "         'No amplification':'no_amp',\n",
    "    })\n",
    "    # unknown in this table are not reported in other AC results afaik\n",
    "    df=df[df.feature != 'unknown']\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fe5e7ee1-6901-47d1-a36c-d80af3b2fc76",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GLOBALS\n",
    "\n",
    "RNA=import_cbtn_rnaseq()\n",
    "BIOSAMPLES=import_biosamples()\n",
    "GENES = import_genes()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1818a9d6-7336-4c4a-86a0-90f3ff3669f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "RNA.head()\n",
    "#BIOSAMPLES.amplicon_class.unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "80e4d1e1-1bfc-40a7-bb8d-ab9dfa6c6350",
   "metadata": {},
   "source": [
    "# create GSEA inputs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e53dec40-9cab-4527-8d83-3dd89b44e00f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def exp2gct(df,outfile):\n",
    "    '''\n",
    "    Format a table of expression values as gct. See https://software.broadinstitute.org/cancer/software/gsea/wiki/index.php/Data_formats\n",
    "    Input df: samples on column names, genes on row names.\n",
    "    Returns: None\n",
    "    '''\n",
    "    # checks\n",
    "    df=df.copy()\n",
    "    assert ~df.columns.duplicated().all()\n",
    "    assert ~df.index.duplicated().all()\n",
    "    # format table\n",
    "    df.index.name = 'Name'\n",
    "    df.insert(loc=0,column='Description',value='')\n",
    "    # write\n",
    "    with open(outfile,'w') as f:\n",
    "        f.write('#1.2\\n')\n",
    "        f.write(f'{len(df.index)}\\t{len(df.columns)-1}\\n')\n",
    "        df.to_csv(f,sep='\\t')\n",
    "    return\n",
    "\n",
    "def labels2cls(series,outfile):\n",
    "    # checks\n",
    "    series=series.copy()\n",
    "    assert(~series.index.duplicated().all())\n",
    "    # format table\n",
    "    df = pd.DataFrame(series).T\n",
    "    # write\n",
    "    with open(outfile,'w') as f:\n",
    "        a = df.iloc[0].unique()\n",
    "        b = df.iloc[0].values\n",
    "        f.write(f'{len(b)} {len(a)} 1\\n')\n",
    "        f.write(f'# {\" \".join(a)}\\n')\n",
    "        f.write(\" \".join(b))\n",
    "    return\n",
    "        \n",
    "\n",
    "\n",
    "def map_ecDNA_status(iterable):\n",
    "    '''\n",
    "    Get ecDNA status for an iterable of samples\n",
    "    Inputs:\n",
    "        iterable: iterable of sample names in the form '7316-XXXX'\n",
    "    Returns:\n",
    "        Series of {sample:status} where status in ['ecDNA','chromosomal','No amplification']\n",
    "    '''\n",
    "    d=OrderedDict()\n",
    "    for i in iterable:\n",
    "        assert i not in d.keys()\n",
    "        bs = BIOSAMPLES[BIOSAMPLES.external_sample_id == i].amplicon_class\n",
    "        for j in ['ecDNA','chromosomal','no_amp']:\n",
    "            if j in bs.values:\n",
    "                d[i] = j\n",
    "                continue\n",
    "    return pd.Series(d)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e46c3e15-e2bd-46cf-a1ad-72a1e32908d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def prepare_gsea_comparison(phenotype1, phenotype2):\n",
    "    d = map_ecDNA_status(RNA.columns)\n",
    "    d = d[d.isin([phenotype1, phenotype2])]\n",
    "    exp = RNA[d.index]\n",
    "    labels = pd.Series(data=exp.columns.map(lambda x: d[x]),index=exp.columns)\n",
    "    exp2gct(exp,f'out/{phenotype1}-{phenotype2}.gct')\n",
    "    labels2cls(labels,f'out/{phenotype1}-{phenotype2}.cls')\n",
    "    return exp, labels\n",
    "\n",
    "def prepare_ecDNA_no_amp_gsea():\n",
    "    return prepare_gsea_comparison('ecDNA','no_amp')\n",
    "\n",
    "def prepare_ecDNA_chromosomal_gsea():\n",
    "    return prepare_gsea_comparison('ecDNA','chromosomal')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dac08766-e9ba-49c2-901a-9a1c1272d521",
   "metadata": {},
   "outputs": [],
   "source": [
    "exp, labels = prepare_ecDNA_chromosomal_gsea()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f42c12bb-de11-4db4-bf4f-3277de3c0d9c",
   "metadata": {},
   "outputs": [],
   "source": [
    "exp, labels = prepare_ecDNA_no_amp_gsea()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7a3a256e-ccc8-4bca-98d7-ac39dcd361b2",
   "metadata": {},
   "source": [
    "## DSBR gene sets from Miin Lin et al., 2024\n",
    "https://elifesciences.org/articles/88895/figures#supp1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7312c062-e810-4cde-a4a6-04d52f9d1c17",
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_lin_etal(file='external/elife-88895-supp1-v1.xlsx',sheet='1L'):\n",
    "    # get this file from https://doi.org/10.7554/eLife.88895.3\n",
    "    df = pd.read_excel(file,sheet_name=sheet)\n",
    "    return df\n",
    "def write_gene_sets(set_dict,outfile):\n",
    "    # dict of the form setname: {set of gene names}\n",
    "    with open(outfile,'w') as f:\n",
    "        for key,value in set_dict.items():\n",
    "            line = ' '.join([key, \"NA\"]+list(value)+['\\n'])\n",
    "            f.write(line)\n",
    "    return\n",
    "def get_gene_sets_from_lin1L(df,outfile):\n",
    "    # parse gene names\n",
    "    df[\"Gene\"] = df.Gene.map(lambda x: x.split('|')[0])\n",
    "    # Use the gene symbols in my expression dataset\n",
    "    df[\"Gene\"] = df.Gene.replace({\n",
    "        \"H2AFX\":\"H2AX\",\n",
    "        \"H3F3A\":\"H3-3A\",\n",
    "        \"BTBD12\":\"SLX4\",\n",
    "        \"FAM175A\":\"ABRAXAS1\",\n",
    "        \"FAM35A\":\"SHLD2\",\n",
    "        \"GIYD2\":\"SLX1B\",\n",
    "        \"HDGFRP2\":\"HDGFL2\",\n",
    "        \"MRE11A\":\"MRE11\",\n",
    "        \"RAD51L1\":\"RAD51B\",\n",
    "        \"RAD51L3\":\"RAD51D\",\n",
    "        \"C16orf75\":\"RMI2\",\n",
    "        \"C20orf196\":\"SHLD1\",\n",
    "        \"C20orf20\":\"MRGBP\",\n",
    "        \"C3orf37\":\"HMCES\",\n",
    "        \"C5orf45\":\"MRNIP\",\n",
    "        \"C9orf142\":\"PAXX\"\n",
    "    })\n",
    "    d = {}\n",
    "    for pathway in [\"c-NHEJ\",\"Alt-EJ\",\"SSA\",\"HR\"]:\n",
    "        d[pathway] = set(df['Gene'][~df[pathway].isna()])\n",
    "    write_gene_sets(d,outfile)\n",
    "    return d\n",
    "\n",
    "df = load_lin_etal()\n",
    "dsbr = get_gene_sets_from_lin1L(df,\"out/dsbr-sets.gmt\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2f8b10a7-8c5e-4b51-b99e-10fe21bf8d2a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# DSBR genes not in my expression dataset\n",
    "set(df.Gene.values) - set(RNA.index)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b7e70919-63ac-4221-a32f-ad5df37522fe",
   "metadata": {},
   "source": [
    "# metaanalysis of GSEA inputs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0f159a55-f0ed-4c0b-bfd3-ba0ab4ec1499",
   "metadata": {},
   "outputs": [],
   "source": [
    "def count_tumor_types(labels):\n",
    "    labels = pd.DataFrame(labels,columns=[\"amp_status\"])\n",
    "    labels['tumor_type'] = labels.index.map(lambda x: BIOSAMPLES[BIOSAMPLES.external_sample_id == x][\"cancer_type\"][0])\n",
    "    return pd.crosstab(labels.amp_status,  labels.tumor_type)\n",
    "\n",
    "count_tumor_types(labels)    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fdee7a3b-fec4-48ee-9c7a-32d40b7f4e65",
   "metadata": {},
   "source": [
    "# metaanalysis of GSEA outputs\n",
    "\n",
    "Many of the gene set hits for enrichment in ecDNA+ are replication-dependent histones, H[1-4]C[1-20]. Literature search says they're tightly controlled and expressed in S phase during DNA replication. ecDNA+ cells undergoing cell cycle more often?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "af4d6d14-3f9c-47ff-b306-0c5d0ef85d1b",
   "metadata": {},
   "outputs": [],
   "source": [
    "HERE='/Users/ochapman/SBP/projects/pedpancan_ecdna/2024-04-19_RNAseq'\n",
    "def import_gsea_result(path1,path2):\n",
    "    df1 = pd.read_csv(path1,sep='\\t',index_col=0)\n",
    "    df2 = pd.read_csv(path2,sep='\\t',index_col=0)\n",
    "    return pd.concat([df1,df2]).sort_values('NOM p-val')\n",
    "def get_ec_up_vs_chromosomal():\n",
    "    path1=f'{HERE}/results/ec_ch_go_584576/gsea_report_for_ecDNA_1717712143472.tsv'\n",
    "    path2=f'{HERE}/results/ec_ch_other_584575/gsea_report_for_ecDNA_1717712138408.tsv'\n",
    "    return import_gsea_result(path1,path2)\n",
    "def get_ec_dn_vs_chromosomal():\n",
    "    path1=f'{HERE}/results/ec_ch_go_584576/gsea_report_for_chromosomal_1717712143472.tsv'\n",
    "    path2=f'{HERE}/results/ec_ch_other_584575/gsea_report_for_chromosomal_1717712138408.tsv'\n",
    "    return import_gsea_result(path1,path2)\n",
    "def get_ec_up_vs_no_amp():\n",
    "    path1=f'{HERE}/results/ec_na_go_584573/gsea_report_for_ecDNA_1717712072261.tsv'\n",
    "    path2=f'{HERE}/results/ec_na_other_584574/gsea_report_for_ecDNA_1717712095569.tsv'\n",
    "    return import_gsea_result(path1, path2)\n",
    "def get_ec_dn_vs_no_amp():\n",
    "    path1=f'{HERE}/results/ec_na_go_584573/gsea_report_for_no_amp_1717712072261.tsv'\n",
    "    path2=f'{HERE}/results/ec_na_other_584574/gsea_report_for_no_amp_1717712095569.tsv'\n",
    "    return import_gsea_result(path1,path2)\n",
    "\n",
    "def print_summary_statistics(df,name):\n",
    "    print(f'Summary statistics for comparison {name}:')\n",
    "    print(f'Number of tests: {len(df)}')\n",
    "    print(f'p < 0.05: {(df[\"NOM p-val\"] < 0.05).sum()}')\n",
    "    print(f'p < 0.01: {(df[\"NOM p-val\"] < 0.01).sum()}')\n",
    "    print(f'FDR q < 0.25: {(df[\"FDR q-val\"] < 0.25).sum()}')\n",
    "    print(f'FWER p < 0.25 {(df[\"FWER p-val\"] < 0.25).sum()}')\n",
    "    print('')\n",
    "    return\n",
    "\n",
    "def summary():\n",
    "    print_summary_statistics(get_ec_up_vs_chromosomal(),\"ecDNA up vs. chromosomal\")\n",
    "    print_summary_statistics(get_ec_dn_vs_chromosomal(),\"ecDNA down vs. chromosomal\")\n",
    "    print_summary_statistics(get_ec_up_vs_no_amp(),\"ecDNA up vs. no amp\")\n",
    "    print_summary_statistics(get_ec_dn_vs_no_amp(),\"ecDNA down vs. no amp\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a4ac6042-5950-4c16-8303-7026971d9a77",
   "metadata": {},
   "outputs": [],
   "source": [
    "summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8cbaa6ea-7e7e-4659-bd06-cc292d0f4e67",
   "metadata": {},
   "outputs": [],
   "source": [
    "def text_venn(set1,set2):\n",
    "    print(f\"Size of set 1: {len(set1)}.\\n Size of set 2: {len(set2)}.\\n Intersection: {len(set1 & set2)}.\")\n",
    "    print(f\"In set 1 but not set 2: {set1 - set2}\\n\")\n",
    "    print(f\"In both: {set1 & set2}\\n\")\n",
    "    #print(f\"In set 2 but not set 1: {set2 - set1}\\n\")\n",
    "\n",
    "def venn_hits(direction, metric, threshold):\n",
    "    '''\n",
    "    Examine GSEA results.\n",
    "    Inputs:\n",
    "        direction: should be \"up\" (overexpressed in ecDNA) or \"down\" (underexpressed)\n",
    "        metric: one of \"NOM p-val\", \"FDR q-val\", \"FWER p-val\"\n",
    "        threshold: float 0 < t < 1\n",
    "    '''\n",
    "    if direction == \"up\":\n",
    "        a = get_ec_up_vs_chromosomal()\n",
    "        b = get_ec_up_vs_no_amp()\n",
    "    elif direction == \"down\":\n",
    "        a = get_ec_dn_vs_chromosomal()\n",
    "        b = get_ec_dn_vs_no_amp()\n",
    "    else:\n",
    "        raise Exception(\"direction should be up or down\")\n",
    "    a = set(a[a[metric] < threshold].index)\n",
    "    b = set(b[b[metric] < threshold].index)\n",
    "    text_venn(a,b)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f35660cb-5301-484d-9941-d33e75e02b60",
   "metadata": {},
   "outputs": [],
   "source": [
    "venn_hits(\"up\",\"NOM p-val\",0.05)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2350319c-bd68-455e-b2be-e132abc62d90",
   "metadata": {},
   "outputs": [],
   "source": [
    "#venn_hits(\"down\",\"NOM p-val\", 0.05)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9511d745-b30a-4448-9493-5b32b6209d2f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_gmt(files):\n",
    "    d = {}\n",
    "    for file in files:\n",
    "        with open(file,'r') as f:\n",
    "            for line in f:\n",
    "                line = line.split()\n",
    "                d[line[0]] = set(line[2:])\n",
    "    \n",
    "    return d\n",
    "\n",
    "def jaccard(a,b):\n",
    "    n = len(a & b)\n",
    "    m = len(a | b)\n",
    "    return n/m\n",
    "\n",
    "def go_jaccard_matrix(metric=\"NOM p-val\",threshold=0.05,go=True,include_histones=False):\n",
    "    \n",
    "    # get sets\n",
    "    a = get_ec_up_vs_chromosomal()\n",
    "    a = set(a[a[metric] < threshold].index)\n",
    "    b = get_ec_up_vs_no_amp()\n",
    "    b = set(b[b[metric] < threshold].index)\n",
    "    if go:\n",
    "        a = set(s for s in a if s.startswith('GO'))\n",
    "        b = set(s for s in b if s.startswith('GO'))\n",
    "    if include_histones:\n",
    "        for s in ['replication_dependent_histones','constitutive_histones','histone_pseudogenes']:\n",
    "            a.add(s)\n",
    "            b.add(s)\n",
    "    # get union, set up distance matrix\n",
    "    union = list(a & b)\n",
    "    d = np.empty((len(union),len(union)))\n",
    "\n",
    "    # compute distances\n",
    "    for i in range(len(union)):\n",
    "        for j in range(len(union)):\n",
    "            i_members = go_sets[union[i]]\n",
    "            j_members = go_sets[union[j]]\n",
    "            d[i][j] = 1 - jaccard(i_members,j_members)\n",
    "\n",
    "    # plot\n",
    "    d = pd.DataFrame(d,index=union,columns=union)\n",
    "    linkage = scipy.cluster.hierarchy.linkage(scipy.spatial.distance.squareform(d), method='average')\n",
    "    plot = sns.clustermap(d, row_linkage=linkage, col_linkage=linkage,cmap='mako')\n",
    "    return plot\n",
    "    \n",
    "def savefig(plot,basename):\n",
    "    pdfName = basename + \".pdf\"\n",
    "    pngName = basename + \".png\"\n",
    "    svgName = basename + \".svg\"\n",
    "    plot.savefig(pdfName,format='pdf')\n",
    "    plot.savefig(pngName,format='png')\n",
    "    plot.savefig(svgName,format='svg')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1cf9cbb9-d239-45fd-8647-c046e3477b48",
   "metadata": {},
   "outputs": [],
   "source": [
    "gmts = [\n",
    "    '/Users/ochapman/SBP/projects/pedpancan_ecdna/2024-04-19_RNAseq/results/ec_ch_go_584576/gene_sets.gmt',\n",
    "]\n",
    "go_sets = parse_gmt(gmts)\n",
    "plt.rcParams['svg.fonttype'] = 'none'\n",
    "sns.set_theme(font='Arial',font_scale=2)\n",
    "plot = go_jaccard_matrix(go=True,include_histones=False)\n",
    "savefig(plot,'figures/go_jaccard_matrix')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2edfda97-4ab6-4f76-8eb3-d5d629f95a9f",
   "metadata": {},
   "outputs": [],
   "source": [
    "gmts = [\n",
    "    '/Users/ochapman/projects/pedpancan_ecdna/2024-04-19_RNAseq/results/ec_ch_go_584576/gene_sets.gmt',\n",
    "    '/Users/ochapman/projects/pedpancan_ecdna/2024-04-19_RNAseq/results/ec_ch_other_584575/gene_sets.gmt',\n",
    "    '/Users/ochapman/projects/pedpancan_ecdna/2024-04-19_RNAseq/out/dsbr-sets.gmt',\n",
    "    '/Users/ochapman/projects/pedpancan_ecdna/2024-04-19_RNAseq/out/histone-sets.gmt',\n",
    "]\n",
    "go_sets = parse_gmt(gmts)\n",
    "plot = go_jaccard_matrix(go=False,include_histones=False)\n",
    "\n",
    "savefig(plot,'figures/jaccard_matrix')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "py3",
   "language": "python",
   "name": "py3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
