{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "422d2eb9-aae2-4a6a-920d-48feff2b073c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# adapted from:\n",
    "# https://ucdavis-bioinformatics-training.github.io/2018-June-RNA-Seq-Workshop/thursday/DE.html\n",
    "# https://rpubs.com/jrgonzalezISGlobal/transcriptomic_analyses\n",
    "# https://ucdavis-bioinformatics-training.github.io/2018-September-Bioinformatics-Prerequisites/friday/limma_biomart_vignettes.html\n",
    "# https://github.com/kevinblighe/EnhancedVolcano"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c53639f1-4678-4a13-b3f6-0cbef7558479",
   "metadata": {},
   "outputs": [],
   "source": [
    "sessionInfo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "803ddef4-0098-4221-9ee7-1fd952a0bad0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Dependencies\n",
    "\n",
    "suppressWarnings(library(edgeR))\n",
    "suppressWarnings(library(EnhancedVolcano))\n",
    "suppressWarnings(library(patchwork)) # combine plots\n",
    "suppressWarnings(library(magrittr))\n",
    "suppressWarnings(library(tibble))\n",
    "suppressWarnings(library(repr))\n",
    "suppressWarnings(library(stringr))\n",
    "suppressWarnings(library(dplyr))\n",
    "suppressWarnings(library(ggplot2))\n",
    "suppressWarnings(library(extrafont))\n",
    "suppressWarnings(library(svglite))\n",
    "\n",
    "suppressMessages(extrafont::font_import(pattern=\"Arial\",prompt=FALSE))\n",
    "suppressMessages(extrafont::loadfonts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "689f10fd-1db9-40ed-8db1-d7b48d86a8e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Data\n",
    "dir.create(\"figures\", showWarnings = FALSE)\n",
    "dir.create(\"out\", showWarnings = FALSE)\n",
    "\n",
    "path_to_counts =  \"/Users/ochapman/Library/CloudStorage/OneDrive-SanfordBurnhamPrebysMedicalDiscoveryInstitute/projects/2023-pedpancan/data/gex/pbta-rsem-genes.expected_count.tsv\"\n",
    "path_to_annotation = \"/Users/ochapman/Library/CloudStorage/OneDrive-SanfordBurnhamPrebysMedicalDiscoveryInstitute/projects/2023-pedpancan/data/gex/sample_phenotypes.csv\"\n",
    "path_to_annotation = \"out/sample_phenotypes.csv\"\n",
    "filter_genes <- function(gene_matrix){\n",
    "    patterns <- \"^ENSG|^LINC|-AS\\\\d$|^PAR_Y_|-DT$|-OT\\\\d$\"\n",
    "    rows_to_remove <- grep(patterns, rownames(gene_matrix))\n",
    "\n",
    "    # Subset the matrix to exclude these rows\n",
    "    filtered_data <- gene_matrix[-rows_to_remove, ]\n",
    "    return (filtered_data)\n",
    "}\n",
    "load_inputs <- function(cts_path,annot_path){\n",
    "    message(\"loading data...\")\n",
    "    cts = as.matrix(read.csv(cts_path,sep='\\t',row.names=\"Gene_or_Transcript_ID\",check.names=FALSE))\n",
    "    annot = read.csv(annot_path,row.names=1)\n",
    "\n",
    "    cts = round(cts) # counts must be integer\n",
    "    cts <- cts[, rownames(annot)] # only include samples with metadata\n",
    "    new_rownames <- sub(\"^ENSG\\\\d*\\\\.\\\\d*_\", \"\", rownames(cts)) # Remove prefixes ending with '_' from the row names\n",
    "    rownames(cts) <- new_rownames # Update the row names of the matrix\n",
    "    cts = filter_genes(cts) # remove common noncoding RNAs\n",
    "    annot$age_at_diagnosis <- (annot$age_at_diagnosis - mean(annot$age_at_diagnosis)) / sd(annot$age_at_diagnosis) \n",
    "    annot <- annot[,c(\"sex\",\"tumor_history\",\"age_at_diagnosis\",\"extent_of_tumor_resection\",\"cancer_type\",\"amplicon_class\")]\n",
    "\n",
    "    stopifnot(all(rownames(annot) == colnames(cts)))\n",
    "    return(list(cts=cts,annot=annot))\n",
    "}\n",
    "data=load_inputs(path_to_counts,path_to_annotation) # this will take awhile"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4469e1e8-a93d-4ea3-93af-982660fbd680",
   "metadata": {},
   "outputs": [],
   "source": [
    "head(data$cts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6dafa16-eeec-49fe-b618-19c997ba3d47",
   "metadata": {},
   "outputs": [],
   "source": [
    "## lots of R pacakges work by creating a data *object* and performing matematical transformations on the structured data within.\n",
    "# https://www.rdocumentation.org/packages/edgeR/versions/3.14.0/topics/DGEList-class\n",
    "dge <- DGEList(data$cts)\n",
    "\n",
    "formula = ~ data$annot$sex + data$annot$tumor_history + data$annot$age_at_diagnosis + data$annot$extent_of_tumor_resection + data$annot$cancer_type + data$annot$amplicon_class\n",
    "#formula = ~ data$annot$extent_of_tumor_resection + data$annot$cancer_type + data$annot$amplicon_class\n",
    "design <- model.matrix(formula)\n",
    "\n",
    "# https://rdrr.io/bioc/edgeR/man/filterByExpr.html\n",
    "keep <- filterByExpr(dge,design=design)\n",
    "dge <- dge[keep,]\n",
    "\n",
    "# gene counts from RNA-seq are influenced by technical factors such as library size and sequencing depth, which do not reflect true\n",
    "# biological variation. There are various algorithms to *normalize* RNA-seq data. Here we use trimmed mean of M values, TMM.\n",
    "# https://www.rdocumentation.org/packages/edgeR/versions/3.14.0/topics/calcNormFactors\n",
    "dge <- calcNormFactors(dge,method=\"TMM\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "973f393b-4049-472f-a3a9-8fb793f38f37",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# This takes a minute\n",
    "options(repr.plot.width=7, repr.plot.height=7)\n",
    "message(\"Fitting voom normalization...\")\n",
    "v <- voom(dge, design, plot=TRUE)\n",
    "\n",
    "message(\"Fitting linear model...\")\n",
    "fit <- lmFit(v,design)\n",
    "\n",
    "message(\"Calculating emperical bayes statistics...\")\n",
    "fit <- eBayes(fit,robust=TRUE)\n",
    "\n",
    "# Apply multiple testing correction and obtain stats\n",
    "comparison = ncol(design)-1\n",
    "print(colnames(design)[comparison]) # This is the \n",
    "stats_df <- topTable(fit,n=Inf,coef=comparison) %>% tibble\n",
    "print(dim(dge))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4d6e5903-e913-4d9c-839f-a18750e2d1e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Seal, R.L., Denny, P., Bruford, E.A. et al. A standardized nomenclature for mammalian histone genes. Epigenetics & Chromatin 15, 34 \n",
    "# (2022). https://doi.org/10.1186/s13072-022-00467-2\n",
    "H_pseudogenes = c('H1-9P','H1-12P','H2AC2P','H2AC3P','H2AC5P','H2AC9P','H2AC10P','H2AQ1P','H2AL1MP','H2BC2P','H2BC19P','H2BC20P',\n",
    "                 'H2BC16P','H2BL1P','H2BU2P','H2BW3P','H2BW4P','H3C5P','H3C9P','H3P16','H3P44','H4C10P',\n",
    "                 'H2AZP7','H2AZ2P1','H2AZP2','H2AZP5','H2BP2','H2BP3','H2BP9','H4P1','H2BP1','H2AL1QP','H3P18','H3P21','H3P3','H3P1','H3P4',\n",
    "                 'H3P37','H3P31','H3P14','H3P47','H3P6','H3P39','H3P13') # H2BU2P = H2BC27P\n",
    "H1 = c('H1-0','H1-1','H1-2','H1-3','H1-4','H1-5','H1-6','H1-7','H1-8','H1-10')\n",
    "H2A_clustered = c('H2AC1','H2AC4','H2AC6','H2AC7','H2AC8','H2AC11','H2AC12',\n",
    "                'H2AC13','H2AC14','H2AC15','H2AC16','H2AC17','H2AC18','H2AC19','H2AC20','H2AC21','H2AW') # H2AW = H2AC25\n",
    "H2A = c('H2AZ1','H2AZ2','MACROH2A1','MACROH2A2','H2AX','H2AJ','H2AB1','H2AB2','H2AB3','H2AP','H2AL3')\n",
    "H2B_clustered = c('H2BC1','H2BC3','H2BC4','H2BC5','H2BC6','H2BC7','H2BC8','H2BC9','H2BC10','H2BC11','H2BC12',\n",
    "                 'H2BC13','H2BC14','H2BC15','H2BC17','H2BC18','H2BC21','H2BU1') # H2BU1 = H2BC26\n",
    "H2B = c('H2BE1','H2BW1','H2BW2','H2BN1','H2BS1') # H2BS1 = H2BC12L; H2BE1 = H2BK1\n",
    "H3_clustered = c('H3C1','H3C2','H3C3','H3C4','H3C6','H3C7','H3C8','H3C10','H3C11','H3C12','H3C13','H3C14','H3C15')\n",
    "H3 = c('H3-3A','H3-3B','H3-4','H3-5','H3-2','H3Y1','H3Y2','CENPA') # H3-2 = H3-7\n",
    "H4_clustered = c('H4C1','H4C2','H4C3','H4C4','H4C5','H4C6','H4C7','H4C8','H4C9','H4C11','H4C12','H4C13','H4C14','H4C15')\n",
    "H4 = c('H4-16')\n",
    "rdh = c(H2A_clustered,H2B_clustered,H3_clustered,H4_clustered)\n",
    "nrdh = c(H1,H2A,H2B,H3,H4)\n",
    "all_histones = c(rdh,nrdh,H_pseudogenes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f710f120-99c9-41af-a294-f3cf76aab57e",
   "metadata": {},
   "outputs": [],
   "source": [
    "hits <- stats_df %>% \n",
    "    #filter(adj.P.Val < 0.10) %>%\n",
    "    arrange(logFC)\n",
    "write.table(hits, file='out/differential_expression.tsv',quote=FALSE,sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "73ada897-0a40-418b-a752-0bfd145ecda5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# These histones are missing in the expression dataset\n",
    "all_histones[!(all_histones %in% stats_df$ID)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "344849d2-12ea-4d6a-896d-bb43a91e6f48",
   "metadata": {},
   "outputs": [],
   "source": [
    "# These genes are not in my histone sets\n",
    "stats_df$ID[str_starts(stats_df$ID,'H[1-4]') & !(stats_df$ID %in% all_histones)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "91674dc8-5987-4de3-8b26-0fe84df0aacb",
   "metadata": {},
   "outputs": [],
   "source": [
    "write_gene_sets <- function(){\n",
    "    vectors = list(rdh, nrdh, H_pseudogenes)\n",
    "    names = c(\"replication_dependent_histones\",\"constitutive_histones\",\"histone_pseudogenes\")\n",
    "    # Open a connection to the output file\n",
    "    file_conn <- file(\"out/histone-sets.gmt\", open = \"wt\")\n",
    "    # Write each set to the file\n",
    "    for (i in seq_along(names)) {\n",
    "      # Combine the name, description, and vector into a single row\n",
    "      row <- c(names[i], \"custom_gene_set\", vectors[[i]])\n",
    "      # Write the row to the file, separated by tabs\n",
    "      writeLines(paste(row, collapse = \"\\t\"), file_conn)\n",
    "    }\n",
    "    \n",
    "    # Close the file connection\n",
    "    close(file_conn)\n",
    "}\n",
    "write_gene_sets()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "53a80a58-23e3-4ac3-8d17-5766bbeb9338",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Plotting code\n",
    "\n",
    "color_code <- list(\n",
    "  \"replication-dependent histones\" = \"blue\",\n",
    "  \"canonical histones\" = \"darkgreen \",\n",
    "  \"histone pseudogenes\" = \"darkorange\",\n",
    "  \"other\" = \"grey50\",\n",
    "  'c-NHEJ' = 'royalblue3',\n",
    "  'Alt-EJ' = 'darkblue',\n",
    "  'SSA' = 'blue3',\n",
    "  'HR' = 'purple4',\n",
    "  'HOXA@' = 'black',\n",
    "  'HOXB@' = 'cyan',\n",
    "  'HOXC@' = 'mediumorchid4',\n",
    "  'HOXD@' = 'slateblue4'\n",
    ")\n",
    "ylabel=expression(-Log[10]*\"(\"*italic(q)*\")\")\n",
    "\n",
    "base_theme <- theme_classic(base_size=7, base_family=\"Arial\",) +\n",
    "    theme(axis.text = element_text(size=7,colour=\"black\"))\n",
    "theme_set(base_theme)\n",
    "\n",
    "osc_volcano_i <- function(stats_df,gene_set,gene_set_name){\n",
    "    highlight <- stats_df$ID %in% gene_set\n",
    "    plt <- EnhancedVolcano(stats_df,\n",
    "                    lab = stats_df$'ID',\n",
    "                    title = NULL,\n",
    "                    subtitle = NULL,\n",
    "                    caption = NULL,\n",
    "                    axisLabSize = 14,\n",
    "                    x = 'logFC',\n",
    "                    y = \"adj.P.Val\",\n",
    "                    xlim = c(-3,3),\n",
    "                    ylim = c(0,3),\n",
    "                    pCutoff = 0.05,\n",
    "                    drawConnectors = TRUE,\n",
    "                    maxoverlapsConnectors = Inf,\n",
    "                    lengthConnectors = unit(2, \"npc\"),\n",
    "                    selectLab = gene_set,\n",
    "                    pointSize = c(ifelse(highlight, 3, 1)),\n",
    "                    )\n",
    "    options(repr.plot.width=18, repr.plot.height=7)\n",
    "    return(plt)\n",
    "}\n",
    "osc_volcano_ii <- function(stats_df,gene_set,gene_set_name){\n",
    "    stats_df$highlight = stats_df$ID %in% gene_set\n",
    "    stats_df <- stats_df[order(stats_df$highlight),]\n",
    "    colCustom <- c(ifelse(stats_df$highlight, color_code[[gene_set_name]], \"grey50\"))\n",
    "    names(colCustom) <- c(ifelse(stats_df$highlight, gene_set_name, \"other\"))\n",
    "    plt <- EnhancedVolcano(stats_df,\n",
    "                    lab = stats_df$'ID',\n",
    "                    title = NULL,\n",
    "                    subtitle = NULL,\n",
    "                    caption = NULL,\n",
    "                    axisLabSize = 14,\n",
    "                    x = 'logFC',\n",
    "                    y = \"adj.P.Val\",\n",
    "                    xlim = c(-2.75,2.75),\n",
    "                    ylim = c(0,3),\n",
    "                    pCutoff = 0.05,\n",
    "                    FCcutoff = 10,\n",
    "                    labSize=0,\n",
    "                    pointSize = c(ifelse(stats_df$highlight, 3, 1)),\n",
    "                    colCustom = colCustom,\n",
    "                    colAlpha = .6\n",
    "                    ) %>% suppressWarnings\n",
    "    options(repr.plot.width=18, repr.plot.height=7)\n",
    "    return(plt + ylab(ylabel)) #+ lims(x=c(0,4),y=c(-4,4))\n",
    "}\n",
    "osc_volcano_iii <- function(stats_df){\n",
    "    stats_df$highlight = ifelse(stats_df$ID %in% rdh,\"replication-dependent histones\",\n",
    "                                ifelse(stats_df$ID %in% nrdh, \"canonical histones\",\n",
    "                                       ifelse(stats_df$ID %in% H_pseudogenes, \"histone pseudogenes\", \"other\")))\n",
    "    stats_df <- stats_df[order(stats_df$highlight=='other',decreasing=TRUE),]\n",
    "    stats_df$color <- sapply(stats_df$highlight, function(x) color_code[[x]])\n",
    "                             \n",
    "    plt <- EnhancedVolcano(stats_df,\n",
    "                    lab = stats_df$'ID',\n",
    "                    title = NULL,\n",
    "                    subtitle = NULL,\n",
    "                    caption = NULL,\n",
    "                    axisLabSize = 14,\n",
    "                    x = 'logFC',\n",
    "                    y = \"adj.P.Val\",\n",
    "                    xlim = c(-2.75,2.75),\n",
    "                    ylim = c(0,3),\n",
    "                    pCutoff = 0.05,\n",
    "                    FCcutoff = 10,\n",
    "                    labSize=0,\n",
    "                    pointSize = c(ifelse(stats_df$highlight == \"other\", 1, 3)),\n",
    "                    colCustom = setNames(stats_df$color,stats_df$highlight)\n",
    "                    )\n",
    "    options(repr.plot.width=8, repr.plot.height=9)\n",
    "    return(plt + ylab(ylabel)) #+ lims(x=c(0,4),y=c(-4,4))\n",
    "}\n",
    "\n",
    "osc_volcano_iv <- function(stats_df){\n",
    "    plt <- EnhancedVolcano(stats_df,\n",
    "                lab = stats_df$'ID',\n",
    "                title = NULL,\n",
    "                subtitle = NULL,\n",
    "                caption = NULL,\n",
    "                axisLabSize = 14,\n",
    "                x = 'logFC',\n",
    "                y = \"adj.P.Val\",\n",
    "                xlim = c(-2.75,2.75),\n",
    "                ylim = c(0,3),\n",
    "                pCutoff = 0.05,\n",
    "                drawConnectors = TRUE,\n",
    "                maxoverlapsConnectors = Inf,\n",
    "                lengthConnectors = unit(0, \"npc\"),   \n",
    "                )\n",
    "    options(repr.plot.width=12, repr.plot.height=12)\n",
    "    return(plt)\n",
    "}\n",
    "\n",
    "osc_volcano_v <- function(stats_df){\n",
    "    highlight <- c('GFAP','HOXA6','HOXA9','HOXA10','S100B','XRCC4','CCNB1','CCNA2','CIP2A','FEN1','MYCN',\n",
    "                  'RAD51','SMC2','RAD17','PRIM2','TIMELESS','CDK7','CDK1',\n",
    "                  #'CENPK','AURKA','CENPN','CENPL',\n",
    "                  'H3C15','H2AW','H2BU1', 'H3C13')\n",
    "    stats_df$highlight = ifelse(stats_df$ID %in% highlight,\"notable significant\",\"other\")\n",
    "    stats_df <- stats_df[order(stats_df$highlight=='other',decreasing=TRUE),]\n",
    "    stats_df$color <- ifelse(stats_df$highlight=='other','grey50','black')\n",
    "    sapply(stats_df$highlight, function(x) color_code[[x]])\n",
    "       \n",
    "    plt <- EnhancedVolcano(stats_df,\n",
    "                lab = stats_df$'ID',\n",
    "                title = NULL,\n",
    "                subtitle = NULL,\n",
    "                caption = NULL,\n",
    "                axisLabSize = 14,\n",
    "                x = 'logFC',\n",
    "                y = \"adj.P.Val\",\n",
    "                xlim = c(-2.75,2.75),\n",
    "                ylim = c(0,3),\n",
    "                pCutoff = 0.05,\n",
    "                selectLab = highlight,\n",
    "                labSize = 4.21644413212,#3.37315530569,#\n",
    "                FCcutoff = 10,\n",
    "                vline = NULL, \n",
    "                vlineType = \"blank\",\n",
    "                legendPosition = \"none\",\n",
    "                pointSize = c(ifelse(stats_df$highlight == \"other\", 1, 3)),\n",
    "                colCustom = setNames(stats_df$color,stats_df$highlight),\n",
    "                drawConnectors = TRUE,\n",
    "                maxoverlapsConnectors = Inf,\n",
    "                lengthConnectors = unit(0, \"npc\"),\n",
    "                colAlpha = .6\n",
    "                ) %>% suppressWarnings\n",
    "    options(repr.plot.width=5, repr.plot.height=5)\n",
    "    return(plt + ylab(ylabel))\n",
    "}\n",
    "\n",
    "write_plot <- function(plt,outfile,width,height){\n",
    "    pdf.options(encoding='ISOLatin2.enc')\n",
    "    pdfName = paste(outfile, \".pdf\", sep=\"\")\n",
    "    pngName = paste(outfile, \".png\", sep=\"\")\n",
    "    svgName = paste(outfile, \".svg\", sep = \"\")\n",
    "    ggsave(path=\"figures\", filename=pdfName, device=\"pdf\", width=width, height=height, units='in')\n",
    "    ggsave(path=\"figures\", device=\"png\", filename=pngName, width=width, height=height, units='in')\n",
    "    ggsave(path=\"figures\", device=\"svg\", filename=svgName, width=width, height=height, units='in')\n",
    "\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb406591-cb66-4b25-8975-f62b178f1919",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read the tab-separated file\n",
    "read_gmt = function(file_path){\n",
    "    data <- readLines(file_path)\n",
    "    \n",
    "    # Initialize an empty list to store named vectors\n",
    "    named_vectors <- list()\n",
    "    \n",
    "    # Process each line\n",
    "    for (line in data) {\n",
    "      # Split the line by tab\n",
    "      parts <- strsplit(line, \"\\t\")[[1]]\n",
    "      \n",
    "      # The first part is the vector name\n",
    "      vector_name <- parts[1]\n",
    "      \n",
    "      # The third part and onwards are the vector values\n",
    "      vector_values <- (parts[3:length(parts)])\n",
    "      \n",
    "      # Create the named vector and add it to the list\n",
    "      named_vectors[[vector_name]] <- vector_values\n",
    "    }\n",
    "    return(named_vectors)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "96e6d33f-78db-4cca-bb1b-fba25f8f791e",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(warn=-1)\n",
    "plt <- osc_volcano_v(stats_df)\n",
    "w=14.4/4;h=3.55\n",
    "options(repr.plot.width=2*w, repr.plot.height=2*h)\n",
    "write_plot(plt,\"selected_volcano\",w,h)\n",
    "plt\n",
    "options(warn=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "92c33065-83a8-4a18-b406-c993a69ea943",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(warn=-1)\n",
    "plt <- osc_volcano_ii(stats_df,rdh,'replication-dependent histones') +\n",
    "osc_volcano_ii(stats_df,nrdh,'canonical histones') + \n",
    "osc_volcano_ii(stats_df,H_pseudogenes, 'histone pseudogenes') %>% suppressWarnings\n",
    "w=14.4/4*3;h=4.5\n",
    "options(repr.plot.width=w, repr.plot.height=h)\n",
    "write_plot(plt,\"rdh_volcano\",w,h)\n",
    "plt\n",
    "options(warn=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d2e21d80-fc54-4eac-91ed-0e55b7025d01",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(warn=-1)\n",
    "dsbr_sets = read_gmt(\"out/dsbr-sets.gmt\")\n",
    "plt <- osc_volcano_ii(stats_df,dsbr_sets[[\"c-NHEJ\"]],'c-NHEJ') |\n",
    "osc_volcano_ii(stats_df,dsbr_sets[[\"Alt-EJ\"]],'Alt-EJ') |\n",
    "osc_volcano_ii(stats_df,dsbr_sets[[\"SSA\"]], 'SSA') | \n",
    "osc_volcano_ii(stats_df,dsbr_sets[[\"HR\"]], 'HR')\n",
    "w=14.4;h=4.5\n",
    "options(repr.plot.width=w, repr.plot.height=h)\n",
    "write_plot(plt,\"dsbr_volcano\",w,h)\n",
    "\n",
    "plt \n",
    "options(warn=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "788e01a3-ba91-4d0d-9c7f-088b5efcce08",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(warn=-1)\n",
    "hox_sets = read_gmt(\"out/hox-sets.gmt\")\n",
    "plt <- osc_volcano_ii(stats_df,hox_sets[[\"HOXA@\"]],'HOXA@') |\n",
    "osc_volcano_ii(stats_df,hox_sets[[\"HOXB@\"]],'HOXB@') |\n",
    "osc_volcano_ii(stats_df,hox_sets[[\"HOXC@\"]], 'HOXC@') | \n",
    "osc_volcano_ii(stats_df,hox_sets[[\"HOXD@\"]], 'HOXD@')\n",
    "w=14.4;h=4.5\n",
    "options(repr.plot.width=w, repr.plot.height=h)\n",
    "write_plot(plt,\"hox_volcano\",w,h)\n",
    "\n",
    "plt\n",
    "options(warn=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "355e2fdf-5a56-4469-b5f7-a9b6462bde0b",
   "metadata": {},
   "source": [
    "# Old plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "73d01d91-29b7-4fe0-a20f-fe8707c0b70a",
   "metadata": {},
   "outputs": [],
   "source": [
    "osc_volcano_i(stats_df,rdh) +\n",
    "osc_volcano_i(stats_df,nrdh) + \n",
    "osc_volcano_i(stats_df,H_pseudogenes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "00c132ca-a80d-43a0-bc4e-974021dcfee7",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt <- osc_volcano_iv(stats_df)\n",
    "#write_plot(plt,\"sig_volcano\",11,11)\n",
    "plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "67edf283-ff11-41b8-9963-9fe263d69594",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "differential-expression",
   "language": "R",
   "name": "differential-expression"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
