{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "0dbf1dcf-a16a-4c9e-90a3-617716bd9de6",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['BS_85Y70BHE', 'BS_A9S5HT6P', 'BS_1EQGGHHQ', 'BS_BR1X3R68',\n",
       "       'BS_JDMM2XG4', 'BS_859AV1DB', 'BS_RCC7ZEC0', 'BS_9KDA3HG6',\n",
       "       'BS_E0S2Y0TS', 'BS_1C4WCNQF',\n",
       "       ...\n",
       "       'SJOS001101_M1', 'SJOS001130_D1', 'SJOS010_D', 'SJOS012_D',\n",
       "       'SJOS030876_D1', 'SJRHB012_D', 'SJST030043_D1', 'SJST030131_D1',\n",
       "       'SJST030383_D1', 'SJST030890_D1'],\n",
       "      dtype='object', name='biosample', length=1711)"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#! python3\n",
    "# Library of functions to load metadata from various places, and merge into unified patient, biosample tables.\n",
    "# Usage: generate_biosample_table()\n",
    "# Requires: \n",
    "# \n",
    "# Owen Chapman\n",
    "\n",
    "import pandas as pd\n",
    "import pathlib\n",
    "import os\n",
    "\n",
    "## Function to load metadata from the AmpliconClassifier results\n",
    "## Get this file from /expanse/lustre/projects/csd677/collab/projects/pedpancan/AmpliconClassifier/batch/inputs\n",
    "def get_pedpancan_biosamples_from_AC(include_x01=False,path='pedpancan_summary_map.txt'):\n",
    "    path = pathlib.Path(path)\n",
    "    df = pd.read_csv(path, sep='\\t', header=None, index_col=0, names = [\"biosample\",\"file\"])\n",
    "    if not include_x01:\n",
    "        df['firstletter']=df.index.map(lambda x: x[0])\n",
    "        df = df[df.firstletter != 'P']\n",
    "    return df.index\n",
    "get_pedpancan_biosamples_from_AC()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b0223a72-3fb5-48d2-9d3a-e93643294747",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "## Functions to load metadata from the CAVATICA API. \n",
    "## See also 2023-11-27_cavatica-api/cavatica-api.ipynb\n",
    "def import_x01_biosample_metadata(path=\"X01-biosample-metadata.tsv\"):\n",
    "    path = pathlib.Path(path)\n",
    "    df = pd.read_csv(path, sep='\\t',index_col=0)\n",
    "    df[\"cohort\"]=\"PBTA-X01\"\n",
    "    return df\n",
    "def import_x00_biosample_metadata(path=\"X00-biosample-metadata.tsv\"):\n",
    "    df = import_x01_biosample_metadata(path)\n",
    "    df[\"cohort\"]=\"PBTA-X00\"\n",
    "    return df\n",
    "def import_pnoc_biosample_metadata(path=\"PNOC-biosample-metadata.tsv\"):\n",
    "    df = import_x01_biosample_metadata(path)\n",
    "    df[\"cohort\"]=\"PNOC\"\n",
    "    return df\n",
    "df = import_x01_biosample_metadata()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "47498484-555d-4b24-ba21-f6b87e538e21",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def clean_cavatica_biosample_metadata(df):\n",
    "    '''\n",
    "    Clean known errors in the x01 metadata, and unify ontologies.\n",
    "    '''\n",
    "    # remove suffix from pnoc sample ids\n",
    "    df.sample_id = df.sample_id.map(lambda x: '-'.join(x.split('-')[:2]) if x.startswith(\"7316-\") else x)\n",
    "    \n",
    "    df = df.replace({\n",
    "        'Tumor Descriptor':{\n",
    "            \"initial CNS Tumor\": \"Diagnosis\",\n",
    "            \"Not Applicable\":None,\n",
    "            \"Unavailable\":None,\n",
    "            \"Initial CNS Tumor\": \"Diagnosis\",\n",
    "            \"Progressive Disease Post-Mortem\":\"Progressive\",\n",
    "        },\n",
    "        'gender':{\n",
    "            \"Not Reported\":None\n",
    "        }\n",
    "    })\n",
    "    # Correct suspected errors\n",
    "    df.loc[\"BS_6Z213H2V\",\"Tumor Descriptor\"] = \"Progressive\" # This tumor was resected 175 days after the initial tumor resection.\n",
    "    df.loc[\"BS_1135HC0V\",\"Tumor Descriptor\"] = \"Second Malignancy\" # This dysplasia was diagnosed 574 days after the first tumor was resected, in a new location.\n",
    "    df.loc[\"BS_ZS1QRMXS\",\"Tumor Descriptor\"] = \"Progressive\" # Tumor resected 128 days after previous resection.\n",
    "    df.loc[\"BS_FVYBGMG1\",\"Tumor Descriptor\"] = \"Progressive\" # Tumor resected 107 days after previous resection.\n",
    "    df.loc[\"BS_5J5VH3X0\",\"Tumor Descriptor\"] = \"Progressive\" # Biopsied 240 days after previous biopsy.\n",
    "    df.loc[\"BS_E9M7TDB6\",\"Tumor Descriptor\"] = \"Progressive\" # Second resection in different location 112 days after previous resection.\n",
    "    df.loc[\"BS_5XZP7F4Q\",\"Tumor Descriptor\"] = \"Progressive\" # Second resection 1845 days after previous\n",
    "    df.loc[\"BS_EXTEGB51\",\"Tumor Descriptor\"] = \"Progressive\" # Third resection 1922 days after previous\n",
    "    df.loc[\"BS_93BV8AY9\",\"Tumor Descriptor\"] = \"Second Malignancy\" # Second diagnosis 2975 days after initial.\n",
    "    df.loc[\"BS_CRKBDAYZ\",\"Tumor Descriptor\"] = \"Progressive\" # Series of progressive diagnoses long after initial.\n",
    "    df.loc[\"BS_B4DY7ET3\",\"Tumor Descriptor\"] = \"Progressive\" # Second resection 119 days after previous.\n",
    "    return df \n",
    "df = clean_cavatica_biosample_metadata(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "d57c9063-01ae-4fe6-a873-8f42e22740a6",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gender</th>\n",
       "      <th>race</th>\n",
       "      <th>ethnicity</th>\n",
       "      <th>Kids First Participant ID</th>\n",
       "      <th>disease_type</th>\n",
       "      <th>sample_id</th>\n",
       "      <th>Tumor Descriptor</th>\n",
       "      <th>primary_site</th>\n",
       "      <th>age_at_diagnosis</th>\n",
       "      <th>Kids First Biospecimen ID Normal</th>\n",
       "      <th>case_id</th>\n",
       "      <th>WGS_UUID</th>\n",
       "      <th>cohort</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Kids First Biospecimen ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>BS_K07KNTFY</th>\n",
       "      <td>Male</td>\n",
       "      <td>White</td>\n",
       "      <td>Not Hispanic or Latino</td>\n",
       "      <td>PT_00G007DM</td>\n",
       "      <td>Other</td>\n",
       "      <td>7316-272</td>\n",
       "      <td>Diagnosis</td>\n",
       "      <td>Cerebellum/Posterior Fossa</td>\n",
       "      <td>464</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C27429</td>\n",
       "      <td>7b06bec0-c64a-4df4-8237-d07de0a5df42</td>\n",
       "      <td>PBTA-X00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BS_AQMKA8NC</th>\n",
       "      <td>Male</td>\n",
       "      <td>White</td>\n",
       "      <td>Not Hispanic or Latino</td>\n",
       "      <td>PT_00G007DM</td>\n",
       "      <td>Other</td>\n",
       "      <td>7316-2577</td>\n",
       "      <td>Recurrence</td>\n",
       "      <td>Cerebellum/Posterior Fossa</td>\n",
       "      <td>3178</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C27429</td>\n",
       "      <td>9ba06668-759e-440c-a22f-957df52dae06</td>\n",
       "      <td>PBTA-X00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BS_TE8QFF7T</th>\n",
       "      <td>Male</td>\n",
       "      <td>White</td>\n",
       "      <td>Not Hispanic or Latino</td>\n",
       "      <td>PT_01MZ62KG</td>\n",
       "      <td>Pineoblastoma</td>\n",
       "      <td>7316-447</td>\n",
       "      <td>Progressive</td>\n",
       "      <td>Cerebellum/Posterior Fossa</td>\n",
       "      <td>546</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C92004</td>\n",
       "      <td>bd68f356-781e-49d8-a5ff-ec6c12514dc1</td>\n",
       "      <td>PBTA-X00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BS_2VB8649B</th>\n",
       "      <td>Male</td>\n",
       "      <td>Black or African American</td>\n",
       "      <td>Not Available</td>\n",
       "      <td>PT_02J5CWN5</td>\n",
       "      <td>Low-grade glioma/astrocytoma (WHO grade I/II)</td>\n",
       "      <td>7316-2989</td>\n",
       "      <td>Diagnosis</td>\n",
       "      <td>Suprasellar/Hypothalamic/Pituitary</td>\n",
       "      <td>2223</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C90159</td>\n",
       "      <td>3cdb9fca-94cf-4938-9d20-cd1b0c59f6a0</td>\n",
       "      <td>PBTA-X00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BS_5TT6TT4K</th>\n",
       "      <td>Male</td>\n",
       "      <td>Black or African American</td>\n",
       "      <td>Not Available</td>\n",
       "      <td>PT_02J5CWN5</td>\n",
       "      <td>Low-grade glioma/astrocytoma (WHO grade I/II)</td>\n",
       "      <td>7316-898</td>\n",
       "      <td>Progressive</td>\n",
       "      <td>Suprasellar/Hypothalamic/Pituitary</td>\n",
       "      <td>3722</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C90159</td>\n",
       "      <td>2c620221-1561-4073-8558-f81e4b3c9dcc</td>\n",
       "      <td>PBTA-X00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BS_VXDGXQKZ</th>\n",
       "      <td>Female</td>\n",
       "      <td>Reported Unknown</td>\n",
       "      <td>Reported Unknown</td>\n",
       "      <td>PT_VPEMAQBN</td>\n",
       "      <td>Anaplastic Astrocytoma</td>\n",
       "      <td>7316-3235</td>\n",
       "      <td>Diagnosis</td>\n",
       "      <td>Brain Stem</td>\n",
       "      <td>1709</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C3078444</td>\n",
       "      <td>7ad281cf-82c0-46b9-9a23-8348d623ba77</td>\n",
       "      <td>PNOC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BS_38CD519Z</th>\n",
       "      <td>Male</td>\n",
       "      <td>Asian</td>\n",
       "      <td>Hispanic or Latino</td>\n",
       "      <td>PT_W5GP3F6B</td>\n",
       "      <td>Diffuse midline glioma; H3K27M mutant; WHO gra...</td>\n",
       "      <td>7316-5003</td>\n",
       "      <td>Diagnosis</td>\n",
       "      <td>Brain Stem</td>\n",
       "      <td>2459</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C3092712</td>\n",
       "      <td>80fd23ab-de59-45da-80f6-49af162fe982</td>\n",
       "      <td>PNOC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BS_4DQAQFQH</th>\n",
       "      <td>Female</td>\n",
       "      <td>Reported Unknown</td>\n",
       "      <td>Hispanic or Latino</td>\n",
       "      <td>PT_WGVEF96B</td>\n",
       "      <td>Diffuse Astrocytoma</td>\n",
       "      <td>7316-4446</td>\n",
       "      <td>Diagnosis</td>\n",
       "      <td>Brain Stem</td>\n",
       "      <td>2853</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C3080043</td>\n",
       "      <td>3d9ec140-cf1f-48e0-b4bd-d9ae775b29b0</td>\n",
       "      <td>PNOC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BS_TQ0J7WJQ</th>\n",
       "      <td>Female</td>\n",
       "      <td>Reported Unknown</td>\n",
       "      <td>Hispanic or Latino</td>\n",
       "      <td>PT_WGVEF96B</td>\n",
       "      <td>Diffuse Astrocytoma</td>\n",
       "      <td>7316-3219</td>\n",
       "      <td>Diagnosis</td>\n",
       "      <td>Brain Stem</td>\n",
       "      <td>2853</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C3080043</td>\n",
       "      <td>c2e040bd-64b7-4928-8428-1da909cb2873</td>\n",
       "      <td>PNOC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BS_5Z0GA574</th>\n",
       "      <td>Female</td>\n",
       "      <td>White</td>\n",
       "      <td>Not Hispanic or Latino</td>\n",
       "      <td>PT_Y74CVASJ</td>\n",
       "      <td>Glioblastoma</td>\n",
       "      <td>7316-5002</td>\n",
       "      <td>Diagnosis</td>\n",
       "      <td>Brain Stem</td>\n",
       "      <td>1717</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C3081519</td>\n",
       "      <td>a7eaf642-8f59-4b80-a1c1-709afc030e16</td>\n",
       "      <td>PNOC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>994 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           gender                       race  \\\n",
       "Kids First Biospecimen ID                                      \n",
       "BS_K07KNTFY                  Male                      White   \n",
       "BS_AQMKA8NC                  Male                      White   \n",
       "BS_TE8QFF7T                  Male                      White   \n",
       "BS_2VB8649B                  Male  Black or African American   \n",
       "BS_5TT6TT4K                  Male  Black or African American   \n",
       "...                           ...                        ...   \n",
       "BS_VXDGXQKZ                Female           Reported Unknown   \n",
       "BS_38CD519Z                  Male                      Asian   \n",
       "BS_4DQAQFQH                Female           Reported Unknown   \n",
       "BS_TQ0J7WJQ                Female           Reported Unknown   \n",
       "BS_5Z0GA574                Female                      White   \n",
       "\n",
       "                                        ethnicity Kids First Participant ID  \\\n",
       "Kids First Biospecimen ID                                                     \n",
       "BS_K07KNTFY                Not Hispanic or Latino               PT_00G007DM   \n",
       "BS_AQMKA8NC                Not Hispanic or Latino               PT_00G007DM   \n",
       "BS_TE8QFF7T                Not Hispanic or Latino               PT_01MZ62KG   \n",
       "BS_2VB8649B                         Not Available               PT_02J5CWN5   \n",
       "BS_5TT6TT4K                         Not Available               PT_02J5CWN5   \n",
       "...                                           ...                       ...   \n",
       "BS_VXDGXQKZ                      Reported Unknown               PT_VPEMAQBN   \n",
       "BS_38CD519Z                    Hispanic or Latino               PT_W5GP3F6B   \n",
       "BS_4DQAQFQH                    Hispanic or Latino               PT_WGVEF96B   \n",
       "BS_TQ0J7WJQ                    Hispanic or Latino               PT_WGVEF96B   \n",
       "BS_5Z0GA574                Not Hispanic or Latino               PT_Y74CVASJ   \n",
       "\n",
       "                                                                disease_type  \\\n",
       "Kids First Biospecimen ID                                                      \n",
       "BS_K07KNTFY                                                            Other   \n",
       "BS_AQMKA8NC                                                            Other   \n",
       "BS_TE8QFF7T                                                    Pineoblastoma   \n",
       "BS_2VB8649B                    Low-grade glioma/astrocytoma (WHO grade I/II)   \n",
       "BS_5TT6TT4K                    Low-grade glioma/astrocytoma (WHO grade I/II)   \n",
       "...                                                                      ...   \n",
       "BS_VXDGXQKZ                                           Anaplastic Astrocytoma   \n",
       "BS_38CD519Z                Diffuse midline glioma; H3K27M mutant; WHO gra...   \n",
       "BS_4DQAQFQH                                              Diffuse Astrocytoma   \n",
       "BS_TQ0J7WJQ                                              Diffuse Astrocytoma   \n",
       "BS_5Z0GA574                                                     Glioblastoma   \n",
       "\n",
       "                           sample_id Tumor Descriptor  \\\n",
       "Kids First Biospecimen ID                               \n",
       "BS_K07KNTFY                 7316-272        Diagnosis   \n",
       "BS_AQMKA8NC                7316-2577       Recurrence   \n",
       "BS_TE8QFF7T                 7316-447      Progressive   \n",
       "BS_2VB8649B                7316-2989        Diagnosis   \n",
       "BS_5TT6TT4K                 7316-898      Progressive   \n",
       "...                              ...              ...   \n",
       "BS_VXDGXQKZ                7316-3235        Diagnosis   \n",
       "BS_38CD519Z                7316-5003        Diagnosis   \n",
       "BS_4DQAQFQH                7316-4446        Diagnosis   \n",
       "BS_TQ0J7WJQ                7316-3219        Diagnosis   \n",
       "BS_5Z0GA574                7316-5002        Diagnosis   \n",
       "\n",
       "                                                 primary_site  \\\n",
       "Kids First Biospecimen ID                                       \n",
       "BS_K07KNTFY                        Cerebellum/Posterior Fossa   \n",
       "BS_AQMKA8NC                        Cerebellum/Posterior Fossa   \n",
       "BS_TE8QFF7T                        Cerebellum/Posterior Fossa   \n",
       "BS_2VB8649B                Suprasellar/Hypothalamic/Pituitary   \n",
       "BS_5TT6TT4K                Suprasellar/Hypothalamic/Pituitary   \n",
       "...                                                       ...   \n",
       "BS_VXDGXQKZ                                        Brain Stem   \n",
       "BS_38CD519Z                                        Brain Stem   \n",
       "BS_4DQAQFQH                                        Brain Stem   \n",
       "BS_TQ0J7WJQ                                        Brain Stem   \n",
       "BS_5Z0GA574                                        Brain Stem   \n",
       "\n",
       "                           age_at_diagnosis  Kids First Biospecimen ID Normal  \\\n",
       "Kids First Biospecimen ID                                                       \n",
       "BS_K07KNTFY                             464                               NaN   \n",
       "BS_AQMKA8NC                            3178                               NaN   \n",
       "BS_TE8QFF7T                             546                               NaN   \n",
       "BS_2VB8649B                            2223                               NaN   \n",
       "BS_5TT6TT4K                            3722                               NaN   \n",
       "...                                     ...                               ...   \n",
       "BS_VXDGXQKZ                            1709                               NaN   \n",
       "BS_38CD519Z                            2459                               NaN   \n",
       "BS_4DQAQFQH                            2853                               NaN   \n",
       "BS_TQ0J7WJQ                            2853                               NaN   \n",
       "BS_5Z0GA574                            1717                               NaN   \n",
       "\n",
       "                            case_id                              WGS_UUID  \\\n",
       "Kids First Biospecimen ID                                                   \n",
       "BS_K07KNTFY                  C27429  7b06bec0-c64a-4df4-8237-d07de0a5df42   \n",
       "BS_AQMKA8NC                  C27429  9ba06668-759e-440c-a22f-957df52dae06   \n",
       "BS_TE8QFF7T                  C92004  bd68f356-781e-49d8-a5ff-ec6c12514dc1   \n",
       "BS_2VB8649B                  C90159  3cdb9fca-94cf-4938-9d20-cd1b0c59f6a0   \n",
       "BS_5TT6TT4K                  C90159  2c620221-1561-4073-8558-f81e4b3c9dcc   \n",
       "...                             ...                                   ...   \n",
       "BS_VXDGXQKZ                C3078444  7ad281cf-82c0-46b9-9a23-8348d623ba77   \n",
       "BS_38CD519Z                C3092712  80fd23ab-de59-45da-80f6-49af162fe982   \n",
       "BS_4DQAQFQH                C3080043  3d9ec140-cf1f-48e0-b4bd-d9ae775b29b0   \n",
       "BS_TQ0J7WJQ                C3080043  c2e040bd-64b7-4928-8428-1da909cb2873   \n",
       "BS_5Z0GA574                C3081519  a7eaf642-8f59-4b80-a1c1-709afc030e16   \n",
       "\n",
       "                             cohort  \n",
       "Kids First Biospecimen ID            \n",
       "BS_K07KNTFY                PBTA-X00  \n",
       "BS_AQMKA8NC                PBTA-X00  \n",
       "BS_TE8QFF7T                PBTA-X00  \n",
       "BS_2VB8649B                PBTA-X00  \n",
       "BS_5TT6TT4K                PBTA-X00  \n",
       "...                             ...  \n",
       "BS_VXDGXQKZ                    PNOC  \n",
       "BS_38CD519Z                    PNOC  \n",
       "BS_4DQAQFQH                    PNOC  \n",
       "BS_TQ0J7WJQ                    PNOC  \n",
       "BS_5Z0GA574                    PNOC  \n",
       "\n",
       "[994 rows x 13 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Function to compile CAVATICA metatdata for all CBTN samples in our cohort.\n",
    "def import_cbtn_biosample_metadata(include_X01=False):\n",
    "    if include_X01:\n",
    "        df = pd.concat([import_x00_biosample_metadata(),import_x01_biosample_metadata(),import_pnoc_biosample_metadata()])\n",
    "    else:\n",
    "        df = pd.concat([import_x00_biosample_metadata(),import_pnoc_biosample_metadata()])\n",
    "    cohort = get_pedpancan_biosamples_from_AC()\n",
    "    df = df[df.index.isin(cohort)]\n",
    "    df = clean_cavatica_biosample_metadata(df)\n",
    "    return df\n",
    "import_cbtn_biosample_metadata()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "54d756d6-c958-435a-b029-27ae2640d4c1",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/vv/261hv1sn1h36hvx1wc1cv9w40000gp/T/ipykernel_49058/3760351257.py:3: ParserWarning: Falling back to the 'python' engine because the 'c' engine does not support regex separators (separators > 1 char and different from '\\s+' are interpreted as regex); you can avoid this warning by specifying engine='python'.\n",
      "  df = pd.read_csv(\"pedpancan_histologies.tsv\", sep='delimiter', header=None)\n"
     ]
    },
    {
     "ename": "AttributeError",
     "evalue": "'DataFrame' object has no attribute 'sample_id'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m/var/folders/vv/261hv1sn1h36hvx1wc1cv9w40000gp/T/ipykernel_49058/3760351257.py\u001b[0m in \u001b[0;36m?\u001b[0;34m()\u001b[0m\n\u001b[1;32m     60\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     61\u001b[0m     \u001b[0;31m# Subset our cohort\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     62\u001b[0m     \u001b[0mdf\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mdf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0misin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcohort\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     63\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mdf\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 64\u001b[0;31m \u001b[0mclean_opentarget_histologies_files\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdf\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m/var/folders/vv/261hv1sn1h36hvx1wc1cv9w40000gp/T/ipykernel_49058/3760351257.py\u001b[0m in \u001b[0;36m?\u001b[0;34m(df)\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0mclean_opentarget_histologies_files\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdf\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m     \u001b[0mcohort\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mimport_cbtn_biosample_metadata\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 6\u001b[0;31m     \u001b[0mdf\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mdf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msample_id\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0misin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcohort\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msample_id\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      7\u001b[0m     \u001b[0mdf\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mdf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msample_type\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m'Tumor'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;31m# Drop normals\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      8\u001b[0m     \u001b[0mdf\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mdf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcomposition\u001b[0m \u001b[0;34m!=\u001b[0m \u001b[0;34m'Derived Cell Line'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;31m# Drop cell lines\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      9\u001b[0m     \u001b[0mdf\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mdf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mexperimental_strategy\u001b[0m \u001b[0;34m!=\u001b[0m \u001b[0;34m\"Targeted Sequencing\"\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;31m# these metadata are very different\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/lib/python3.11/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36m?\u001b[0;34m(self, name)\u001b[0m\n\u001b[1;32m   5985\u001b[0m             \u001b[0;32mand\u001b[0m \u001b[0mname\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_accessors\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   5986\u001b[0m             \u001b[0;32mand\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_info_axis\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_can_hold_identifiers_and_holds_name\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   5987\u001b[0m         ):\n\u001b[1;32m   5988\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 5989\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mobject\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__getattribute__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m: 'DataFrame' object has no attribute 'sample_id'"
     ]
    }
   ],
   "source": [
    "## Functions to open & preprocess opentarget histology data.\n",
    "## Get histologies.tsv from https://github.com/d3b-center/OpenPedCan-analysis/blob/dev/analyses/molecular-subtyping-integrate/results/histologies.tsv\n",
    "df = pd.read_csv(\"pedpancan_histologies.tsv\", sep='delimiter', header=None)\n",
    "def clean_opentarget_histologies_files(df):\n",
    "    cohort = import_cbtn_biosample_metadata()\n",
    "    df = df[df.sample_id.isin(cohort.sample_id)]\n",
    "    df = df[df.sample_type == 'Tumor'] # Drop normals\n",
    "    df = df[df.composition != 'Derived Cell Line'] # Drop cell lines\n",
    "    df = df[df.experimental_strategy != \"Targeted Sequencing\"] # these metadata are very different\n",
    "    df = df.drop([\"RNA_library\",\"seq_center\",\"pathology_free_text_diagnosis\",\"gtex_group\",\"gtex_subgroup\",\"normal_fraction\",\n",
    "                  \"cell_line_composition\",\"cell_line_passage\",\"tumor_fraction_RFpurify_ABSOLUTE\",\n",
    "                  \"tumor_fraction_RFpurify_ESTIMATE\",\"tumor_fraction_LUMP\",\"dkfz_v12_methylation_mgmt_status\",\n",
    "                  \"dkfz_v12_methylation_mgmt_estimated\",\"integrated_diagnosis\",\n",
    "                  \"tumor_fraction\",\"tumor_ploidy\",\"cohort\"],axis=1) # drop columns we know we don't want\n",
    "    df = df.replace({\n",
    "        'composition':{\n",
    "            \"Not Available\": None,\n",
    "        },\n",
    "        'extent_of_tumor_resection':{\n",
    "            \"Not Reported\":None,\n",
    "            'Unavailable':None,\n",
    "            'Not Applicable':None\n",
    "        },\n",
    "    })\n",
    "    \n",
    "    # correct known errors\n",
    "    df.loc[\"BS_K14VJ1E3\",\"age_at_diagnosis_days\"] = 2778\n",
    "    df = df.drop([\"BS_03G6PJKJ\",\"BS_HJJPT3NR\",\"BS_15R0SQRN\",\"BS_GGDMSB26\",\"BS_VQGR0D61\"]) # lots of biosamples for the same tumor at the same timepoint\n",
    "    \n",
    "    # Propagate metadata from the same sample_id\n",
    "    g = df.groupby('sample_id')\n",
    "    df = []\n",
    "    for name, group in g:\n",
    "        columns = [col for col in group.columns if col not in ['sample_id','aliquot_id','experimental_strategy']]\n",
    "        for column in columns:\n",
    "            unique_values = group[column].dropna().unique()\n",
    "            if len(unique_values) == 0:\n",
    "                continue\n",
    "            elif len(unique_values) == 1:\n",
    "                non_na_value = unique_values[0]\n",
    "                group[column].fillna(non_na_value, inplace=True)\n",
    "            else:\n",
    "                print(f\"Warning: The column '{column}' for sample {name} differs between CAVATICA and opentarget annotations.\")\n",
    "        group=group.sort_values('experimental_strategy')\n",
    "        df.append(group)\n",
    "    df = pd.concat(df)\n",
    "    \n",
    "    # Add entries missing a KF biospecimen ID, but with a matching external biosample id.\n",
    "    missing_bs = (cohort[~cohort.index.isin(df.index)][\"sample_id\"]).sort_values()\n",
    "    print(f\"{len(missing_bs)} KF biospecimens missing from the opentarget histologies table...\")\n",
    "    missing_bs = missing_bs[missing_bs.isin(df.sample_id)]\n",
    "    print(f\"found {len(missing_bs)} matching external sample IDs, adding to table...\")\n",
    "    newdf = []\n",
    "    for biospecimen, sample in missing_bs.items():\n",
    "        newentry=df[df.sample_id==sample].iloc[0]\n",
    "        newentry.name = biospecimen\n",
    "        newdf.append(newentry)\n",
    "    newdf = pd.DataFrame(newdf)\n",
    "    df = pd.concat([df,newdf])\n",
    "    \n",
    "    # Subset our cohort\n",
    "    df = df[df.index.isin(cohort.index)]\n",
    "    return df\n",
    "clean_opentarget_histologies_files(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "80a281cd-1477-4ba8-b109-e876b0801032",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/vv/261hv1sn1h36hvx1wc1cv9w40000gp/T/ipykernel_49058/2192161946.py:6: DtypeWarning: Columns (1,21,24,25,26,27,33,36,37,41,43,45,47,48,49,51) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  df = pd.read_csv(path,sep='\\t',index_col=0)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning: The column 'harmonized_diagnosis' for sample 7316-1101 differs between CAVATICA and opentarget annotations.\n",
      "Warning: The column 'harmonized_diagnosis' for sample 7316-2935 differs between CAVATICA and opentarget annotations.\n",
      "Warning: The column 'tumor_descriptor' for sample 7316-3217 differs between CAVATICA and opentarget annotations.\n",
      "Warning: The column 'primary_site' for sample 7316-3229 differs between CAVATICA and opentarget annotations.\n",
      "Warning: The column 'age_at_diagnosis_days' for sample 7316-323 differs between CAVATICA and opentarget annotations.\n",
      "Warning: The column 'primary_site' for sample 7316-3230 differs between CAVATICA and opentarget annotations.\n",
      "Warning: The column 'CNS_region' for sample 7316-3230 differs between CAVATICA and opentarget annotations.\n",
      "Warning: The column 'primary_site' for sample 7316-3231 differs between CAVATICA and opentarget annotations.\n",
      "Warning: The column 'harmonized_diagnosis' for sample 7316-3295 differs between CAVATICA and opentarget annotations.\n",
      "Warning: The column 'harmonized_diagnosis' for sample 7316-3632 differs between CAVATICA and opentarget annotations.\n",
      "Warning: The column 'harmonized_diagnosis' for sample 7316-4062 differs between CAVATICA and opentarget annotations.\n",
      "Warning: The column 'harmonized_diagnosis' for sample 7316-921 differs between CAVATICA and opentarget annotations.\n",
      "53 KF biospecimens missing from the opentarget histologies table...\n",
      "found 41 matching external sample IDs, adding to table...\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>aliquot_id</th>\n",
       "      <th>Kids_First_Participant_ID</th>\n",
       "      <th>experimental_strategy</th>\n",
       "      <th>sample_type</th>\n",
       "      <th>composition</th>\n",
       "      <th>tumor_descriptor</th>\n",
       "      <th>primary_site</th>\n",
       "      <th>reported_gender</th>\n",
       "      <th>race</th>\n",
       "      <th>...</th>\n",
       "      <th>dkfz_v11_methylation_subclass_score</th>\n",
       "      <th>dkfz_v12_methylation_subclass</th>\n",
       "      <th>dkfz_v12_methylation_subclass_score</th>\n",
       "      <th>molecular_subtype</th>\n",
       "      <th>Notes</th>\n",
       "      <th>harmonized_diagnosis</th>\n",
       "      <th>molecular_subtype_methyl</th>\n",
       "      <th>broad_histology</th>\n",
       "      <th>short_histology</th>\n",
       "      <th>cancer_group</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>BS_1RFBH1SP</th>\n",
       "      <td>7316-10</td>\n",
       "      <td>717017</td>\n",
       "      <td>PT_K8ZV7APT</td>\n",
       "      <td>WGS</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>Solid Tissue</td>\n",
       "      <td>Initial CNS Tumor</td>\n",
       "      <td>Spinal Cord- Lumbar/Thecal Sac</td>\n",
       "      <td>Female</td>\n",
       "      <td>Black or African American</td>\n",
       "      <td>...</td>\n",
       "      <td>0.598</td>\n",
       "      <td>NFIB_PLEX</td>\n",
       "      <td>0.507953</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Neurofibroma/Plexiform</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Tumor of cranial and paraspinal nerves</td>\n",
       "      <td>Neurofibroma</td>\n",
       "      <td>Neurofibroma/Plexiform</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BS_SFZ3A07S</th>\n",
       "      <td>7316-100</td>\n",
       "      <td>67852</td>\n",
       "      <td>PT_6TZR2DH1</td>\n",
       "      <td>WGS</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>Solid Tissue</td>\n",
       "      <td>Initial CNS Tumor</td>\n",
       "      <td>Suprasellar/Hypothalamic/Pituitary</td>\n",
       "      <td>Female</td>\n",
       "      <td>Black or African American</td>\n",
       "      <td>...</td>\n",
       "      <td>0.668</td>\n",
       "      <td>CPH_ADM</td>\n",
       "      <td>0.580744</td>\n",
       "      <td>CRANIO, ADAM</td>\n",
       "      <td>Updated via OpenPedCan subtyping</td>\n",
       "      <td>Adamantinomatous craniopharyngioma</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Tumors of sellar region</td>\n",
       "      <td>Craniopharyngioma</td>\n",
       "      <td>Adamantinomatous Craniopharyngioma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BS_4RS1SC48</th>\n",
       "      <td>7316-101</td>\n",
       "      <td>588044</td>\n",
       "      <td>PT_CWD717Q0</td>\n",
       "      <td>WGS</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>Solid Tissue</td>\n",
       "      <td>Initial CNS Tumor</td>\n",
       "      <td>Temporal Lobe</td>\n",
       "      <td>Male</td>\n",
       "      <td>Black or African American</td>\n",
       "      <td>...</td>\n",
       "      <td>0.479</td>\n",
       "      <td>PA_CORT</td>\n",
       "      <td>0.825647</td>\n",
       "      <td>GNG, other MAPK</td>\n",
       "      <td>Updated via OpenPedCan subtyping</td>\n",
       "      <td>Ganglioglioma, other MAPK</td>\n",
       "      <td>GNT, MAPK</td>\n",
       "      <td>Low-grade astrocytic tumor</td>\n",
       "      <td>Ganglioglioma</td>\n",
       "      <td>Ganglioglioma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BS_MRX1SPFB</th>\n",
       "      <td>7316-104</td>\n",
       "      <td>717047</td>\n",
       "      <td>PT_H024RK87</td>\n",
       "      <td>WGS</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>Solid Tissue</td>\n",
       "      <td>Recurrence</td>\n",
       "      <td>Brain Stem- Midbrain/Tectum;Pineal Gland;Thala...</td>\n",
       "      <td>Male</td>\n",
       "      <td>Asian</td>\n",
       "      <td>...</td>\n",
       "      <td>0.230</td>\n",
       "      <td>GCT_TERA</td>\n",
       "      <td>0.972105</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Non-germinomatous germ cell tumor;Teratoma</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Germ cell tumor</td>\n",
       "      <td>Germ cell tumor</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BS_A9S5HT6P</th>\n",
       "      <td>7316-1069</td>\n",
       "      <td>717042</td>\n",
       "      <td>PT_9XAA8PF6</td>\n",
       "      <td>WGS</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>Solid Tissue</td>\n",
       "      <td>Initial CNS Tumor</td>\n",
       "      <td>Suprasellar/Hypothalamic/Pituitary</td>\n",
       "      <td>Female</td>\n",
       "      <td>White</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Adenoma</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Benign tumor</td>\n",
       "      <td>Adenoma</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BS_FJEZ3ASV</th>\n",
       "      <td>7316-913</td>\n",
       "      <td>1108019</td>\n",
       "      <td>PT_4347ZBEX</td>\n",
       "      <td>Methylation</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>Solid Tissue</td>\n",
       "      <td>Second Malignancy</td>\n",
       "      <td>Parietal Lobe;Temporal Lobe</td>\n",
       "      <td>Female</td>\n",
       "      <td>White</td>\n",
       "      <td>...</td>\n",
       "      <td>0.999</td>\n",
       "      <td>pedHGG_RTK1C</td>\n",
       "      <td>0.982681</td>\n",
       "      <td>HGG, H3 wildtype</td>\n",
       "      <td>Updated via OpenPedCan subtyping</td>\n",
       "      <td>High-grade glioma, IDH-wildtype and H3-wildtype</td>\n",
       "      <td>HGG, H3 wildtype</td>\n",
       "      <td>Diffuse astrocytic and oligodendroglial tumor</td>\n",
       "      <td>HGAT</td>\n",
       "      <td>High-grade glioma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BS_5DHZ0T1A</th>\n",
       "      <td>7316-925</td>\n",
       "      <td>1112611</td>\n",
       "      <td>PT_3HRS5CWF</td>\n",
       "      <td>Methylation</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>Solid Tissue</td>\n",
       "      <td>Initial CNS Tumor</td>\n",
       "      <td>Cerebellum/Posterior Fossa</td>\n",
       "      <td>Female</td>\n",
       "      <td>White</td>\n",
       "      <td>...</td>\n",
       "      <td>0.993</td>\n",
       "      <td>PA_INF</td>\n",
       "      <td>0.999797</td>\n",
       "      <td>LGG, KIAA1549-BRAF</td>\n",
       "      <td>Updated via OpenPedCan subtyping</td>\n",
       "      <td>Low-grade glioma, KIAA1549-BRAF</td>\n",
       "      <td>LGG, FGFR</td>\n",
       "      <td>Low-grade astrocytic tumor</td>\n",
       "      <td>LGAT</td>\n",
       "      <td>Low-grade glioma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BS_W37QBA12</th>\n",
       "      <td>7316-931</td>\n",
       "      <td>1112899</td>\n",
       "      <td>PT_XA98HG1C</td>\n",
       "      <td>Methylation</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>Solid Tissue</td>\n",
       "      <td>Initial CNS Tumor</td>\n",
       "      <td>Cerebellum/Posterior Fossa</td>\n",
       "      <td>Male</td>\n",
       "      <td>White</td>\n",
       "      <td>...</td>\n",
       "      <td>0.906</td>\n",
       "      <td>MB_SHH_3</td>\n",
       "      <td>0.480522</td>\n",
       "      <td>MB, SHH</td>\n",
       "      <td>Subtype based on prediction;Updated via OpenPe...</td>\n",
       "      <td>Medulloblastoma, SHH-activated</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Embryonal tumor</td>\n",
       "      <td>Medulloblastoma</td>\n",
       "      <td>Medulloblastoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BS_ZVM1B6BF</th>\n",
       "      <td>7316-95</td>\n",
       "      <td>608916</td>\n",
       "      <td>PT_9S6WMQ92</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>Solid Tissue</td>\n",
       "      <td>Progressive</td>\n",
       "      <td>Cerebellum/Posterior Fossa</td>\n",
       "      <td>Female</td>\n",
       "      <td>White</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MB, SHH</td>\n",
       "      <td>Subtype based on prediction;Updated via OpenPe...</td>\n",
       "      <td>Medulloblastoma, SHH-activated</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Embryonal tumor</td>\n",
       "      <td>Medulloblastoma</td>\n",
       "      <td>Medulloblastoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BS_6D37YSGS</th>\n",
       "      <td>7316-968</td>\n",
       "      <td>1030646</td>\n",
       "      <td>PT_62VNQ8C4</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>Solid Tissue</td>\n",
       "      <td>Initial CNS Tumor</td>\n",
       "      <td>Cerebellum/Posterior Fossa;Spinal Cord- Cervical</td>\n",
       "      <td>Female</td>\n",
       "      <td>White</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MB, SHH</td>\n",
       "      <td>Subtype based on prediction;Updated via OpenPe...</td>\n",
       "      <td>Medulloblastoma, SHH-activated</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Embryonal tumor</td>\n",
       "      <td>Medulloblastoma</td>\n",
       "      <td>Medulloblastoma</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>982 rows × 37 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             sample_id aliquot_id Kids_First_Participant_ID  \\\n",
       "BS_1RFBH1SP    7316-10     717017               PT_K8ZV7APT   \n",
       "BS_SFZ3A07S   7316-100      67852               PT_6TZR2DH1   \n",
       "BS_4RS1SC48   7316-101     588044               PT_CWD717Q0   \n",
       "BS_MRX1SPFB   7316-104     717047               PT_H024RK87   \n",
       "BS_A9S5HT6P  7316-1069     717042               PT_9XAA8PF6   \n",
       "...                ...        ...                       ...   \n",
       "BS_FJEZ3ASV   7316-913    1108019               PT_4347ZBEX   \n",
       "BS_5DHZ0T1A   7316-925    1112611               PT_3HRS5CWF   \n",
       "BS_W37QBA12   7316-931    1112899               PT_XA98HG1C   \n",
       "BS_ZVM1B6BF    7316-95     608916               PT_9S6WMQ92   \n",
       "BS_6D37YSGS   7316-968    1030646               PT_62VNQ8C4   \n",
       "\n",
       "            experimental_strategy sample_type   composition  \\\n",
       "BS_1RFBH1SP                   WGS       Tumor  Solid Tissue   \n",
       "BS_SFZ3A07S                   WGS       Tumor  Solid Tissue   \n",
       "BS_4RS1SC48                   WGS       Tumor  Solid Tissue   \n",
       "BS_MRX1SPFB                   WGS       Tumor  Solid Tissue   \n",
       "BS_A9S5HT6P                   WGS       Tumor  Solid Tissue   \n",
       "...                           ...         ...           ...   \n",
       "BS_FJEZ3ASV           Methylation       Tumor  Solid Tissue   \n",
       "BS_5DHZ0T1A           Methylation       Tumor  Solid Tissue   \n",
       "BS_W37QBA12           Methylation       Tumor  Solid Tissue   \n",
       "BS_ZVM1B6BF               RNA-Seq       Tumor  Solid Tissue   \n",
       "BS_6D37YSGS               RNA-Seq       Tumor  Solid Tissue   \n",
       "\n",
       "              tumor_descriptor  \\\n",
       "BS_1RFBH1SP  Initial CNS Tumor   \n",
       "BS_SFZ3A07S  Initial CNS Tumor   \n",
       "BS_4RS1SC48  Initial CNS Tumor   \n",
       "BS_MRX1SPFB         Recurrence   \n",
       "BS_A9S5HT6P  Initial CNS Tumor   \n",
       "...                        ...   \n",
       "BS_FJEZ3ASV  Second Malignancy   \n",
       "BS_5DHZ0T1A  Initial CNS Tumor   \n",
       "BS_W37QBA12  Initial CNS Tumor   \n",
       "BS_ZVM1B6BF        Progressive   \n",
       "BS_6D37YSGS  Initial CNS Tumor   \n",
       "\n",
       "                                                  primary_site  \\\n",
       "BS_1RFBH1SP                     Spinal Cord- Lumbar/Thecal Sac   \n",
       "BS_SFZ3A07S                 Suprasellar/Hypothalamic/Pituitary   \n",
       "BS_4RS1SC48                                      Temporal Lobe   \n",
       "BS_MRX1SPFB  Brain Stem- Midbrain/Tectum;Pineal Gland;Thala...   \n",
       "BS_A9S5HT6P                 Suprasellar/Hypothalamic/Pituitary   \n",
       "...                                                        ...   \n",
       "BS_FJEZ3ASV                        Parietal Lobe;Temporal Lobe   \n",
       "BS_5DHZ0T1A                         Cerebellum/Posterior Fossa   \n",
       "BS_W37QBA12                         Cerebellum/Posterior Fossa   \n",
       "BS_ZVM1B6BF                         Cerebellum/Posterior Fossa   \n",
       "BS_6D37YSGS   Cerebellum/Posterior Fossa;Spinal Cord- Cervical   \n",
       "\n",
       "            reported_gender                       race  ...  \\\n",
       "BS_1RFBH1SP          Female  Black or African American  ...   \n",
       "BS_SFZ3A07S          Female  Black or African American  ...   \n",
       "BS_4RS1SC48            Male  Black or African American  ...   \n",
       "BS_MRX1SPFB            Male                      Asian  ...   \n",
       "BS_A9S5HT6P          Female                      White  ...   \n",
       "...                     ...                        ...  ...   \n",
       "BS_FJEZ3ASV          Female                      White  ...   \n",
       "BS_5DHZ0T1A          Female                      White  ...   \n",
       "BS_W37QBA12            Male                      White  ...   \n",
       "BS_ZVM1B6BF          Female                      White  ...   \n",
       "BS_6D37YSGS          Female                      White  ...   \n",
       "\n",
       "            dkfz_v11_methylation_subclass_score  \\\n",
       "BS_1RFBH1SP                               0.598   \n",
       "BS_SFZ3A07S                               0.668   \n",
       "BS_4RS1SC48                               0.479   \n",
       "BS_MRX1SPFB                               0.230   \n",
       "BS_A9S5HT6P                                 NaN   \n",
       "...                                         ...   \n",
       "BS_FJEZ3ASV                               0.999   \n",
       "BS_5DHZ0T1A                               0.993   \n",
       "BS_W37QBA12                               0.906   \n",
       "BS_ZVM1B6BF                                 NaN   \n",
       "BS_6D37YSGS                                 NaN   \n",
       "\n",
       "             dkfz_v12_methylation_subclass  \\\n",
       "BS_1RFBH1SP                      NFIB_PLEX   \n",
       "BS_SFZ3A07S                        CPH_ADM   \n",
       "BS_4RS1SC48                        PA_CORT   \n",
       "BS_MRX1SPFB                       GCT_TERA   \n",
       "BS_A9S5HT6P                            NaN   \n",
       "...                                    ...   \n",
       "BS_FJEZ3ASV                   pedHGG_RTK1C   \n",
       "BS_5DHZ0T1A                         PA_INF   \n",
       "BS_W37QBA12                       MB_SHH_3   \n",
       "BS_ZVM1B6BF                            NaN   \n",
       "BS_6D37YSGS                            NaN   \n",
       "\n",
       "            dkfz_v12_methylation_subclass_score   molecular_subtype  \\\n",
       "BS_1RFBH1SP                            0.507953                 NaN   \n",
       "BS_SFZ3A07S                            0.580744        CRANIO, ADAM   \n",
       "BS_4RS1SC48                            0.825647     GNG, other MAPK   \n",
       "BS_MRX1SPFB                            0.972105                 NaN   \n",
       "BS_A9S5HT6P                                 NaN                 NaN   \n",
       "...                                         ...                 ...   \n",
       "BS_FJEZ3ASV                            0.982681    HGG, H3 wildtype   \n",
       "BS_5DHZ0T1A                            0.999797  LGG, KIAA1549-BRAF   \n",
       "BS_W37QBA12                            0.480522             MB, SHH   \n",
       "BS_ZVM1B6BF                                 NaN             MB, SHH   \n",
       "BS_6D37YSGS                                 NaN             MB, SHH   \n",
       "\n",
       "                                                         Notes  \\\n",
       "BS_1RFBH1SP                                                NaN   \n",
       "BS_SFZ3A07S                   Updated via OpenPedCan subtyping   \n",
       "BS_4RS1SC48                   Updated via OpenPedCan subtyping   \n",
       "BS_MRX1SPFB                                                NaN   \n",
       "BS_A9S5HT6P                                                NaN   \n",
       "...                                                        ...   \n",
       "BS_FJEZ3ASV                   Updated via OpenPedCan subtyping   \n",
       "BS_5DHZ0T1A                   Updated via OpenPedCan subtyping   \n",
       "BS_W37QBA12  Subtype based on prediction;Updated via OpenPe...   \n",
       "BS_ZVM1B6BF  Subtype based on prediction;Updated via OpenPe...   \n",
       "BS_6D37YSGS  Subtype based on prediction;Updated via OpenPe...   \n",
       "\n",
       "                                        harmonized_diagnosis  \\\n",
       "BS_1RFBH1SP                           Neurofibroma/Plexiform   \n",
       "BS_SFZ3A07S               Adamantinomatous craniopharyngioma   \n",
       "BS_4RS1SC48                        Ganglioglioma, other MAPK   \n",
       "BS_MRX1SPFB       Non-germinomatous germ cell tumor;Teratoma   \n",
       "BS_A9S5HT6P                                          Adenoma   \n",
       "...                                                      ...   \n",
       "BS_FJEZ3ASV  High-grade glioma, IDH-wildtype and H3-wildtype   \n",
       "BS_5DHZ0T1A                  Low-grade glioma, KIAA1549-BRAF   \n",
       "BS_W37QBA12                   Medulloblastoma, SHH-activated   \n",
       "BS_ZVM1B6BF                   Medulloblastoma, SHH-activated   \n",
       "BS_6D37YSGS                   Medulloblastoma, SHH-activated   \n",
       "\n",
       "             molecular_subtype_methyl  \\\n",
       "BS_1RFBH1SP                       NaN   \n",
       "BS_SFZ3A07S                       NaN   \n",
       "BS_4RS1SC48                 GNT, MAPK   \n",
       "BS_MRX1SPFB                       NaN   \n",
       "BS_A9S5HT6P                       NaN   \n",
       "...                               ...   \n",
       "BS_FJEZ3ASV          HGG, H3 wildtype   \n",
       "BS_5DHZ0T1A                 LGG, FGFR   \n",
       "BS_W37QBA12                       NaN   \n",
       "BS_ZVM1B6BF                       NaN   \n",
       "BS_6D37YSGS                       NaN   \n",
       "\n",
       "                                           broad_histology    short_histology  \\\n",
       "BS_1RFBH1SP         Tumor of cranial and paraspinal nerves       Neurofibroma   \n",
       "BS_SFZ3A07S                        Tumors of sellar region  Craniopharyngioma   \n",
       "BS_4RS1SC48                     Low-grade astrocytic tumor      Ganglioglioma   \n",
       "BS_MRX1SPFB                                Germ cell tumor    Germ cell tumor   \n",
       "BS_A9S5HT6P                                   Benign tumor            Adenoma   \n",
       "...                                                    ...                ...   \n",
       "BS_FJEZ3ASV  Diffuse astrocytic and oligodendroglial tumor               HGAT   \n",
       "BS_5DHZ0T1A                     Low-grade astrocytic tumor               LGAT   \n",
       "BS_W37QBA12                                Embryonal tumor    Medulloblastoma   \n",
       "BS_ZVM1B6BF                                Embryonal tumor    Medulloblastoma   \n",
       "BS_6D37YSGS                                Embryonal tumor    Medulloblastoma   \n",
       "\n",
       "                                   cancer_group  \n",
       "BS_1RFBH1SP              Neurofibroma/Plexiform  \n",
       "BS_SFZ3A07S  Adamantinomatous Craniopharyngioma  \n",
       "BS_4RS1SC48                       Ganglioglioma  \n",
       "BS_MRX1SPFB                                 NaN  \n",
       "BS_A9S5HT6P                                 NaN  \n",
       "...                                         ...  \n",
       "BS_FJEZ3ASV                   High-grade glioma  \n",
       "BS_5DHZ0T1A                    Low-grade glioma  \n",
       "BS_W37QBA12                     Medulloblastoma  \n",
       "BS_ZVM1B6BF                     Medulloblastoma  \n",
       "BS_6D37YSGS                     Medulloblastoma  \n",
       "\n",
       "[982 rows x 37 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def import_opentarget_histologies_files(path='owen_histologies.tsv'):\n",
    "    '''\n",
    "    Get this file from /Users/ochapman/Library/CloudStorage/OneDrive-SanfordBurnhamPrebysMedicalDiscoveryInstitute/projects/2023-pedpancan/data/opentarget/histologies.tsv\n",
    "    '''\n",
    "    path = pathlib.Path(path)\n",
    "    df = pd.read_csv(path,sep='\\t',index_col=0)\n",
    "    df = clean_opentarget_histologies_files(df)\n",
    "    return df\n",
    "import_opentarget_histologies_files()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "7455c8d0-6cfb-4133-8876-1c9ce31fefbf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sampleId</th>\n",
       "      <th>CANCER_TYPE</th>\n",
       "      <th>CANCER_TYPE_DETAILED</th>\n",
       "      <th>CNS_REGION</th>\n",
       "      <th>EXPERIMENTAL_STRATEGY</th>\n",
       "      <th>EXTENT_OF_TUMOR_RESECTION</th>\n",
       "      <th>HISTOLOGY</th>\n",
       "      <th>MUTATION_COUNT</th>\n",
       "      <th>PATHOLOGY_FREE_TEXT_DIAGNOSIS</th>\n",
       "      <th>SAMPLE_TYPE</th>\n",
       "      <th>SPECIMEN_ID</th>\n",
       "      <th>TUMOR_FRACTION</th>\n",
       "      <th>TUMOR_PLOIDY</th>\n",
       "      <th>TUMOR_TISSUE_SITE</th>\n",
       "      <th>TUMOR_TYPE</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>patientId</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>PT_00G007DM</th>\n",
       "      <td>7316-2577</td>\n",
       "      <td>Embryonal tumor</td>\n",
       "      <td>CNS Embryonal tumor, NOS</td>\n",
       "      <td>Posterior fossa</td>\n",
       "      <td>WGS;RNA-Seq</td>\n",
       "      <td>Gross/Near total resection</td>\n",
       "      <td>Embryonal tumor with multilayer rosettes</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ependymoblastoma</td>\n",
       "      <td>Solid Tissue</td>\n",
       "      <td>BS_AQMKA8NC;BS_FEPRNEXX</td>\n",
       "      <td>0.613054</td>\n",
       "      <td>4.0</td>\n",
       "      <td>Cerebellum/Posterior Fossa</td>\n",
       "      <td>recurrence</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PT_00G007DM</th>\n",
       "      <td>7316-272</td>\n",
       "      <td>Embryonal tumor</td>\n",
       "      <td>Embryonal tumor with multilayer rosettes, C19M...</td>\n",
       "      <td>Posterior fossa</td>\n",
       "      <td>WGS;RNA-Seq</td>\n",
       "      <td>Gross/Near total resection</td>\n",
       "      <td>Embryonal tumor with multilayer rosettes</td>\n",
       "      <td>4.0</td>\n",
       "      <td>embryonal tumor with abundant neuropil and tru...</td>\n",
       "      <td>Solid Tissue</td>\n",
       "      <td>BS_K07KNTFY;BS_QWNBZ9RJ</td>\n",
       "      <td>0.328185</td>\n",
       "      <td>4.0</td>\n",
       "      <td>Cerebellum/Posterior Fossa</td>\n",
       "      <td>primary</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PT_01MZ62KG</th>\n",
       "      <td>7316-447</td>\n",
       "      <td>Embryonal tumor</td>\n",
       "      <td>Embryonal tumor with multilayer rosettes, C19M...</td>\n",
       "      <td>Posterior fossa</td>\n",
       "      <td>RNA-Seq;WGS</td>\n",
       "      <td>Partial resection</td>\n",
       "      <td>Embryonal tumor with multilayer rosettes</td>\n",
       "      <td>11.0</td>\n",
       "      <td>pineoblastoma</td>\n",
       "      <td>Solid Tissue</td>\n",
       "      <td>BS_TE8QFF7T;BS_P39SQPTS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cerebellum/Posterior Fossa</td>\n",
       "      <td>progression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PT_02J5CWN5</th>\n",
       "      <td>7316-2989</td>\n",
       "      <td>Low-grade astrocytic tumor</td>\n",
       "      <td>Low-grade glioma/astrocytoma, FGFR</td>\n",
       "      <td>Suprasellar</td>\n",
       "      <td>WGS;RNA-Seq</td>\n",
       "      <td>Partial resection</td>\n",
       "      <td>Low-grade glioma astrocytoma</td>\n",
       "      <td>7.0</td>\n",
       "      <td>low-grade glioma</td>\n",
       "      <td>Solid Tissue</td>\n",
       "      <td>BS_2VB8649B;BS_M8WP5T16</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Suprasellar/Hypothalamic/Pituitary</td>\n",
       "      <td>primary</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PT_02J5CWN5</th>\n",
       "      <td>7316-898</td>\n",
       "      <td>Low-grade astrocytic tumor</td>\n",
       "      <td>Low-grade glioma/astrocytoma, wildtype</td>\n",
       "      <td>Suprasellar</td>\n",
       "      <td>RNA-Seq;WGS</td>\n",
       "      <td>Partial resection</td>\n",
       "      <td>Low-grade glioma astrocytoma</td>\n",
       "      <td>1.0</td>\n",
       "      <td>low-grade glioma</td>\n",
       "      <td>Solid Tissue</td>\n",
       "      <td>BS_5TT6TT4K;BS_1RXZ3BP7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Suprasellar/Hypothalamic/Pituitary</td>\n",
       "      <td>progression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PT_ZZRBX5JT</th>\n",
       "      <td>7316-2661</td>\n",
       "      <td>Ependymal tumor</td>\n",
       "      <td>Ependymoma</td>\n",
       "      <td>Posterior fossa</td>\n",
       "      <td>RNA-Seq;WGS</td>\n",
       "      <td>Gross/Near total resection</td>\n",
       "      <td>Ependymoma</td>\n",
       "      <td>17.0</td>\n",
       "      <td>ependymoma</td>\n",
       "      <td>Solid Tissue</td>\n",
       "      <td>BS_JE1DF581;BS_HZ707C55</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cerebellum/Posterior Fossa</td>\n",
       "      <td>recurrence</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PT_ZZRBX5JT</th>\n",
       "      <td>7316-3319</td>\n",
       "      <td>Ependymal tumor</td>\n",
       "      <td>Ependymoma</td>\n",
       "      <td>Spine</td>\n",
       "      <td>WGS;RNA-Seq</td>\n",
       "      <td>Gross/Near total resection</td>\n",
       "      <td>Ependymoma</td>\n",
       "      <td>5.0</td>\n",
       "      <td>ependymoma</td>\n",
       "      <td>Solid Tissue</td>\n",
       "      <td>BS_9N3B3HZB;BS_YE1MAQYJ</td>\n",
       "      <td>0.503286</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Spinal Cord- Lumbar/Thecal Sac</td>\n",
       "      <td>recurrence</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PT_ZZRBX5JT</th>\n",
       "      <td>7316-496</td>\n",
       "      <td>Ependymal tumor</td>\n",
       "      <td>Ependymoma</td>\n",
       "      <td>Ventricles</td>\n",
       "      <td>WGS;RNA-Seq</td>\n",
       "      <td>Gross/Near total resection</td>\n",
       "      <td>Ependymoma</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ependymoma</td>\n",
       "      <td>Solid Tissue</td>\n",
       "      <td>BS_AGD2ATY1;BS_PJPGVD62</td>\n",
       "      <td>0.256150</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Ventricles</td>\n",
       "      <td>primary</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PT_ZZTQQVCT</th>\n",
       "      <td>7316-2723</td>\n",
       "      <td>Low-grade astrocytic tumor</td>\n",
       "      <td>Low-grade glioma/astrocytoma, wildtype</td>\n",
       "      <td>Posterior fossa</td>\n",
       "      <td>WGS;RNA-Seq</td>\n",
       "      <td>Gross/Near total resection</td>\n",
       "      <td>Low-grade glioma astrocytoma</td>\n",
       "      <td>1.0</td>\n",
       "      <td>low-grade glioma</td>\n",
       "      <td>Solid Tissue</td>\n",
       "      <td>BS_78TQS4Y3;BS_Y7F9E2E9</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Cerebellum/Posterior Fossa</td>\n",
       "      <td>primary</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PT_ZZWYKZZ8</th>\n",
       "      <td>7316-2758</td>\n",
       "      <td>Ependymal tumor</td>\n",
       "      <td>Ependymoma</td>\n",
       "      <td>Mixed</td>\n",
       "      <td>WGS;RNA-Seq</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>Ependymoma</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ependymoma</td>\n",
       "      <td>Solid Tissue</td>\n",
       "      <td>BS_2E81A3FT;BS_7B89YR6K</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Cerebellum/Posterior Fossa;Ventricles</td>\n",
       "      <td>progression</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1121 rows × 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              sampleId                 CANCER_TYPE  \\\n",
       "patientId                                            \n",
       "PT_00G007DM  7316-2577             Embryonal tumor   \n",
       "PT_00G007DM   7316-272             Embryonal tumor   \n",
       "PT_01MZ62KG   7316-447             Embryonal tumor   \n",
       "PT_02J5CWN5  7316-2989  Low-grade astrocytic tumor   \n",
       "PT_02J5CWN5   7316-898  Low-grade astrocytic tumor   \n",
       "...                ...                         ...   \n",
       "PT_ZZRBX5JT  7316-2661             Ependymal tumor   \n",
       "PT_ZZRBX5JT  7316-3319             Ependymal tumor   \n",
       "PT_ZZRBX5JT   7316-496             Ependymal tumor   \n",
       "PT_ZZTQQVCT  7316-2723  Low-grade astrocytic tumor   \n",
       "PT_ZZWYKZZ8  7316-2758             Ependymal tumor   \n",
       "\n",
       "                                          CANCER_TYPE_DETAILED  \\\n",
       "patientId                                                        \n",
       "PT_00G007DM                           CNS Embryonal tumor, NOS   \n",
       "PT_00G007DM  Embryonal tumor with multilayer rosettes, C19M...   \n",
       "PT_01MZ62KG  Embryonal tumor with multilayer rosettes, C19M...   \n",
       "PT_02J5CWN5                 Low-grade glioma/astrocytoma, FGFR   \n",
       "PT_02J5CWN5             Low-grade glioma/astrocytoma, wildtype   \n",
       "...                                                        ...   \n",
       "PT_ZZRBX5JT                                         Ependymoma   \n",
       "PT_ZZRBX5JT                                         Ependymoma   \n",
       "PT_ZZRBX5JT                                         Ependymoma   \n",
       "PT_ZZTQQVCT             Low-grade glioma/astrocytoma, wildtype   \n",
       "PT_ZZWYKZZ8                                         Ependymoma   \n",
       "\n",
       "                  CNS_REGION EXPERIMENTAL_STRATEGY  \\\n",
       "patientId                                            \n",
       "PT_00G007DM  Posterior fossa           WGS;RNA-Seq   \n",
       "PT_00G007DM  Posterior fossa           WGS;RNA-Seq   \n",
       "PT_01MZ62KG  Posterior fossa           RNA-Seq;WGS   \n",
       "PT_02J5CWN5      Suprasellar           WGS;RNA-Seq   \n",
       "PT_02J5CWN5      Suprasellar           RNA-Seq;WGS   \n",
       "...                      ...                   ...   \n",
       "PT_ZZRBX5JT  Posterior fossa           RNA-Seq;WGS   \n",
       "PT_ZZRBX5JT            Spine           WGS;RNA-Seq   \n",
       "PT_ZZRBX5JT       Ventricles           WGS;RNA-Seq   \n",
       "PT_ZZTQQVCT  Posterior fossa           WGS;RNA-Seq   \n",
       "PT_ZZWYKZZ8            Mixed           WGS;RNA-Seq   \n",
       "\n",
       "              EXTENT_OF_TUMOR_RESECTION  \\\n",
       "patientId                                 \n",
       "PT_00G007DM  Gross/Near total resection   \n",
       "PT_00G007DM  Gross/Near total resection   \n",
       "PT_01MZ62KG           Partial resection   \n",
       "PT_02J5CWN5           Partial resection   \n",
       "PT_02J5CWN5           Partial resection   \n",
       "...                                 ...   \n",
       "PT_ZZRBX5JT  Gross/Near total resection   \n",
       "PT_ZZRBX5JT  Gross/Near total resection   \n",
       "PT_ZZRBX5JT  Gross/Near total resection   \n",
       "PT_ZZTQQVCT  Gross/Near total resection   \n",
       "PT_ZZWYKZZ8                Not Reported   \n",
       "\n",
       "                                            HISTOLOGY  MUTATION_COUNT  \\\n",
       "patientId                                                               \n",
       "PT_00G007DM  Embryonal tumor with multilayer rosettes             2.0   \n",
       "PT_00G007DM  Embryonal tumor with multilayer rosettes             4.0   \n",
       "PT_01MZ62KG  Embryonal tumor with multilayer rosettes            11.0   \n",
       "PT_02J5CWN5              Low-grade glioma astrocytoma             7.0   \n",
       "PT_02J5CWN5              Low-grade glioma astrocytoma             1.0   \n",
       "...                                               ...             ...   \n",
       "PT_ZZRBX5JT                                Ependymoma            17.0   \n",
       "PT_ZZRBX5JT                                Ependymoma             5.0   \n",
       "PT_ZZRBX5JT                                Ependymoma             2.0   \n",
       "PT_ZZTQQVCT              Low-grade glioma astrocytoma             1.0   \n",
       "PT_ZZWYKZZ8                                Ependymoma             NaN   \n",
       "\n",
       "                                 PATHOLOGY_FREE_TEXT_DIAGNOSIS   SAMPLE_TYPE  \\\n",
       "patientId                                                                      \n",
       "PT_00G007DM                                   ependymoblastoma  Solid Tissue   \n",
       "PT_00G007DM  embryonal tumor with abundant neuropil and tru...  Solid Tissue   \n",
       "PT_01MZ62KG                                      pineoblastoma  Solid Tissue   \n",
       "PT_02J5CWN5                                   low-grade glioma  Solid Tissue   \n",
       "PT_02J5CWN5                                   low-grade glioma  Solid Tissue   \n",
       "...                                                        ...           ...   \n",
       "PT_ZZRBX5JT                                         ependymoma  Solid Tissue   \n",
       "PT_ZZRBX5JT                                         ependymoma  Solid Tissue   \n",
       "PT_ZZRBX5JT                                         ependymoma  Solid Tissue   \n",
       "PT_ZZTQQVCT                                   low-grade glioma  Solid Tissue   \n",
       "PT_ZZWYKZZ8                                         ependymoma  Solid Tissue   \n",
       "\n",
       "                         SPECIMEN_ID  TUMOR_FRACTION  TUMOR_PLOIDY  \\\n",
       "patientId                                                            \n",
       "PT_00G007DM  BS_AQMKA8NC;BS_FEPRNEXX        0.613054           4.0   \n",
       "PT_00G007DM  BS_K07KNTFY;BS_QWNBZ9RJ        0.328185           4.0   \n",
       "PT_01MZ62KG  BS_TE8QFF7T;BS_P39SQPTS             NaN           NaN   \n",
       "PT_02J5CWN5  BS_2VB8649B;BS_M8WP5T16        1.000000           2.0   \n",
       "PT_02J5CWN5  BS_5TT6TT4K;BS_1RXZ3BP7             NaN           NaN   \n",
       "...                              ...             ...           ...   \n",
       "PT_ZZRBX5JT  BS_JE1DF581;BS_HZ707C55             NaN           NaN   \n",
       "PT_ZZRBX5JT  BS_9N3B3HZB;BS_YE1MAQYJ        0.503286           2.0   \n",
       "PT_ZZRBX5JT  BS_AGD2ATY1;BS_PJPGVD62        0.256150           2.0   \n",
       "PT_ZZTQQVCT  BS_78TQS4Y3;BS_Y7F9E2E9        1.000000           2.0   \n",
       "PT_ZZWYKZZ8  BS_2E81A3FT;BS_7B89YR6K        1.000000           2.0   \n",
       "\n",
       "                                 TUMOR_TISSUE_SITE   TUMOR_TYPE  \n",
       "patientId                                                        \n",
       "PT_00G007DM             Cerebellum/Posterior Fossa   recurrence  \n",
       "PT_00G007DM             Cerebellum/Posterior Fossa      primary  \n",
       "PT_01MZ62KG             Cerebellum/Posterior Fossa  progression  \n",
       "PT_02J5CWN5     Suprasellar/Hypothalamic/Pituitary      primary  \n",
       "PT_02J5CWN5     Suprasellar/Hypothalamic/Pituitary  progression  \n",
       "...                                            ...          ...  \n",
       "PT_ZZRBX5JT             Cerebellum/Posterior Fossa   recurrence  \n",
       "PT_ZZRBX5JT         Spinal Cord- Lumbar/Thecal Sac   recurrence  \n",
       "PT_ZZRBX5JT                             Ventricles      primary  \n",
       "PT_ZZTQQVCT             Cerebellum/Posterior Fossa      primary  \n",
       "PT_ZZWYKZZ8  Cerebellum/Posterior Fossa;Ventricles  progression  \n",
       "\n",
       "[1121 rows x 15 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def import_pedcbioportal_metadata(path=\"openpbta-biosample-metadata.tsv\"):\n",
    "    path = pathlib.Path(path)\n",
    "    df = pd.read_csv(path, sep='\\t',index_col=0)\n",
    "    return df\n",
    "import_pedcbioportal_metadata()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "11be5b6c-efc5-4e1d-a1d0-ced22ddf24fa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['BS_CZRA594T',\n",
       " 'BS_59ZJWJTF',\n",
       " 'BS_ERAWW3H7',\n",
       " 'BS_QYPHA40N',\n",
       " 'BS_PKZ1HWNB',\n",
       " 'BS_M8EA6R2A',\n",
       " 'BS_FJEZ3ASV',\n",
       " 'BS_8ZD6J47V',\n",
       " 'BS_68TZMZH1',\n",
       " 'BS_0RQ4P069',\n",
       " 'BS_AFBPM6CN',\n",
       " 'BS_PNYN0AYD',\n",
       " 'BS_TX8C5VAJ',\n",
       " 'BS_GXTFW99H',\n",
       " 'BS_QWM9BPDY',\n",
       " 'BS_BWBDH9GM',\n",
       " 'BS_HM5GFJN8',\n",
       " 'BS_2A162JH9',\n",
       " 'BS_RXP2ZRQT',\n",
       " 'BS_40MP5BWR',\n",
       " 'BS_QZRP3NSG',\n",
       " 'BS_DVDT4VXQ',\n",
       " 'BS_6JBE0947',\n",
       " 'BS_YZD4SSMA',\n",
       " 'BS_M659G06J',\n",
       " 'BS_5GNQC2FF',\n",
       " 'BS_VXDGXQKZ',\n",
       " 'BS_ERFMPQN3',\n",
       " 'BS_PGK832G2',\n",
       " 'BS_4DQAQFQH',\n",
       " 'BS_XMP9XNR9',\n",
       " 'BS_P9JP6JFA',\n",
       " 'BS_JGKRN7NA',\n",
       " 'BS_MX23ZY0Y',\n",
       " 'BS_853PNV7P',\n",
       " 'BS_DRY58DTF',\n",
       " 'BS_TF5TTEXH',\n",
       " 'BS_E60JZ9Z3']"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_cbtn_cell_lines():\n",
    "    df = import_pedcbioportal_metadata()\n",
    "    df = df[df.SAMPLE_TYPE == \"Derived Cell Line\"]\n",
    "    return df.SPECIMEN_ID.str.cat(sep=';').split(';')\n",
    "\n",
    "#duplicated_cbtn_samples = [\"BS_DRVEFVQ5\",\"BS_169P1QCA\",\"BS_QG6V29H7\",\n",
    "#                           \"BS_6GV08HTE\",\"BS_B4PPG3X5\",\"BS_2X60Q1ET\",\"BS_S791VC80\",\"BS_STNH7YSX\",\"BS_3Z40EZHD\",\"BS_ZR75EKKX\",\n",
    "#                           \"BS_KQPCYZ2K\",\"BS_Z64NEPNE\",\"BS_KQRAHH6Y\",\"BS_KH3859M5\",\"BS_HJ7HYZ7N\",\"BS_AH3RVK53\",\n",
    "#                           \"BS_5S8VXASX\",\"BS_JEZBA2EW\",\"BS_XNYQS1WG\",\"BS_P99S85CY\",\n",
    "#                           \"BS_M5FM63EB\",\"BS_M0B42FPR\",\"BS_TX2WGF8K\",\"BS_RENPFFNK\",\"BS_R6CKWZW6\",\n",
    "#                           \"BS_ZSH09N84\",\"BS_6TMPZKSZ\",\"BS_B91XGSA5\",\"BS_XQF18WZP\",\"BS_0TCRV9AC\",\n",
    "#                           \"BS_2J4FG4HV\",\n",
    "#                           \"BS_EE73VE7V\",\"BS_5968GBGT\",\"BS_BQ81D2BP\",\"BS_3VKW5988\", # duplicate samples from PT_KTRJ8TFY autopsy\n",
    "#                           \"BS_AK9BV52G\",\"BS_X5VN0FW0\",\"BS_D6STCMQS\",\"BS_22VCR7DF\",\"BS_1Q524P3B\" # duplicate samples from PT_KZ56XHJT autopsy\n",
    "#                          ]\n",
    "#nontumor_samples = [\"BS_MCM78YPC\",\"BS_886M7JMG\",\"BS_TPX7YY57\"] # Epilepsy, Arteriovenous malformation, and Reactive connective tissue respectively\n",
    "get_cbtn_cell_lines()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "6bbdc8a7-0087-4458-b1d8-a450787afe01",
   "metadata": {},
   "outputs": [],
   "source": [
    "def propagate(df,dest,source,rename=False):\n",
    "    '''\n",
    "    Replace NA values in dest with those in source, then drop source and rename dest\n",
    "    '''\n",
    "    df[dest].fillna(df[source], inplace=True)\n",
    "    df.drop(source, inplace=True, axis=1)\n",
    "    if rename:\n",
    "        df = df.rename(columns={dest:rename})\n",
    "    return df\n",
    "def consensus(df,dest,source,rename=False):\n",
    "    '''\n",
    "    Check that values in dest and source agree, if not then set to NA. Drop source and rename dest.\n",
    "    '''\n",
    "    df.loc[df[dest] != df[source], dest] = None\n",
    "    df.drop(source, inplace=True, axis=1)\n",
    "    if rename:\n",
    "        df = df.rename(columns={dest:rename})\n",
    "    return df    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "68b7d2c6-d9ec-4796-bad4-fe0eef1498d7",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/vv/261hv1sn1h36hvx1wc1cv9w40000gp/T/ipykernel_49058/2192161946.py:6: DtypeWarning: Columns (1,21,24,25,26,27,33,36,37,41,43,45,47,48,49,51) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  df = pd.read_csv(path,sep='\\t',index_col=0)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning: The column 'harmonized_diagnosis' for sample 7316-1101 differs between CAVATICA and opentarget annotations.\n",
      "Warning: The column 'harmonized_diagnosis' for sample 7316-2935 differs between CAVATICA and opentarget annotations.\n",
      "Warning: The column 'tumor_descriptor' for sample 7316-3217 differs between CAVATICA and opentarget annotations.\n",
      "Warning: The column 'primary_site' for sample 7316-3229 differs between CAVATICA and opentarget annotations.\n",
      "Warning: The column 'age_at_diagnosis_days' for sample 7316-323 differs between CAVATICA and opentarget annotations.\n",
      "Warning: The column 'primary_site' for sample 7316-3230 differs between CAVATICA and opentarget annotations.\n",
      "Warning: The column 'CNS_region' for sample 7316-3230 differs between CAVATICA and opentarget annotations.\n",
      "Warning: The column 'primary_site' for sample 7316-3231 differs between CAVATICA and opentarget annotations.\n",
      "Warning: The column 'harmonized_diagnosis' for sample 7316-3295 differs between CAVATICA and opentarget annotations.\n",
      "Warning: The column 'harmonized_diagnosis' for sample 7316-3632 differs between CAVATICA and opentarget annotations.\n",
      "Warning: The column 'harmonized_diagnosis' for sample 7316-4062 differs between CAVATICA and opentarget annotations.\n",
      "Warning: The column 'harmonized_diagnosis' for sample 7316-921 differs between CAVATICA and opentarget annotations.\n",
      "53 KF biospecimens missing from the opentarget histologies table...\n",
      "found 41 matching external sample IDs, adding to table...\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sex</th>\n",
       "      <th>patient_id</th>\n",
       "      <th>disease_type</th>\n",
       "      <th>external_sample_id</th>\n",
       "      <th>tumor_history</th>\n",
       "      <th>age_at_diagnosis</th>\n",
       "      <th>cohort</th>\n",
       "      <th>extent_of_tumor_resection</th>\n",
       "      <th>dkfz_v11_methylation_subclass</th>\n",
       "      <th>dkfz_v11_methylation_subclass_score</th>\n",
       "      <th>dkfz_v12_methylation_subclass</th>\n",
       "      <th>dkfz_v12_methylation_subclass_score</th>\n",
       "      <th>molecular_subtype</th>\n",
       "      <th>harmonized_diagnosis</th>\n",
       "      <th>broad_histology</th>\n",
       "      <th>short_histology</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>biosample_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>BS_85Y70BHE</th>\n",
       "      <td>Female</td>\n",
       "      <td>PT_124TTG01</td>\n",
       "      <td>Adenoma</td>\n",
       "      <td>7316-1960</td>\n",
       "      <td>Diagnosis</td>\n",
       "      <td>5916</td>\n",
       "      <td>PBTA-X00</td>\n",
       "      <td>Gross/Near total resection</td>\n",
       "      <td>PITAD, STH DNS A</td>\n",
       "      <td>0.857</td>\n",
       "      <td>PITAD_STH_DENSE1</td>\n",
       "      <td>0.967892</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Adenoma</td>\n",
       "      <td>Benign tumor</td>\n",
       "      <td>Adenoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BS_A9S5HT6P</th>\n",
       "      <td>Female</td>\n",
       "      <td>PT_9XAA8PF6</td>\n",
       "      <td>Adenoma</td>\n",
       "      <td>7316-1069</td>\n",
       "      <td>Diagnosis</td>\n",
       "      <td>4964</td>\n",
       "      <td>PBTA-X00</td>\n",
       "      <td>Gross/Near total resection</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Adenoma</td>\n",
       "      <td>Benign tumor</td>\n",
       "      <td>Adenoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BS_1EQGGHHQ</th>\n",
       "      <td>Female</td>\n",
       "      <td>PT_XP8AM1CH</td>\n",
       "      <td>Adenoma</td>\n",
       "      <td>7316-1096</td>\n",
       "      <td>Diagnosis</td>\n",
       "      <td>5713</td>\n",
       "      <td>PBTA-X00</td>\n",
       "      <td>Gross/Near total resection</td>\n",
       "      <td>PITAD, ACTH</td>\n",
       "      <td>1.000</td>\n",
       "      <td>PITAD_ACTH</td>\n",
       "      <td>0.999998</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Adenoma</td>\n",
       "      <td>Benign tumor</td>\n",
       "      <td>Adenoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BS_BR1X3R68</th>\n",
       "      <td>Female</td>\n",
       "      <td>PT_ZN1F7RMP</td>\n",
       "      <td>Adenoma</td>\n",
       "      <td>7316-1080</td>\n",
       "      <td>Diagnosis</td>\n",
       "      <td>7032</td>\n",
       "      <td>PBTA-X00</td>\n",
       "      <td>Partial resection</td>\n",
       "      <td>PITAD, TSH</td>\n",
       "      <td>0.981</td>\n",
       "      <td>PITAD_TSH</td>\n",
       "      <td>0.996686</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Adenoma</td>\n",
       "      <td>Benign tumor</td>\n",
       "      <td>Adenoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BS_JDMM2XG4</th>\n",
       "      <td>Male</td>\n",
       "      <td>PT_32J909WM</td>\n",
       "      <td>Atypical Teratoid Rhabdoid Tumor (ATRT)</td>\n",
       "      <td>7316-2688</td>\n",
       "      <td>Diagnosis</td>\n",
       "      <td>860</td>\n",
       "      <td>PBTA-X00</td>\n",
       "      <td>Gross/Near total resection</td>\n",
       "      <td>ATRT, TYR</td>\n",
       "      <td>0.594</td>\n",
       "      <td>ATRT_SHH</td>\n",
       "      <td>0.105467</td>\n",
       "      <td>ATRT, To be classified</td>\n",
       "      <td>Atypical Teratoid Rhabdoid Tumor (ATRT)</td>\n",
       "      <td>Embryonal tumor</td>\n",
       "      <td>ATRT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 sex   patient_id                             disease_type  \\\n",
       "biosample_id                                                                 \n",
       "BS_85Y70BHE   Female  PT_124TTG01                                  Adenoma   \n",
       "BS_A9S5HT6P   Female  PT_9XAA8PF6                                  Adenoma   \n",
       "BS_1EQGGHHQ   Female  PT_XP8AM1CH                                  Adenoma   \n",
       "BS_BR1X3R68   Female  PT_ZN1F7RMP                                  Adenoma   \n",
       "BS_JDMM2XG4     Male  PT_32J909WM  Atypical Teratoid Rhabdoid Tumor (ATRT)   \n",
       "\n",
       "             external_sample_id tumor_history  age_at_diagnosis    cohort  \\\n",
       "biosample_id                                                                \n",
       "BS_85Y70BHE           7316-1960     Diagnosis              5916  PBTA-X00   \n",
       "BS_A9S5HT6P           7316-1069     Diagnosis              4964  PBTA-X00   \n",
       "BS_1EQGGHHQ           7316-1096     Diagnosis              5713  PBTA-X00   \n",
       "BS_BR1X3R68           7316-1080     Diagnosis              7032  PBTA-X00   \n",
       "BS_JDMM2XG4           7316-2688     Diagnosis               860  PBTA-X00   \n",
       "\n",
       "               extent_of_tumor_resection dkfz_v11_methylation_subclass  \\\n",
       "biosample_id                                                             \n",
       "BS_85Y70BHE   Gross/Near total resection              PITAD, STH DNS A   \n",
       "BS_A9S5HT6P   Gross/Near total resection                           NaN   \n",
       "BS_1EQGGHHQ   Gross/Near total resection                   PITAD, ACTH   \n",
       "BS_BR1X3R68            Partial resection                    PITAD, TSH   \n",
       "BS_JDMM2XG4   Gross/Near total resection                     ATRT, TYR   \n",
       "\n",
       "              dkfz_v11_methylation_subclass_score  \\\n",
       "biosample_id                                        \n",
       "BS_85Y70BHE                                 0.857   \n",
       "BS_A9S5HT6P                                   NaN   \n",
       "BS_1EQGGHHQ                                 1.000   \n",
       "BS_BR1X3R68                                 0.981   \n",
       "BS_JDMM2XG4                                 0.594   \n",
       "\n",
       "             dkfz_v12_methylation_subclass  \\\n",
       "biosample_id                                 \n",
       "BS_85Y70BHE               PITAD_STH_DENSE1   \n",
       "BS_A9S5HT6P                            NaN   \n",
       "BS_1EQGGHHQ                     PITAD_ACTH   \n",
       "BS_BR1X3R68                      PITAD_TSH   \n",
       "BS_JDMM2XG4                       ATRT_SHH   \n",
       "\n",
       "              dkfz_v12_methylation_subclass_score       molecular_subtype  \\\n",
       "biosample_id                                                                \n",
       "BS_85Y70BHE                              0.967892                     NaN   \n",
       "BS_A9S5HT6P                                   NaN                     NaN   \n",
       "BS_1EQGGHHQ                              0.999998                     NaN   \n",
       "BS_BR1X3R68                              0.996686                     NaN   \n",
       "BS_JDMM2XG4                              0.105467  ATRT, To be classified   \n",
       "\n",
       "                                 harmonized_diagnosis  broad_histology  \\\n",
       "biosample_id                                                             \n",
       "BS_85Y70BHE                                   Adenoma     Benign tumor   \n",
       "BS_A9S5HT6P                                   Adenoma     Benign tumor   \n",
       "BS_1EQGGHHQ                                   Adenoma     Benign tumor   \n",
       "BS_BR1X3R68                                   Adenoma     Benign tumor   \n",
       "BS_JDMM2XG4   Atypical Teratoid Rhabdoid Tumor (ATRT)  Embryonal tumor   \n",
       "\n",
       "             short_histology  \n",
       "biosample_id                  \n",
       "BS_85Y70BHE          Adenoma  \n",
       "BS_A9S5HT6P          Adenoma  \n",
       "BS_1EQGGHHQ          Adenoma  \n",
       "BS_BR1X3R68          Adenoma  \n",
       "BS_JDMM2XG4             ATRT  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "## Integrate all CBTN data available\n",
    "def generate_cbtn_biosample_table(verbose=0):\n",
    "    '''\n",
    "    Generate a metadata table of cbtn biosamples.\n",
    "    verbose:\n",
    "        0: most useful metadata only. These are included in the Supplmentary Table.\n",
    "        1: includes some extra columns. Useful for generating the patient metadata table.\n",
    "        2: includes lots of extra columns\n",
    "    '''\n",
    "    df = pd.DataFrame(index=get_pedpancan_biosamples_from_AC())\n",
    "    cavatica_data = import_cbtn_biosample_metadata()\n",
    "    df = pd.merge(left=df,how='inner',right=cavatica_data,left_index=True,right_index=True)\n",
    "    opentarget_data = import_opentarget_histologies_files()\n",
    "    df = pd.merge(left=df,how='left',right=opentarget_data,left_index=True,right_index=True,suffixes=(None,\"_y\"))\n",
    "\n",
    "    # For CAVATICA annotations which are missing, propagate those from opentarget.\n",
    "    df = propagate(df,\"primary_site\",\"primary_site_y\")\n",
    "    df = propagate(df,\"age_at_diagnosis\",\"age_at_diagnosis_days\")\n",
    "    df = propagate(df,\"Tumor Descriptor\",\"tumor_descriptor\")\n",
    "    df = consensus(df,\"gender\",\"reported_gender\",\"sex\")\n",
    "    \n",
    "    # Rename columns\n",
    "    df.index.name = \"biosample_id\"\n",
    "    df = df.rename(columns={\n",
    "        'Kids First Participant ID':'patient_id',\n",
    "        'Tumor Descriptor':'tumor_history',\n",
    "        'case_id':'external_patient_id',\n",
    "        'sample_id':'external_sample_id',\n",
    "    })\n",
    "\n",
    "    # drop columns\n",
    "    if verbose < 2:\n",
    "        df = df.drop([\"race\",\"ethnicity\",\"external_patient_id\",\"WGS_UUID\",\"Kids First Biospecimen ID Normal\",\n",
    "                      \"sample_id_y\",\"composition\",\"Kids_First_Participant_ID\",\"experimental_strategy\",\"sample_type\",\n",
    "                      \"germline_sex_estimate\",\"race_y\",\"ethnicity_y\",\"molecular_subtype_methyl\",\"cohort_participant_id\",\"Notes\"\n",
    "                     ],axis=1)  \n",
    "    if verbose < 1:\n",
    "        df = df.drop([\"primary_site\",\"pathology_diagnosis\",\"OS_days\",\"OS_status\",\"EFS_days\",\"age_last_update_days\",\"aliquot_id\",\n",
    "                      \"cancer_predispositions\",\"CNS_region\",\"age_at_chemo_start\",\"age_at_radiation_start\",\"cancer_group\",\n",
    "                      \"age_at_event_days\",\"clinical_status_at_event\"\n",
    "                     ],axis=1)\n",
    "    # Drop nontumor samples\n",
    "    #df = df.drop(nontumor_samples)\n",
    "    \n",
    "    # Drop cell lines\n",
    "    #df = df[~df.index.isin(get_cbtn_cell_lines())]\n",
    "    \n",
    "    # Mark duplicates\n",
    "    #df['in_deduplicated_sample_cohort'] = True\n",
    "    #df.loc[duplicated_cbtn_samples,'in_deduplicated_sample_cohort'] = False\n",
    "    return df\n",
    "\n",
    "## SJ data\n",
    "# duplicated_sj_samples are biosamples where we suspect that the same tumor was sequenced twice at the same timepoint. \n",
    "# These are flagged and removed arbitrarily.\n",
    "#duplicated_sj_samples = ['SJST030043_D1','SJST030131_D1','SJMEL001003_D2','SJOS001115_D1','SJMB009_E',\n",
    "#                         'SJDSRCT030041_D3','SJBT030081_D2','SJWLM030180_D2','SJHGG017_D','SJEWS030228_D2',\n",
    "#                         'SJST030383_D1','SJLGG017_D','SJLGG030611_D2','SJOS030876_D2','SJOS001101_M1','SJHM030702_D2']\n",
    "# what to do with SJOS001101_M1?\n",
    "CBTN_df = generate_cbtn_biosample_table()\n",
    "CBTN_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "61343b81-56fd-4030-bbdc-1bfd4c44db84",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>patient_id</th>\n",
       "      <th>tumor_history</th>\n",
       "      <th>sj_long_disease_name</th>\n",
       "      <th>cohort</th>\n",
       "      <th>sj_diseases</th>\n",
       "      <th>age_at_diagnosis</th>\n",
       "      <th>sex</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>biosample_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SJACT001_D</th>\n",
       "      <td>SJ006401</td>\n",
       "      <td>Diagnosis</td>\n",
       "      <td>Adrenocortical Carcinoma</td>\n",
       "      <td>SJC-DS-1001</td>\n",
       "      <td>ACC</td>\n",
       "      <td>1088.0</td>\n",
       "      <td>Female</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SJACT002_D</th>\n",
       "      <td>SJ006402</td>\n",
       "      <td>Diagnosis</td>\n",
       "      <td>Adrenocortical Carcinoma</td>\n",
       "      <td>SJC-DS-1001</td>\n",
       "      <td>ACC</td>\n",
       "      <td>961.0</td>\n",
       "      <td>Female</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SJACT003_D</th>\n",
       "      <td>SJ006403</td>\n",
       "      <td>Diagnosis</td>\n",
       "      <td>Adrenocortical Carcinoma</td>\n",
       "      <td>SJC-DS-1001</td>\n",
       "      <td>ACC</td>\n",
       "      <td>475.0</td>\n",
       "      <td>Female</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SJACT004_D</th>\n",
       "      <td>SJ006404</td>\n",
       "      <td>Diagnosis</td>\n",
       "      <td>Adrenocortical Carcinoma</td>\n",
       "      <td>SJC-DS-1001</td>\n",
       "      <td>ACC</td>\n",
       "      <td>1819.0</td>\n",
       "      <td>Female</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SJACT005_D</th>\n",
       "      <td>SJ006405</td>\n",
       "      <td>Diagnosis</td>\n",
       "      <td>Adrenocortical Carcinoma</td>\n",
       "      <td>SJC-DS-1001</td>\n",
       "      <td>ACC</td>\n",
       "      <td>5621.0</td>\n",
       "      <td>Female</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             patient_id tumor_history      sj_long_disease_name       cohort  \\\n",
       "biosample_id                                                                   \n",
       "SJACT001_D     SJ006401     Diagnosis  Adrenocortical Carcinoma  SJC-DS-1001   \n",
       "SJACT002_D     SJ006402     Diagnosis  Adrenocortical Carcinoma  SJC-DS-1001   \n",
       "SJACT003_D     SJ006403     Diagnosis  Adrenocortical Carcinoma  SJC-DS-1001   \n",
       "SJACT004_D     SJ006404     Diagnosis  Adrenocortical Carcinoma  SJC-DS-1001   \n",
       "SJACT005_D     SJ006405     Diagnosis  Adrenocortical Carcinoma  SJC-DS-1001   \n",
       "\n",
       "             sj_diseases  age_at_diagnosis     sex  \n",
       "biosample_id                                        \n",
       "SJACT001_D           ACC            1088.0  Female  \n",
       "SJACT002_D           ACC             961.0  Female  \n",
       "SJACT003_D           ACC             475.0  Female  \n",
       "SJACT004_D           ACC            1819.0  Female  \n",
       "SJACT005_D           ACC            5621.0  Female  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def import_sj_sample_info(path=\"SAMPLE_INFO_2022-03-02.tsv\"):\n",
    "    path = pathlib.Path(path)\n",
    "    df = pd.read_csv(path,sep='\\t',index_col=\"sample_name\")\n",
    "    return df\n",
    "def clean_sj_biosample_metadata(df):\n",
    "    '''\n",
    "    Clean known errors in the sj metadata, and unify ontologies, units etc.\n",
    "    '''\n",
    "    df = df.replace({\n",
    "        'attr_age_at_diagnosis':{\n",
    "            \"Not Available\": None\n",
    "        },\n",
    "        'attr_sex':{\n",
    "            \"Not Available\":None\n",
    "        }\n",
    "    })\n",
    "    \n",
    "    # Convert age from years to days\n",
    "    df['attr_age_at_diagnosis'] = (pd.to_numeric(df['attr_age_at_diagnosis'],errors='coerce')*365.25).round()\n",
    "    \n",
    "    return df\n",
    "def generate_sj_biosample_table(verbose=0):\n",
    "    '''\n",
    "    Notes:\n",
    "    sj_diseases != attr_oncotree_disease_code = sj_associated_diagnoses_disease_code\n",
    "    attr_diagnosis != sj_long_disease_name != sj_associated_diagnoses\n",
    "    '''\n",
    "    df = pd.DataFrame(index=get_pedpancan_biosamples_from_AC())\n",
    "    columns = ['subject_name','sample_type','attr_age_at_diagnosis','attr_sex','sj_long_disease_name','sj_diseases','attr_oncotree_disease_code','sj_dataset_accessions']\n",
    "    add = import_sj_sample_info()\n",
    "    add = clean_sj_biosample_metadata(add)\n",
    "    add = add[(add.sequencing_type == 'WGS') & (add.file_type == 'BAM') & add.file_path.str.endswith('.bam')]\n",
    "#    add = add.sort_values(columns)\n",
    "#    add = add.loc[~add.index.duplicated()]\n",
    "    df = pd.merge(left=df,how='inner',right=add, left_index=True, right_index=True)\n",
    "    \n",
    "    # Rename columns\n",
    "    df.index.name = \"biosample_id\"\n",
    "    df = df.rename(columns={\n",
    "        'subject_name':'patient_id',\n",
    "        'sample_type':'tumor_history',\n",
    "        'attr_sex':'sex',\n",
    "        'sj_dataset_accessions':'cohort',\n",
    "        'attr_age_at_diagnosis':'age_at_diagnosis',\n",
    "    })\n",
    "    # drop columns\n",
    "    if verbose < 2:\n",
    "        df = df.drop([\"file_path\",\"file_id\",\"sequencing_type\",\"file_type\",\"description\",\"sj_embargo_date\",\"attr_ethnicity\",\"attr_race\",\n",
    "                      \"sj_genome_build\",\"sj_pipeline_name\",\"sj_pipeline_version\",\"attr_library_selection_protocol\",\"attr_read_length\",\n",
    "                      \"attr_sequencing_platform\",\"attr_read_type\",\"attr_tissue_preservative\",\"attr_inferred_strandedness\",\n",
    "                      \"attr_lab_strandedness\",\"attr_germline_sample\"\n",
    "        ],axis=1)\n",
    "    if verbose < 1:\n",
    "        df = df.drop([\"sj_pmid_accessions\",\"sj_publication_titles\",\"sj_pub_accessions\",\"sj_datasets\",\"sj_ega_accessions\",\"attr_diagnosis\",\n",
    "                      \"attr_diagnosis_group\",\"attr_oncotree_disease_code\",\"attr_subtype_biomarkers\",\"sj_associated_diagnoses\",\n",
    "                      \"sj_associated_diagnoses_disease_code\"\n",
    "        ],axis=1)\n",
    "    \n",
    "    # Mark duplicates\n",
    "    #df['in_deduplicated_sample_cohort'] = True\n",
    "    #df.loc[duplicated_sj_samples,'in_deduplicated_sample_cohort'] = False\n",
    "    return df\n",
    "\n",
    "SJ_df = import_sj_sample_info()\n",
    "SJ_df = clean_sj_biosample_metadata(SJ_df)\n",
    "SJ_df = generate_sj_biosample_table()\n",
    "SJ_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "577a17cd-b348-47a2-8105-e0696d969a81",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sex</th>\n",
       "      <th>patient_id</th>\n",
       "      <th>external_sample_id</th>\n",
       "      <th>tumor_history</th>\n",
       "      <th>age_at_diagnosis</th>\n",
       "      <th>cohort</th>\n",
       "      <th>extent_of_tumor_resection</th>\n",
       "      <th>cancer_subtype</th>\n",
       "      <th>cancer_type</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>biosample_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>BS_85Y70BHE</th>\n",
       "      <td>Female</td>\n",
       "      <td>PT_124TTG01</td>\n",
       "      <td>7316-1960</td>\n",
       "      <td>Diagnosis</td>\n",
       "      <td>5916.0</td>\n",
       "      <td>PBTA-X00</td>\n",
       "      <td>Gross/Near total resection</td>\n",
       "      <td>PITAD_STH_DENSE1</td>\n",
       "      <td>BENG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BS_A9S5HT6P</th>\n",
       "      <td>Female</td>\n",
       "      <td>PT_9XAA8PF6</td>\n",
       "      <td>7316-1069</td>\n",
       "      <td>Diagnosis</td>\n",
       "      <td>4964.0</td>\n",
       "      <td>PBTA-X00</td>\n",
       "      <td>Gross/Near total resection</td>\n",
       "      <td>Adenoma</td>\n",
       "      <td>BENG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BS_1EQGGHHQ</th>\n",
       "      <td>Female</td>\n",
       "      <td>PT_XP8AM1CH</td>\n",
       "      <td>7316-1096</td>\n",
       "      <td>Diagnosis</td>\n",
       "      <td>5713.0</td>\n",
       "      <td>PBTA-X00</td>\n",
       "      <td>Gross/Near total resection</td>\n",
       "      <td>PITAD_ACTH</td>\n",
       "      <td>BENG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BS_BR1X3R68</th>\n",
       "      <td>Female</td>\n",
       "      <td>PT_ZN1F7RMP</td>\n",
       "      <td>7316-1080</td>\n",
       "      <td>Diagnosis</td>\n",
       "      <td>7032.0</td>\n",
       "      <td>PBTA-X00</td>\n",
       "      <td>Partial resection</td>\n",
       "      <td>PITAD_TSH</td>\n",
       "      <td>BENG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BS_JDMM2XG4</th>\n",
       "      <td>Male</td>\n",
       "      <td>PT_32J909WM</td>\n",
       "      <td>7316-2688</td>\n",
       "      <td>Diagnosis</td>\n",
       "      <td>860.0</td>\n",
       "      <td>PBTA-X00</td>\n",
       "      <td>Gross/Near total resection</td>\n",
       "      <td>ATRT, To be classified</td>\n",
       "      <td>ATRT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 sex   patient_id external_sample_id tumor_history  \\\n",
       "biosample_id                                                         \n",
       "BS_85Y70BHE   Female  PT_124TTG01          7316-1960     Diagnosis   \n",
       "BS_A9S5HT6P   Female  PT_9XAA8PF6          7316-1069     Diagnosis   \n",
       "BS_1EQGGHHQ   Female  PT_XP8AM1CH          7316-1096     Diagnosis   \n",
       "BS_BR1X3R68   Female  PT_ZN1F7RMP          7316-1080     Diagnosis   \n",
       "BS_JDMM2XG4     Male  PT_32J909WM          7316-2688     Diagnosis   \n",
       "\n",
       "              age_at_diagnosis    cohort   extent_of_tumor_resection  \\\n",
       "biosample_id                                                           \n",
       "BS_85Y70BHE             5916.0  PBTA-X00  Gross/Near total resection   \n",
       "BS_A9S5HT6P             4964.0  PBTA-X00  Gross/Near total resection   \n",
       "BS_1EQGGHHQ             5713.0  PBTA-X00  Gross/Near total resection   \n",
       "BS_BR1X3R68             7032.0  PBTA-X00           Partial resection   \n",
       "BS_JDMM2XG4              860.0  PBTA-X00  Gross/Near total resection   \n",
       "\n",
       "                      cancer_subtype cancer_type  \n",
       "biosample_id                                      \n",
       "BS_85Y70BHE         PITAD_STH_DENSE1        BENG  \n",
       "BS_A9S5HT6P                  Adenoma        BENG  \n",
       "BS_1EQGGHHQ               PITAD_ACTH        BENG  \n",
       "BS_BR1X3R68                PITAD_TSH        BENG  \n",
       "BS_JDMM2XG4   ATRT, To be classified        ATRT  "
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Function to create the cancer_subtype column based on priority\n",
    "\n",
    "def get_subtype(row):\n",
    "    priority_columns = ['molecular_subtype','dkfz_v12_methylation_subclass',\n",
    "                    'dkfz_v11_methylation_subclass', 'harmonized_diagnosis', 'disease_type', \"sj_diseases\"]  # Add other columns as needed\n",
    "    for col in priority_columns:\n",
    "        if col in ['dkfz_v12_methylation_subclass', 'dkfz_v11_methylation_subclass'] and pd.notnull(row[col]) and row[f\"{col}_score\"] > 0.9:\n",
    "            return row[col]\n",
    "        elif col not in ['dkfz_v12_methylation_subclass', 'dkfz_v11_methylation_subclass'] and pd.notnull(row[col]):\n",
    "            return row[col]\n",
    "    return None\n",
    "def unify_tumor_diagnoses(df):\n",
    "    # Apply the function to create the cancer_subtype column\n",
    "    mapping = pd.read_excel(\"pedpancan_mapping.xlsx\", sheet_name='filtered_mapping')\n",
    "    \n",
    "    # Create a dictionary from the mapping\n",
    "    mapping_dict = dict(zip(mapping['Classification'], mapping['Owen']))\n",
    "    \n",
    "    # Create 'cancer_type_pre' column using the get_subtype function\n",
    "    df['cancer_subtype'] = df.apply(get_subtype, axis=1)\n",
    "    \n",
    "    # Map 'cancer_type_pre' to 'owen' using mapping_dict\n",
    "    df['cancer_type'] = df['cancer_subtype'].map(mapping_dict)\n",
    "    \n",
    "    # # Create 'cancer_type' column as a copy of 'cancer_type_pre'\n",
    "    # df['cancer_type'] = df['cancer_type_pre'].copy()\n",
    "    \n",
    "    # # if owen column is na for a specific sample it gets assigned to get_subtype() column//\n",
    "    # df['cancer_type'] = df['cancer_type_post'].combine_first(df['cancer_type'])\n",
    "\n",
    "    df = df.drop([\"disease_type\",\"dkfz_v11_methylation_subclass\",\"dkfz_v11_methylation_subclass_score\",\n",
    "                  \"dkfz_v12_methylation_subclass\",\"dkfz_v12_methylation_subclass_score\",\"molecular_subtype\",\"harmonized_diagnosis\",\n",
    "                  \"broad_histology\",\"short_histology\", \"sj_long_disease_name\", \"sj_diseases\"\n",
    "                 ],axis=1)\n",
    "    return df\n",
    "df = pd.concat([CBTN_df, SJ_df])\n",
    "df = unify_tumor_diagnoses(df)\n",
    "df.head()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "7a0dae6a-52f6-420c-9f68-eb04681d3440",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Annotate with ecDNA status\n",
    "def annotate_with_ecDNA(df,path=\"Supplementary_Tables.xlsx\"):\n",
    "    '''\n",
    "    Annotate biosamples with ecDNA status.\n",
    "    Inputs:\n",
    "        df: pd.DataFrame. Must be indexed by biosample.\n",
    "        path: path to AmpliconClassifier results.\n",
    "    '''\n",
    "    # load AC results\n",
    "    if path.endswith(\"Supplementary_Tables.xlsx\"):\n",
    "        ac = pd.read_excel(path,sheet_name=\"3. Amplicons\")\n",
    "    else:\n",
    "        ac = pd.read_excel(path,index_col=0)\n",
    "    \n",
    "    # Aggregate by biosample\n",
    "    ac_agg = ac.groupby(\"sample_name\").sum().ecDNA_amplicons\n",
    "    df = df.join(ac_agg)\n",
    "    df = df.rename(columns={\"ecDNA_amplicons\":\"ecDNA_sequences_detected\"})\n",
    "    df[\"ecDNA_sequences_detected\"].fillna(0,inplace=True)\n",
    "    return df\n",
    "\n",
    "def generate_biosample_table():\n",
    "    df = pd.concat([SJ_df, CBTN_df])\n",
    "    df = unify_tumor_diagnoses(df)\n",
    "    df = annotate_with_ecDNA(df)\n",
    "    return df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "ed9d7ecb-5894-40d1-b143-53267961489e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>patient_id</th>\n",
       "      <th>tumor_history</th>\n",
       "      <th>cohort</th>\n",
       "      <th>age_at_diagnosis</th>\n",
       "      <th>sex</th>\n",
       "      <th>external_sample_id</th>\n",
       "      <th>extent_of_tumor_resection</th>\n",
       "      <th>cancer_subtype</th>\n",
       "      <th>cancer_type</th>\n",
       "      <th>ecDNA_sequences_detected</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>biosample_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>BS_007JTNB8</th>\n",
       "      <td>PT_1MW98VR1</td>\n",
       "      <td>Diagnosis</td>\n",
       "      <td>PBTA-X00</td>\n",
       "      <td>1872.0</td>\n",
       "      <td>Male</td>\n",
       "      <td>7316-2558</td>\n",
       "      <td>Gross/Near total resection</td>\n",
       "      <td>EPN, PF A</td>\n",
       "      <td>EP</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BS_00TRPEQX</th>\n",
       "      <td>PT_MDWPRDBT</td>\n",
       "      <td>Recurrence</td>\n",
       "      <td>PBTA-X00</td>\n",
       "      <td>2146.0</td>\n",
       "      <td>Female</td>\n",
       "      <td>7316-2660</td>\n",
       "      <td>Gross/Near total resection</td>\n",
       "      <td>HGG, H3 wildtype</td>\n",
       "      <td>HGG</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BS_01DQH017</th>\n",
       "      <td>PT_CYVVA9AB</td>\n",
       "      <td>Diagnosis</td>\n",
       "      <td>PBTA-X00</td>\n",
       "      <td>4738.0</td>\n",
       "      <td>Female</td>\n",
       "      <td>7316-3299</td>\n",
       "      <td>Gross/Near total resection</td>\n",
       "      <td>EPN, ST ZFTA</td>\n",
       "      <td>EP</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BS_01Y5F4PN</th>\n",
       "      <td>PT_QK4XB83F</td>\n",
       "      <td>Diagnosis</td>\n",
       "      <td>PBTA-X00</td>\n",
       "      <td>1546.0</td>\n",
       "      <td>Female</td>\n",
       "      <td>7316-663</td>\n",
       "      <td>Gross/Near total resection</td>\n",
       "      <td>SEGA, To be classified</td>\n",
       "      <td>SEGA</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BS_02W5H7K5</th>\n",
       "      <td>PT_J06X17XH</td>\n",
       "      <td>Diagnosis</td>\n",
       "      <td>PBTA-X00</td>\n",
       "      <td>5146.0</td>\n",
       "      <td>Male</td>\n",
       "      <td>7316-1649</td>\n",
       "      <td>Gross/Near total resection</td>\n",
       "      <td>Rosai-dorfman</td>\n",
       "      <td>BENG</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               patient_id tumor_history    cohort  age_at_diagnosis     sex  \\\n",
       "biosample_id                                                                  \n",
       "BS_007JTNB8   PT_1MW98VR1     Diagnosis  PBTA-X00            1872.0    Male   \n",
       "BS_00TRPEQX   PT_MDWPRDBT    Recurrence  PBTA-X00            2146.0  Female   \n",
       "BS_01DQH017   PT_CYVVA9AB     Diagnosis  PBTA-X00            4738.0  Female   \n",
       "BS_01Y5F4PN   PT_QK4XB83F     Diagnosis  PBTA-X00            1546.0  Female   \n",
       "BS_02W5H7K5   PT_J06X17XH     Diagnosis  PBTA-X00            5146.0    Male   \n",
       "\n",
       "             external_sample_id   extent_of_tumor_resection  \\\n",
       "biosample_id                                                  \n",
       "BS_007JTNB8           7316-2558  Gross/Near total resection   \n",
       "BS_00TRPEQX           7316-2660  Gross/Near total resection   \n",
       "BS_01DQH017           7316-3299  Gross/Near total resection   \n",
       "BS_01Y5F4PN            7316-663  Gross/Near total resection   \n",
       "BS_02W5H7K5           7316-1649  Gross/Near total resection   \n",
       "\n",
       "                      cancer_subtype cancer_type  ecDNA_sequences_detected  \n",
       "biosample_id                                                                \n",
       "BS_007JTNB8                EPN, PF A          EP                       0.0  \n",
       "BS_00TRPEQX         HGG, H3 wildtype         HGG                       1.0  \n",
       "BS_01DQH017             EPN, ST ZFTA          EP                       0.0  \n",
       "BS_01Y5F4PN   SEGA, To be classified        SEGA                       0.0  \n",
       "BS_02W5H7K5            Rosai-dorfman        BENG                       0.0  "
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = generate_biosample_table()\n",
    "df.head()\n",
    "# df.to_csv(\"test_biosample_table.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "ba81c4d7-20b5-451d-bdb1-3777d3f34399",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1717\n",
      "178\n",
      "0\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>patient_id</th>\n",
       "      <th>tumor_history</th>\n",
       "      <th>cohort</th>\n",
       "      <th>age_at_diagnosis</th>\n",
       "      <th>sex</th>\n",
       "      <th>external_sample_id</th>\n",
       "      <th>extent_of_tumor_resection</th>\n",
       "      <th>cancer_subtype</th>\n",
       "      <th>cancer_type</th>\n",
       "      <th>ecDNA_sequences_detected</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>biosample_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [patient_id, tumor_history, cohort, age_at_diagnosis, sex, external_sample_id, extent_of_tumor_resection, cancer_subtype, cancer_type, ecDNA_sequences_detected]\n",
       "Index: []"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "pd.set_option('display.max_rows', None)\n",
    "pd.set_option('display.max_columns', None)\n",
    "print(len(df))\n",
    "print(len(df[df.ecDNA_sequences_detected > 0.0]))\n",
    "\n",
    "def manually_fill_nas(df):\n",
    "    '''\n",
    "    Function to fill samples with NAs in cancer_type (those without Classification) manually based on Owen\n",
    "    '''\n",
    "    samples_to_assign = [\"BS_AH3RVK53\", \"BS_JEZBA2EW\", \"BS_KQPCYZ2K\"]\n",
    "\n",
    "    # Assign values manually based on biosample_id\n",
    "    df.loc[df.index.isin(samples_to_assign), 'cancer_type'] = ['BTNOS', 'HGG', 'BTNOS']\n",
    "    return df\n",
    "df = manually_fill_nas(df)\n",
    "# df.to_csv(\"biosample_table.csv\")\n",
    "nas = df[df.cancer_type.isna()]\n",
    "print(len(nas))\n",
    "nas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d4ca5d96-85a0-4734-a701-5ccf9397f0a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Imports for Sunita's data\n",
    "def import_sunita_classifications(path='../data/combinedamplicons.xlsx'):\n",
    "    df = pd.read_excel(path)\n",
    "    df = df[[\"sample_ID\",\"cancer_type\"]]\n",
    "    df = df.drop_duplicates()\n",
    "    df = df.set_index(\"sample_ID\")\n",
    "    return df"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
