{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "ef5b06bb",
   "metadata": {
    "heading_collapsed": true
   },
   "source": [
    "# Create GCT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "10b71c21",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-05-19T22:41:47.426842Z",
     "start_time": "2023-05-19T22:41:47.422543Z"
    },
    "hidden": true
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import sys"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "9a150571",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-05-19T22:41:47.803682Z",
     "start_time": "2023-05-19T22:41:47.797818Z"
    },
    "hidden": true
   },
   "outputs": [],
   "source": [
    "data='/mnt/c/Users/ochapman/Documents/Mesirov/pedpancan_ecdna/data'\n",
    "sys.path.append(data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "3dc86a5a",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-05-19T22:41:48.176411Z",
     "start_time": "2023-05-19T22:41:48.166430Z"
    },
    "hidden": true
   },
   "outputs": [],
   "source": [
    "# source https://medium.com/intothegenomics/annotate-genes-and-genomic-coordinates-using-python-9259efa6ffc2\n",
    "def gene_info(x):\n",
    "# Extract gene names, gene_type, gene_status and level\n",
    "    g_name = list(filter(lambda x: 'gene_name' in x,  x.split(\";\")))[0].split(\"=\")[1]\n",
    "    g_type = list(filter(lambda x: 'gene_type' in x,  x.split(\";\")))[0].split(\"=\")[1]\n",
    "    g_leve = int(list(filter(lambda x: 'level' in x,  x.split(\";\")))[0].split(\"=\")[1])\n",
    "    return (g_name, g_type, g_leve)\n",
    "\n",
    "def get_protein_coding_genes():\n",
    "    path='/mnt/c/Users/ochapman/Documents/circos/genes/gencode.v33.basic.annotation.gff3'\n",
    "    gencode = pd.read_table(path, comment=\"#\",\n",
    "                            sep = \"\\t\", names = ['seqname', 'source', 'feature', 'start' , 'end', 'score', 'strand', 'frame', 'attribute'])\n",
    "    gencode_genes = gencode[(gencode.feature == \"gene\")][['seqname', 'start', 'end', 'attribute']].copy().reset_index().drop('index', axis=1)\n",
    "    gencode_genes[\"gene_name\"], gencode_genes[\"gene_type\"], gencode_genes[\"gene_level\"] = zip(*gencode_genes.attribute.apply(lambda x: gene_info(x)))\n",
    "    gencode_genes = gencode_genes[gencode_genes['gene_type'] == 'protein_coding'].reset_index().drop('index', axis=1)\n",
    "    gencode_genes = gencode_genes.sort_values(['gene_level', 'seqname'], ascending=True).drop_duplicates('gene_name', keep='first').reset_index().drop('index', axis=1)\n",
    "    return gencode_genes\n",
    "# Example usage\n",
    "#gene_ids = get_protein_coding_genes()\n",
    "#print(\"Protein-coding gene IDs:\", gene_ids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "e89583e8",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-05-19T22:41:49.631434Z",
     "start_time": "2023-05-19T22:41:49.622400Z"
    },
    "hidden": true
   },
   "outputs": [],
   "source": [
    "def write_gct_from_df(df,outfile):\n",
    "    rows,cols = df.shape\n",
    "    cols-=1\n",
    "    header=f'#1.2\\n{rows}\\t{cols}\\n'\n",
    "    with open(outfile,'w') as f:\n",
    "        f.write(header)\n",
    "    df.to_csv(outfile,sep='\\t',mode='a')\n",
    "    \n",
    "def gct_from_tsv(path):\n",
    "    print('reading gene expression matrix...')\n",
    "    gex = pd.read_csv(path,sep='\\t',index_col=0)\n",
    "    print('sorting by expression...')\n",
    "    rowsums = gex.sum(axis=1)\n",
    "    gex = gex.iloc[rowsums.argsort()]\n",
    "    gex = gex.iloc[::-1]\n",
    "    # drop duplicates\n",
    "    gex.insert(0,'NAME',gex.index.map(lambda x: '_'.join(x.split('_')[1:])))\n",
    "    l0=len(gex)\n",
    "    gex.drop_duplicates(subset=['NAME'],inplace=True)\n",
    "    print(f'Dropped {l0-len(gex)} duplicate gene entries...')\n",
    "    # drop nonexpressed genes\n",
    "    l0=len(gex)\n",
    "    gex = gex[gex.sum(axis=1) > 100]\n",
    "    print(f'Dropped {l0-len(gex)} nonexpressed genes...')\n",
    "    # only protein-coding genes\n",
    "    print('subset only protein-coding genes...')\n",
    "    gene_ids = get_protein_coding_genes()\n",
    "    gene_ids['gene_name']\n",
    "    l0 = len(gex)\n",
    "    gex = gex[gex.NAME.isin(gene_ids.gene_name.values)]\n",
    "    print(f'Dropped {l0-len(gex)} non-protein-coding genes...')\n",
    "    # set gene names to index\n",
    "    gex = gex.copy()\n",
    "    gex.insert(0,'DESCRIPTION',gex.index)\n",
    "    gex.set_index('NAME',inplace=True)\n",
    "    return gex"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "40389c70",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-05-19T22:43:00.795660Z",
     "start_time": "2023-05-19T22:41:53.291606Z"
    },
    "hidden": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "reading gene expression matrix...\n",
      "sorting by expression...\n",
      "Dropped 1216 duplicate gene entries...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_2702/3892606996.py:23: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.\n",
      "  gex = gex[gex.sum(axis=1) > 100]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dropped 29300 nonexpressed genes...\n",
      "subset only protein-coding genes...\n",
      "Dropped 13730 non-protein-coding genes...\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DESCRIPTION</th>\n",
       "      <th>7316-235</th>\n",
       "      <th>7316-1953</th>\n",
       "      <th>7316-4032</th>\n",
       "      <th>7316-216</th>\n",
       "      <th>7316-2566</th>\n",
       "      <th>7316-637</th>\n",
       "      <th>7316-7541</th>\n",
       "      <th>7316UP-1104</th>\n",
       "      <th>7316-9815</th>\n",
       "      <th>...</th>\n",
       "      <th>7316-393</th>\n",
       "      <th>7316-2720</th>\n",
       "      <th>7316-6884</th>\n",
       "      <th>7316-2899</th>\n",
       "      <th>7316-3204</th>\n",
       "      <th>7316-612</th>\n",
       "      <th>7316-5277</th>\n",
       "      <th>7316-3768</th>\n",
       "      <th>7316-9066.6</th>\n",
       "      <th>7316-1955</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NAME</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MT-ATP8</th>\n",
       "      <td>ENSG00000228253.1_MT-ATP8</td>\n",
       "      <td>2716.02</td>\n",
       "      <td>3275.83</td>\n",
       "      <td>15625.20</td>\n",
       "      <td>1421.59</td>\n",
       "      <td>1326.61</td>\n",
       "      <td>8632.86</td>\n",
       "      <td>22770.45</td>\n",
       "      <td>772.03</td>\n",
       "      <td>32027.25</td>\n",
       "      <td>...</td>\n",
       "      <td>1497.96</td>\n",
       "      <td>4120.38</td>\n",
       "      <td>63463.09</td>\n",
       "      <td>1551.12</td>\n",
       "      <td>1559.29</td>\n",
       "      <td>1861.08</td>\n",
       "      <td>39262.79</td>\n",
       "      <td>13591.82</td>\n",
       "      <td>17889.55</td>\n",
       "      <td>1230.14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-ND4L</th>\n",
       "      <td>ENSG00000212907.2_MT-ND4L</td>\n",
       "      <td>801.39</td>\n",
       "      <td>2278.33</td>\n",
       "      <td>9205.29</td>\n",
       "      <td>507.85</td>\n",
       "      <td>464.87</td>\n",
       "      <td>4686.80</td>\n",
       "      <td>13914.41</td>\n",
       "      <td>764.67</td>\n",
       "      <td>24669.84</td>\n",
       "      <td>...</td>\n",
       "      <td>830.92</td>\n",
       "      <td>3401.61</td>\n",
       "      <td>64374.74</td>\n",
       "      <td>596.30</td>\n",
       "      <td>667.76</td>\n",
       "      <td>746.16</td>\n",
       "      <td>27760.11</td>\n",
       "      <td>5440.01</td>\n",
       "      <td>13765.19</td>\n",
       "      <td>177.25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-CO1</th>\n",
       "      <td>ENSG00000198804.2_MT-CO1</td>\n",
       "      <td>2520.55</td>\n",
       "      <td>5609.30</td>\n",
       "      <td>6859.23</td>\n",
       "      <td>1415.10</td>\n",
       "      <td>1640.34</td>\n",
       "      <td>3750.15</td>\n",
       "      <td>9140.80</td>\n",
       "      <td>1977.47</td>\n",
       "      <td>15204.08</td>\n",
       "      <td>...</td>\n",
       "      <td>2812.45</td>\n",
       "      <td>13262.54</td>\n",
       "      <td>15432.33</td>\n",
       "      <td>1239.22</td>\n",
       "      <td>1402.13</td>\n",
       "      <td>1643.32</td>\n",
       "      <td>12802.08</td>\n",
       "      <td>13084.47</td>\n",
       "      <td>2440.27</td>\n",
       "      <td>1262.23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-ND3</th>\n",
       "      <td>ENSG00000198840.2_MT-ND3</td>\n",
       "      <td>1004.62</td>\n",
       "      <td>2023.32</td>\n",
       "      <td>7308.71</td>\n",
       "      <td>608.32</td>\n",
       "      <td>707.26</td>\n",
       "      <td>2614.15</td>\n",
       "      <td>7735.32</td>\n",
       "      <td>613.88</td>\n",
       "      <td>12914.83</td>\n",
       "      <td>...</td>\n",
       "      <td>866.30</td>\n",
       "      <td>3112.91</td>\n",
       "      <td>48507.84</td>\n",
       "      <td>729.49</td>\n",
       "      <td>699.20</td>\n",
       "      <td>821.11</td>\n",
       "      <td>18601.98</td>\n",
       "      <td>5000.52</td>\n",
       "      <td>6234.32</td>\n",
       "      <td>192.67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-CO3</th>\n",
       "      <td>ENSG00000198938.2_MT-CO3</td>\n",
       "      <td>1924.40</td>\n",
       "      <td>3669.34</td>\n",
       "      <td>8384.68</td>\n",
       "      <td>1209.34</td>\n",
       "      <td>1568.75</td>\n",
       "      <td>3960.95</td>\n",
       "      <td>8618.47</td>\n",
       "      <td>490.25</td>\n",
       "      <td>13683.26</td>\n",
       "      <td>...</td>\n",
       "      <td>2095.27</td>\n",
       "      <td>7914.33</td>\n",
       "      <td>27804.84</td>\n",
       "      <td>1129.44</td>\n",
       "      <td>1090.15</td>\n",
       "      <td>1433.76</td>\n",
       "      <td>13312.40</td>\n",
       "      <td>10692.34</td>\n",
       "      <td>7380.77</td>\n",
       "      <td>949.65</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 1923 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       DESCRIPTION  7316-235  7316-1953  7316-4032  7316-216  \\\n",
       "NAME                                                                           \n",
       "MT-ATP8  ENSG00000228253.1_MT-ATP8   2716.02    3275.83   15625.20   1421.59   \n",
       "MT-ND4L  ENSG00000212907.2_MT-ND4L    801.39    2278.33    9205.29    507.85   \n",
       "MT-CO1    ENSG00000198804.2_MT-CO1   2520.55    5609.30    6859.23   1415.10   \n",
       "MT-ND3    ENSG00000198840.2_MT-ND3   1004.62    2023.32    7308.71    608.32   \n",
       "MT-CO3    ENSG00000198938.2_MT-CO3   1924.40    3669.34    8384.68   1209.34   \n",
       "\n",
       "         7316-2566  7316-637  7316-7541  7316UP-1104  7316-9815  ...  \\\n",
       "NAME                                                             ...   \n",
       "MT-ATP8    1326.61   8632.86   22770.45       772.03   32027.25  ...   \n",
       "MT-ND4L     464.87   4686.80   13914.41       764.67   24669.84  ...   \n",
       "MT-CO1     1640.34   3750.15    9140.80      1977.47   15204.08  ...   \n",
       "MT-ND3      707.26   2614.15    7735.32       613.88   12914.83  ...   \n",
       "MT-CO3     1568.75   3960.95    8618.47       490.25   13683.26  ...   \n",
       "\n",
       "         7316-393  7316-2720  7316-6884  7316-2899  7316-3204  7316-612  \\\n",
       "NAME                                                                      \n",
       "MT-ATP8   1497.96    4120.38   63463.09    1551.12    1559.29   1861.08   \n",
       "MT-ND4L    830.92    3401.61   64374.74     596.30     667.76    746.16   \n",
       "MT-CO1    2812.45   13262.54   15432.33    1239.22    1402.13   1643.32   \n",
       "MT-ND3     866.30    3112.91   48507.84     729.49     699.20    821.11   \n",
       "MT-CO3    2095.27    7914.33   27804.84    1129.44    1090.15   1433.76   \n",
       "\n",
       "         7316-5277  7316-3768  7316-9066.6  7316-1955  \n",
       "NAME                                                   \n",
       "MT-ATP8   39262.79   13591.82     17889.55    1230.14  \n",
       "MT-ND4L   27760.11    5440.01     13765.19     177.25  \n",
       "MT-CO1    12802.08   13084.47      2440.27    1262.23  \n",
       "MT-ND3    18601.98    5000.52      6234.32     192.67  \n",
       "MT-CO3    13312.40   10692.34      7380.77     949.65  \n",
       "\n",
       "[5 rows x 1923 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "path='/mnt/c/Users/ochapman/Documents/Mesirov/pedpancan_ecdna/data/gex/CBTN-gex.tpm.tsv'\n",
    "gex = gct_from_tsv(path)\n",
    "gex.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "9858cede",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-05-18T23:00:26.915312Z",
     "start_time": "2023-05-18T23:00:03.232746Z"
    },
    "hidden": true
   },
   "outputs": [],
   "source": [
    "outfile='CBTN-gex.gct'\n",
    "write_gct_from_df(gex,outfile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "43f9389d",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-05-19T22:45:15.189609Z",
     "start_time": "2023-05-19T22:44:53.015033Z"
    },
    "hidden": true
   },
   "outputs": [],
   "source": [
    "gex.pop('DESCRIPTION')\n",
    "path = '/mnt/c/Users/ochapman/Documents/Mesirov/pedpancan_ecdna/data/gex/CBTN-gex-protein-coding.tpm.tsv'\n",
    "gex.to_csv(path,sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "44f491b5",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-05-18T21:55:58.341089Z",
     "start_time": "2023-05-18T21:55:57.826447Z"
    },
    "heading_collapsed": true
   },
   "source": [
    "# Clustering\n",
    "(Matrix too large)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "8fedcc03",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-05-19T19:21:12.126096Z",
     "start_time": "2023-05-19T19:21:09.564142Z"
    },
    "hidden": true
   },
   "outputs": [],
   "source": [
    "from scipy.cluster.hierarchy import dendrogram, linkage\n",
    "from scipy.stats import kendalltau\n",
    "from matplotlib import pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "15e2889b",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-05-19T19:09:01.831147Z",
     "start_time": "2023-05-19T19:09:01.824668Z"
    },
    "hidden": true
   },
   "outputs": [],
   "source": [
    "#df = gex.pop('DESCRIPTION')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ff5d1eb2",
   "metadata": {
    "ExecuteTime": {
     "start_time": "2023-05-19T19:25:09.643Z"
    },
    "hidden": true
   },
   "outputs": [],
   "source": [
    "# This didn't work\n",
    "ordinal_gex = gex.apply(lambda x: x.argsort(),axis=0)\n",
    "Z = linkage(ordinal_gex,method='weighted',metric=kendalltau)\n",
    "fig = plt.figure(figsize=(25, 10))\n",
    "dn = dendrogram(Z)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8f3756c2",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-05-19T22:38:07.774305Z",
     "start_time": "2023-05-19T22:38:07.764513Z"
    }
   },
   "source": [
    "# Metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "efebad48",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-05-19T22:38:59.696598Z",
     "start_time": "2023-05-19T22:38:59.175349Z"
    }
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import sys\n",
    "\n",
    "data='/mnt/c/Users/ochapman/Documents/Mesirov/pedpancan_ecdna/data'\n",
    "sys.path.append(data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "9ae4d13d",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-05-19T23:44:15.774595Z",
     "start_time": "2023-05-19T23:44:15.770458Z"
    }
   },
   "outputs": [],
   "source": [
    "def get_metadata():\n",
    "    master_table = '/mnt/c/Users/ochapman/Documents/Mesirov/pedpancan_ecdna/data/metadata/histologies.tsv'\n",
    "    df = pd.read_csv(master_table,sep='\\t',dtype='str')\n",
    "    return df\n",
    "def get_gex():\n",
    "    gex_table = '/mnt/c/Users/ochapman/Documents/Mesirov/pedpancan_ecdna/data/gex/CBTN-gex-protein-coding.tpm.tsv'\n",
    "    df = pd.read_csv(gex_table,sep='\\t',index_col=0)\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "1d3ca549",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-05-19T23:44:20.920441Z",
     "start_time": "2023-05-19T23:44:16.130937Z"
    }
   },
   "outputs": [],
   "source": [
    "meta = get_metadata()\n",
    "gex = get_gex()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "7a4b08d0",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-05-19T23:44:20.928516Z",
     "start_time": "2023-05-19T23:44:20.922401Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['7316-235', '7316-1953', '7316-4032', '7316-216', '7316-2566',\n",
       "       '7316-637', '7316-7541', '7316UP-1104', '7316-9815', '7316-3202',\n",
       "       ...\n",
       "       '7316-393', '7316-2720', '7316-6884', '7316-2899', '7316-3204',\n",
       "       '7316-612', '7316-5277', '7316-3768', '7316-9066.6', '7316-1955'],\n",
       "      dtype='object', length=1922)"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples = gex.columns\n",
    "samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "e6cd014a",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-05-19T23:49:38.708843Z",
     "start_time": "2023-05-19T23:49:38.578120Z"
    }
   },
   "outputs": [],
   "source": [
    "def preprocess_metadata(meta):\n",
    "    meta=meta.sort_values(['sample_id','experimental_strategy']) #array(['RNA-Seq', 'WGS', 'Methylation', 'WXS', 'Targeted Sequencing'],\n",
    "    meta=meta[meta.sample_type == 'Tumor']\n",
    "    meta=meta[meta.Kids_First_Participant_ID.map(lambda x:x.startswith('PT_'))]\n",
    "    Kids_First_Biospecimen_IDs=meta[['Kids_First_Biospecimen_ID','sample_id']].groupby('sample_id').agg({'Kids_First_Biospecimen_ID':';'.join})\n",
    "    meta=meta.drop_duplicates('sample_id',keep='first')\n",
    "    meta.set_index('sample_id',inplace=True)\n",
    "    meta.pop('Kids_First_Biospecimen_ID')\n",
    "    meta=meta.merge(Kids_First_Biospecimen_IDs,left_index=True,right_index=True)\n",
    "    meta=meta[meta.index.isin(samples)]\n",
    "    return meta\n",
    "submeta=preprocess_metadata(meta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "edc908dd",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-05-19T23:52:40.167491Z",
     "start_time": "2023-05-19T23:52:40.077244Z"
    }
   },
   "outputs": [],
   "source": [
    "path = '/mnt/c/Users/ochapman/Documents/Mesirov/pedpancan_ecdna/data/gex/CBTN-gex-histologies.tsv'\n",
    "submeta.to_csv(path,sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "b43c197b",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-05-19T23:45:35.614346Z",
     "start_time": "2023-05-19T23:45:35.605719Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['7316-2189.1', '7316-14.1', '7316-9066.1', '7316-8053.1', '7316-161.1',\n",
       "       '7316-6477.1', '7316-8051.1', '7316-4509.1', '7316-8053.2',\n",
       "       '7316-2186.1',\n",
       "       ...\n",
       "       '7316-1893.1', '7316-9062.2', '7316-278.1', '7316-7955.1',\n",
       "       '7316-3776.2', '7316-8023.2', '7316-2582.1', '7316-1455.1',\n",
       "       '7316-2176.2', '7316-9066.6'],\n",
       "      dtype='object', length=159)"
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples[~samples.isin(submeta.index)]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d71cf9d2",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-05-19T23:44:42.492278Z",
     "start_time": "2023-05-19T23:44:42.452812Z"
    }
   },
   "source": [
    "# Load data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "06f5aa9a",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-05-22T19:28:26.573375Z",
     "start_time": "2023-05-22T19:28:26.210967Z"
    }
   },
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f07ac772",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-05-22T19:28:07.202769Z",
     "start_time": "2023-05-22T19:28:07.194216Z"
    },
    "code_folding": [
     0,
     5
    ]
   },
   "outputs": [],
   "source": [
    "def get_gex():\n",
    "    gex_table = '/mnt/c/Users/ochapman/Documents/Mesirov/pedpancan_ecdna/data/gex/CBTN-gex-protein-coding.tpm.tsv'\n",
    "    df = pd.read_csv(gex_table,sep='\\t',index_col=0)\n",
    "    return df\n",
    "def get_metadata():\n",
    "    master_table = '/mnt/c/Users/ochapman/Documents/Mesirov/pedpancan_ecdna/data/gex/CBTN-gex-histologies.tsv'\n",
    "    df = pd.read_csv(master_table,sep='\\t',dtype='str')\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "65990a03",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-05-22T19:28:47.059782Z",
     "start_time": "2023-05-22T19:28:40.175033Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>7316-235</th>\n",
       "      <th>7316-1953</th>\n",
       "      <th>7316-4032</th>\n",
       "      <th>7316-216</th>\n",
       "      <th>7316-2566</th>\n",
       "      <th>7316-637</th>\n",
       "      <th>7316-7541</th>\n",
       "      <th>7316UP-1104</th>\n",
       "      <th>7316-9815</th>\n",
       "      <th>7316-3202</th>\n",
       "      <th>...</th>\n",
       "      <th>7316-393</th>\n",
       "      <th>7316-2720</th>\n",
       "      <th>7316-6884</th>\n",
       "      <th>7316-2899</th>\n",
       "      <th>7316-3204</th>\n",
       "      <th>7316-612</th>\n",
       "      <th>7316-5277</th>\n",
       "      <th>7316-3768</th>\n",
       "      <th>7316-9066.6</th>\n",
       "      <th>7316-1955</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NAME</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MT-ATP8</th>\n",
       "      <td>2716.02</td>\n",
       "      <td>3275.83</td>\n",
       "      <td>15625.20</td>\n",
       "      <td>1421.59</td>\n",
       "      <td>1326.61</td>\n",
       "      <td>8632.86</td>\n",
       "      <td>22770.45</td>\n",
       "      <td>772.03</td>\n",
       "      <td>32027.25</td>\n",
       "      <td>1657.10</td>\n",
       "      <td>...</td>\n",
       "      <td>1497.96</td>\n",
       "      <td>4120.38</td>\n",
       "      <td>63463.09</td>\n",
       "      <td>1551.12</td>\n",
       "      <td>1559.29</td>\n",
       "      <td>1861.08</td>\n",
       "      <td>39262.79</td>\n",
       "      <td>13591.82</td>\n",
       "      <td>17889.55</td>\n",
       "      <td>1230.14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-ND4L</th>\n",
       "      <td>801.39</td>\n",
       "      <td>2278.33</td>\n",
       "      <td>9205.29</td>\n",
       "      <td>507.85</td>\n",
       "      <td>464.87</td>\n",
       "      <td>4686.80</td>\n",
       "      <td>13914.41</td>\n",
       "      <td>764.67</td>\n",
       "      <td>24669.84</td>\n",
       "      <td>793.74</td>\n",
       "      <td>...</td>\n",
       "      <td>830.92</td>\n",
       "      <td>3401.61</td>\n",
       "      <td>64374.74</td>\n",
       "      <td>596.30</td>\n",
       "      <td>667.76</td>\n",
       "      <td>746.16</td>\n",
       "      <td>27760.11</td>\n",
       "      <td>5440.01</td>\n",
       "      <td>13765.19</td>\n",
       "      <td>177.25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-CO1</th>\n",
       "      <td>2520.55</td>\n",
       "      <td>5609.30</td>\n",
       "      <td>6859.23</td>\n",
       "      <td>1415.10</td>\n",
       "      <td>1640.34</td>\n",
       "      <td>3750.15</td>\n",
       "      <td>9140.80</td>\n",
       "      <td>1977.47</td>\n",
       "      <td>15204.08</td>\n",
       "      <td>2112.39</td>\n",
       "      <td>...</td>\n",
       "      <td>2812.45</td>\n",
       "      <td>13262.54</td>\n",
       "      <td>15432.33</td>\n",
       "      <td>1239.22</td>\n",
       "      <td>1402.13</td>\n",
       "      <td>1643.32</td>\n",
       "      <td>12802.08</td>\n",
       "      <td>13084.47</td>\n",
       "      <td>2440.27</td>\n",
       "      <td>1262.23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-ND3</th>\n",
       "      <td>1004.62</td>\n",
       "      <td>2023.32</td>\n",
       "      <td>7308.71</td>\n",
       "      <td>608.32</td>\n",
       "      <td>707.26</td>\n",
       "      <td>2614.15</td>\n",
       "      <td>7735.32</td>\n",
       "      <td>613.88</td>\n",
       "      <td>12914.83</td>\n",
       "      <td>545.43</td>\n",
       "      <td>...</td>\n",
       "      <td>866.30</td>\n",
       "      <td>3112.91</td>\n",
       "      <td>48507.84</td>\n",
       "      <td>729.49</td>\n",
       "      <td>699.20</td>\n",
       "      <td>821.11</td>\n",
       "      <td>18601.98</td>\n",
       "      <td>5000.52</td>\n",
       "      <td>6234.32</td>\n",
       "      <td>192.67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-CO3</th>\n",
       "      <td>1924.40</td>\n",
       "      <td>3669.34</td>\n",
       "      <td>8384.68</td>\n",
       "      <td>1209.34</td>\n",
       "      <td>1568.75</td>\n",
       "      <td>3960.95</td>\n",
       "      <td>8618.47</td>\n",
       "      <td>490.25</td>\n",
       "      <td>13683.26</td>\n",
       "      <td>1816.44</td>\n",
       "      <td>...</td>\n",
       "      <td>2095.27</td>\n",
       "      <td>7914.33</td>\n",
       "      <td>27804.84</td>\n",
       "      <td>1129.44</td>\n",
       "      <td>1090.15</td>\n",
       "      <td>1433.76</td>\n",
       "      <td>13312.40</td>\n",
       "      <td>10692.34</td>\n",
       "      <td>7380.77</td>\n",
       "      <td>949.65</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TP53TG3F</th>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.59</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IZUMO1R</th>\n",
       "      <td>0.04</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.01</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LIPK</th>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ASB15</th>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.05</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KRTAP10-5</th>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17341 rows × 1922 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           7316-235  7316-1953  7316-4032  7316-216  7316-2566  7316-637  \\\n",
       "NAME                                                                       \n",
       "MT-ATP8     2716.02    3275.83   15625.20   1421.59    1326.61   8632.86   \n",
       "MT-ND4L      801.39    2278.33    9205.29    507.85     464.87   4686.80   \n",
       "MT-CO1      2520.55    5609.30    6859.23   1415.10    1640.34   3750.15   \n",
       "MT-ND3      1004.62    2023.32    7308.71    608.32     707.26   2614.15   \n",
       "MT-CO3      1924.40    3669.34    8384.68   1209.34    1568.75   3960.95   \n",
       "...             ...        ...        ...       ...        ...       ...   \n",
       "TP53TG3F       0.00       0.00       0.00      0.00       0.00      0.00   \n",
       "IZUMO1R        0.04       0.00       0.00      0.01       0.00      0.00   \n",
       "LIPK           0.00       0.00       0.00      0.00       0.00      0.00   \n",
       "ASB15          0.00       0.00       0.00      0.05       0.00      0.00   \n",
       "KRTAP10-5      0.00       0.00       0.00      0.00       0.00      0.00   \n",
       "\n",
       "           7316-7541  7316UP-1104  7316-9815  7316-3202  ...  7316-393  \\\n",
       "NAME                                                     ...             \n",
       "MT-ATP8     22770.45       772.03   32027.25    1657.10  ...   1497.96   \n",
       "MT-ND4L     13914.41       764.67   24669.84     793.74  ...    830.92   \n",
       "MT-CO1       9140.80      1977.47   15204.08    2112.39  ...   2812.45   \n",
       "MT-ND3       7735.32       613.88   12914.83     545.43  ...    866.30   \n",
       "MT-CO3       8618.47       490.25   13683.26    1816.44  ...   2095.27   \n",
       "...              ...          ...        ...        ...  ...       ...   \n",
       "TP53TG3F        0.00         0.00       0.02       0.00  ...      0.00   \n",
       "IZUMO1R         0.00         0.00       0.01       0.01  ...      0.00   \n",
       "LIPK            0.00         0.00       0.00       0.00  ...      0.00   \n",
       "ASB15           0.00         0.00       0.01       0.05  ...      0.00   \n",
       "KRTAP10-5       0.00         0.00       0.00       0.00  ...      0.00   \n",
       "\n",
       "           7316-2720  7316-6884  7316-2899  7316-3204  7316-612  7316-5277  \\\n",
       "NAME                                                                         \n",
       "MT-ATP8      4120.38   63463.09    1551.12    1559.29   1861.08   39262.79   \n",
       "MT-ND4L      3401.61   64374.74     596.30     667.76    746.16   27760.11   \n",
       "MT-CO1      13262.54   15432.33    1239.22    1402.13   1643.32   12802.08   \n",
       "MT-ND3       3112.91   48507.84     729.49     699.20    821.11   18601.98   \n",
       "MT-CO3       7914.33   27804.84    1129.44    1090.15   1433.76   13312.40   \n",
       "...              ...        ...        ...        ...       ...        ...   \n",
       "TP53TG3F        0.00       0.59       0.00       0.00      0.00       0.00   \n",
       "IZUMO1R         0.00       0.00       0.03       0.00      0.00       0.00   \n",
       "LIPK            0.00       0.00       0.00       0.00      0.00       0.00   \n",
       "ASB15           0.00       0.00       0.00       0.00      0.00       0.00   \n",
       "KRTAP10-5       0.00       0.00       0.00       0.00      0.00       0.00   \n",
       "\n",
       "           7316-3768  7316-9066.6  7316-1955  \n",
       "NAME                                          \n",
       "MT-ATP8     13591.82     17889.55    1230.14  \n",
       "MT-ND4L      5440.01     13765.19     177.25  \n",
       "MT-CO1      13084.47      2440.27    1262.23  \n",
       "MT-ND3       5000.52      6234.32     192.67  \n",
       "MT-CO3      10692.34      7380.77     949.65  \n",
       "...              ...          ...        ...  \n",
       "TP53TG3F        0.00         0.00       0.00  \n",
       "IZUMO1R         0.01         0.00       0.00  \n",
       "LIPK            0.00         0.00       0.00  \n",
       "ASB15           0.00         0.00       0.00  \n",
       "KRTAP10-5       0.01         0.00       0.00  \n",
       "\n",
       "[17341 rows x 1922 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gex = get_gex()\n",
    "gex"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "77572502",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-05-22T19:28:55.124234Z",
     "start_time": "2023-05-22T19:28:55.062548Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>aliquot_id</th>\n",
       "      <th>Kids_First_Participant_ID</th>\n",
       "      <th>experimental_strategy</th>\n",
       "      <th>sample_type</th>\n",
       "      <th>composition</th>\n",
       "      <th>tumor_descriptor</th>\n",
       "      <th>primary_site</th>\n",
       "      <th>reported_gender</th>\n",
       "      <th>race</th>\n",
       "      <th>...</th>\n",
       "      <th>dkfz_v12_methylation_mgmt_estimated</th>\n",
       "      <th>molecular_subtype</th>\n",
       "      <th>integrated_diagnosis</th>\n",
       "      <th>Notes</th>\n",
       "      <th>harmonized_diagnosis</th>\n",
       "      <th>molecular_subtype_methyl</th>\n",
       "      <th>broad_histology</th>\n",
       "      <th>short_histology</th>\n",
       "      <th>cancer_group</th>\n",
       "      <th>Kids_First_Biospecimen_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>7316-10</td>\n",
       "      <td>1114560</td>\n",
       "      <td>PT_K8ZV7APT</td>\n",
       "      <td>Methylation</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>Solid Tissue</td>\n",
       "      <td>Initial CNS Tumor</td>\n",
       "      <td>Spinal Cord- Lumbar/Thecal Sac</td>\n",
       "      <td>Female</td>\n",
       "      <td>Black or African American</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00278215388511586</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Neurofibroma/Plexiform</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Tumor of cranial and paraspinal nerves</td>\n",
       "      <td>Neurofibroma</td>\n",
       "      <td>Neurofibroma/Plexiform</td>\n",
       "      <td>BS_458X47PJ;BS_GDHH6T5A;BS_1RFBH1SP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>7316-100</td>\n",
       "      <td>1112049</td>\n",
       "      <td>PT_6TZR2DH1</td>\n",
       "      <td>Methylation</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>Solid Tissue</td>\n",
       "      <td>Initial CNS Tumor</td>\n",
       "      <td>Suprasellar/Hypothalamic/Pituitary</td>\n",
       "      <td>Female</td>\n",
       "      <td>Black or African American</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00398842742564261</td>\n",
       "      <td>CRANIO, ADAM</td>\n",
       "      <td>Adamantinomatous craniopharyngioma</td>\n",
       "      <td>Updated via OpenPedCan subtyping</td>\n",
       "      <td>Adamantinomatous craniopharyngioma</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Tumors of sellar region</td>\n",
       "      <td>Craniopharyngioma</td>\n",
       "      <td>Adamantinomatous Craniopharyngioma</td>\n",
       "      <td>BS_V4PBW4WH;BS_BHR08WGW;BS_SFZ3A07S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>7316-101</td>\n",
       "      <td>1112551</td>\n",
       "      <td>PT_CWD717Q0</td>\n",
       "      <td>Methylation</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>Solid Tissue</td>\n",
       "      <td>Initial CNS Tumor</td>\n",
       "      <td>Temporal Lobe</td>\n",
       "      <td>Male</td>\n",
       "      <td>Black or African American</td>\n",
       "      <td>...</td>\n",
       "      <td>0.02092587213451</td>\n",
       "      <td>GNG, other MAPK</td>\n",
       "      <td>Ganglioglioma, other MAPK</td>\n",
       "      <td>Updated via OpenPedCan subtyping</td>\n",
       "      <td>Ganglioglioma, other MAPK</td>\n",
       "      <td>GNT, MAPK</td>\n",
       "      <td>Low-grade astrocytic tumor</td>\n",
       "      <td>Ganglioglioma</td>\n",
       "      <td>Ganglioglioma</td>\n",
       "      <td>BS_Q37DRN94;BS_QV51J756;BS_4RS1SC48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>7316-1017</td>\n",
       "      <td>1251886</td>\n",
       "      <td>PT_0DWRY9ZX</td>\n",
       "      <td>Methylation</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>Solid Tissue</td>\n",
       "      <td>Recurrence</td>\n",
       "      <td>Frontal Lobe</td>\n",
       "      <td>Male</td>\n",
       "      <td>White</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0220976046770112</td>\n",
       "      <td>MB, Group4</td>\n",
       "      <td>Medulloblastoma, group 4</td>\n",
       "      <td>Subtype based on prediction;Updated via OpenPe...</td>\n",
       "      <td>Medulloblastoma, group 4</td>\n",
       "      <td>MB, Group4</td>\n",
       "      <td>Embryonal tumor</td>\n",
       "      <td>Medulloblastoma</td>\n",
       "      <td>Medulloblastoma</td>\n",
       "      <td>BS_24KYA5PH;BS_FB6ADVMD;BS_SD4SNCCX;BS_ZGSTZCYN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>7316-1038</td>\n",
       "      <td>1116046</td>\n",
       "      <td>PT_STRDJQ01</td>\n",
       "      <td>Methylation</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>Solid Tissue</td>\n",
       "      <td>Initial CNS Tumor</td>\n",
       "      <td>Cerebellum/Posterior Fossa</td>\n",
       "      <td>Female</td>\n",
       "      <td>Reported Unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00977657395132449</td>\n",
       "      <td>MB, SHH</td>\n",
       "      <td>Medulloblastoma, SHH-activated</td>\n",
       "      <td>Subtype based on prediction;Updated via OpenPe...</td>\n",
       "      <td>Medulloblastoma, SHH-activated</td>\n",
       "      <td>MB, SHH</td>\n",
       "      <td>Embryonal tumor</td>\n",
       "      <td>Medulloblastoma</td>\n",
       "      <td>Medulloblastoma</td>\n",
       "      <td>BS_Y0CPEZ59;BS_VPBMDMQX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1758</th>\n",
       "      <td>7316-9975</td>\n",
       "      <td>1277236</td>\n",
       "      <td>PT_GFERC3PT</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>Solid Tissue</td>\n",
       "      <td>Initial CNS Tumor</td>\n",
       "      <td>Skull</td>\n",
       "      <td>Female</td>\n",
       "      <td>White</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Langerhans Cell histiocytosis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Histiocytic tumor</td>\n",
       "      <td>Langerhans cell histiocytosis</td>\n",
       "      <td>Langerhans Cell histiocytosis</td>\n",
       "      <td>BS_4RAHHT12;BS_4PFG3PX6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1759</th>\n",
       "      <td>7316UP-1104</td>\n",
       "      <td>1105139</td>\n",
       "      <td>PT_X5SGG610</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>Solid Tissue</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>Cerebrum</td>\n",
       "      <td>Female</td>\n",
       "      <td>White/Caucasian</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGG, To be classified</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Astrocytoma</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Diffuse astrocytic and oligodendroglial tumor</td>\n",
       "      <td>HGAT</td>\n",
       "      <td>Astrocytoma</td>\n",
       "      <td>BS_8E8JKP01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1760</th>\n",
       "      <td>7316UP-2035</td>\n",
       "      <td>1141315</td>\n",
       "      <td>PT_6RZ363TJ</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>Solid Tissue</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>Cerebrum</td>\n",
       "      <td>Female</td>\n",
       "      <td>White/Caucasian</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGG, To be classified</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Glioblastoma</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Diffuse astrocytic and oligodendroglial tumor</td>\n",
       "      <td>HGAT</td>\n",
       "      <td>Glioblastoma</td>\n",
       "      <td>BS_0C00AYTE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1761</th>\n",
       "      <td>7316UP-310</td>\n",
       "      <td>1105138</td>\n",
       "      <td>PT_DB281Z4K</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>Solid Tissue</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>Cerebrum</td>\n",
       "      <td>Female</td>\n",
       "      <td>White/Caucasian</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGG, To be classified</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Glioblastoma</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Diffuse astrocytic and oligodendroglial tumor</td>\n",
       "      <td>HGAT</td>\n",
       "      <td>Glioblastoma</td>\n",
       "      <td>BS_0XHT9W4Q</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1762</th>\n",
       "      <td>7316UP-759</td>\n",
       "      <td>1141316</td>\n",
       "      <td>PT_Y34NXFN6</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>Solid Tissue</td>\n",
       "      <td>Recurrence</td>\n",
       "      <td>Cerebrum</td>\n",
       "      <td>Male</td>\n",
       "      <td>White/Caucasian</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGG, To be classified</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Astrocytoma;Oligoastrocytoma</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Diffuse astrocytic and oligodendroglial tumor</td>\n",
       "      <td>HGAT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BS_67V4F080</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1763 rows × 55 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        sample_id aliquot_id Kids_First_Participant_ID experimental_strategy  \\\n",
       "0         7316-10    1114560               PT_K8ZV7APT           Methylation   \n",
       "1        7316-100    1112049               PT_6TZR2DH1           Methylation   \n",
       "2        7316-101    1112551               PT_CWD717Q0           Methylation   \n",
       "3       7316-1017    1251886               PT_0DWRY9ZX           Methylation   \n",
       "4       7316-1038    1116046               PT_STRDJQ01           Methylation   \n",
       "...           ...        ...                       ...                   ...   \n",
       "1758    7316-9975    1277236               PT_GFERC3PT               RNA-Seq   \n",
       "1759  7316UP-1104    1105139               PT_X5SGG610               RNA-Seq   \n",
       "1760  7316UP-2035    1141315               PT_6RZ363TJ               RNA-Seq   \n",
       "1761   7316UP-310    1105138               PT_DB281Z4K               RNA-Seq   \n",
       "1762   7316UP-759    1141316               PT_Y34NXFN6               RNA-Seq   \n",
       "\n",
       "     sample_type   composition   tumor_descriptor  \\\n",
       "0          Tumor  Solid Tissue  Initial CNS Tumor   \n",
       "1          Tumor  Solid Tissue  Initial CNS Tumor   \n",
       "2          Tumor  Solid Tissue  Initial CNS Tumor   \n",
       "3          Tumor  Solid Tissue         Recurrence   \n",
       "4          Tumor  Solid Tissue  Initial CNS Tumor   \n",
       "...          ...           ...                ...   \n",
       "1758       Tumor  Solid Tissue  Initial CNS Tumor   \n",
       "1759       Tumor  Solid Tissue      Primary Tumor   \n",
       "1760       Tumor  Solid Tissue      Primary Tumor   \n",
       "1761       Tumor  Solid Tissue      Primary Tumor   \n",
       "1762       Tumor  Solid Tissue         Recurrence   \n",
       "\n",
       "                            primary_site reported_gender  \\\n",
       "0         Spinal Cord- Lumbar/Thecal Sac          Female   \n",
       "1     Suprasellar/Hypothalamic/Pituitary          Female   \n",
       "2                          Temporal Lobe            Male   \n",
       "3                           Frontal Lobe            Male   \n",
       "4             Cerebellum/Posterior Fossa          Female   \n",
       "...                                  ...             ...   \n",
       "1758                               Skull          Female   \n",
       "1759                            Cerebrum          Female   \n",
       "1760                            Cerebrum          Female   \n",
       "1761                            Cerebrum          Female   \n",
       "1762                            Cerebrum            Male   \n",
       "\n",
       "                           race  ... dkfz_v12_methylation_mgmt_estimated  \\\n",
       "0     Black or African American  ...                 0.00278215388511586   \n",
       "1     Black or African American  ...                 0.00398842742564261   \n",
       "2     Black or African American  ...                    0.02092587213451   \n",
       "3                         White  ...                  0.0220976046770112   \n",
       "4              Reported Unknown  ...                 0.00977657395132449   \n",
       "...                         ...  ...                                 ...   \n",
       "1758                      White  ...                                 NaN   \n",
       "1759            White/Caucasian  ...                                 NaN   \n",
       "1760            White/Caucasian  ...                                 NaN   \n",
       "1761            White/Caucasian  ...                                 NaN   \n",
       "1762            White/Caucasian  ...                                 NaN   \n",
       "\n",
       "          molecular_subtype                integrated_diagnosis  \\\n",
       "0                       NaN                                 NaN   \n",
       "1              CRANIO, ADAM  Adamantinomatous craniopharyngioma   \n",
       "2           GNG, other MAPK           Ganglioglioma, other MAPK   \n",
       "3                MB, Group4            Medulloblastoma, group 4   \n",
       "4                   MB, SHH      Medulloblastoma, SHH-activated   \n",
       "...                     ...                                 ...   \n",
       "1758                    NaN                                 NaN   \n",
       "1759  HGG, To be classified                                 NaN   \n",
       "1760  HGG, To be classified                                 NaN   \n",
       "1761  HGG, To be classified                                 NaN   \n",
       "1762  HGG, To be classified                                 NaN   \n",
       "\n",
       "                                                  Notes  \\\n",
       "0                                                   NaN   \n",
       "1                      Updated via OpenPedCan subtyping   \n",
       "2                      Updated via OpenPedCan subtyping   \n",
       "3     Subtype based on prediction;Updated via OpenPe...   \n",
       "4     Subtype based on prediction;Updated via OpenPe...   \n",
       "...                                                 ...   \n",
       "1758                                                NaN   \n",
       "1759                                                NaN   \n",
       "1760                                                NaN   \n",
       "1761                                                NaN   \n",
       "1762                                                NaN   \n",
       "\n",
       "                    harmonized_diagnosis molecular_subtype_methyl  \\\n",
       "0                 Neurofibroma/Plexiform                      NaN   \n",
       "1     Adamantinomatous craniopharyngioma                      NaN   \n",
       "2              Ganglioglioma, other MAPK                GNT, MAPK   \n",
       "3               Medulloblastoma, group 4               MB, Group4   \n",
       "4         Medulloblastoma, SHH-activated                  MB, SHH   \n",
       "...                                  ...                      ...   \n",
       "1758       Langerhans Cell histiocytosis                      NaN   \n",
       "1759                         Astrocytoma                      NaN   \n",
       "1760                        Glioblastoma                      NaN   \n",
       "1761                        Glioblastoma                      NaN   \n",
       "1762        Astrocytoma;Oligoastrocytoma                      NaN   \n",
       "\n",
       "                                    broad_histology  \\\n",
       "0            Tumor of cranial and paraspinal nerves   \n",
       "1                           Tumors of sellar region   \n",
       "2                        Low-grade astrocytic tumor   \n",
       "3                                   Embryonal tumor   \n",
       "4                                   Embryonal tumor   \n",
       "...                                             ...   \n",
       "1758                              Histiocytic tumor   \n",
       "1759  Diffuse astrocytic and oligodendroglial tumor   \n",
       "1760  Diffuse astrocytic and oligodendroglial tumor   \n",
       "1761  Diffuse astrocytic and oligodendroglial tumor   \n",
       "1762  Diffuse astrocytic and oligodendroglial tumor   \n",
       "\n",
       "                    short_histology                        cancer_group  \\\n",
       "0                      Neurofibroma              Neurofibroma/Plexiform   \n",
       "1                 Craniopharyngioma  Adamantinomatous Craniopharyngioma   \n",
       "2                     Ganglioglioma                       Ganglioglioma   \n",
       "3                   Medulloblastoma                     Medulloblastoma   \n",
       "4                   Medulloblastoma                     Medulloblastoma   \n",
       "...                             ...                                 ...   \n",
       "1758  Langerhans cell histiocytosis       Langerhans Cell histiocytosis   \n",
       "1759                           HGAT                         Astrocytoma   \n",
       "1760                           HGAT                        Glioblastoma   \n",
       "1761                           HGAT                        Glioblastoma   \n",
       "1762                           HGAT                                 NaN   \n",
       "\n",
       "                            Kids_First_Biospecimen_ID  \n",
       "0                 BS_458X47PJ;BS_GDHH6T5A;BS_1RFBH1SP  \n",
       "1                 BS_V4PBW4WH;BS_BHR08WGW;BS_SFZ3A07S  \n",
       "2                 BS_Q37DRN94;BS_QV51J756;BS_4RS1SC48  \n",
       "3     BS_24KYA5PH;BS_FB6ADVMD;BS_SD4SNCCX;BS_ZGSTZCYN  \n",
       "4                             BS_Y0CPEZ59;BS_VPBMDMQX  \n",
       "...                                               ...  \n",
       "1758                          BS_4RAHHT12;BS_4PFG3PX6  \n",
       "1759                                      BS_8E8JKP01  \n",
       "1760                                      BS_0C00AYTE  \n",
       "1761                                      BS_0XHT9W4Q  \n",
       "1762                                      BS_67V4F080  \n",
       "\n",
       "[1763 rows x 55 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta = get_metadata()\n",
    "meta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d1fcd939",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:py3]",
   "language": "python",
   "name": "conda-env-py3-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
